// @anna-ai/cli/dist/test/index.js
//#region src/test/host-api-acl.ts
const NAMESPACED = {
	tools: true,
	chat: true,
	artifact: true,
	llm: true,
	fs: true,
	storage: true,
	prefs: true,
	window: true
};
function deriveAcl(manifest) {
	const ha = manifest.ui?.host_api ?? {};
	const allowed = new Set();
	const allowedTools = new Set();
	let toolsWildcard = false;
	for (const ref of ha.tools ?? []) {
		if (ref === "required:*" || ref === "optional:*" || ref === "*") {
			toolsWildcard = true;
			continue;
		}
		const bare = ref.includes(":") ? ref.split(":", 2)[1] : ref;
		allowedTools.add(bare);
	}
	if (allowedTools.size === 0 && !toolsWildcard) {
		for (const r of [...manifest.required_executas ?? [], ...manifest.optional_executas ?? []]) if (r.tool_id) allowedTools.add(r.tool_id);
	}
	if (allowedTools.size > 0 || toolsWildcard) allowed.add("tools.invoke");
	for (const ns of Object.keys(NAMESPACED)) {
		if (ns === "tools") continue;
		for (const method of ha[ns] ?? []) allowed.add(`${ns}.${method}`);
	}
	return {
		allowed,
		allowedTools,
		toolsWildcard
	};
}
function isToolAllowed(acl, toolId) {
	if (acl.toolsWildcard) return true;
	return acl.allowedTools.has(toolId);
}

//#endregion
//#region src/test/runtime.ts
const DEFAULT_TOKEN_TTL_MS = 5 * 60 * 1e3;
function randHex(n) {
	let s = "";
	for (let i = 0; i < n; i += 1) s += Math.floor(Math.random() * 16).toString(16);
	return s;
}
var CallLogImpl = class {
	records = [];
	push(r) {
		this.records.push(r);
	}
	all() {
		return this.records.slice();
	}
	byNs(prefix) {
		return this.records.filter((r) => `${r.ns}.${r.method}` === prefix || r.ns === prefix);
	}
	last() {
		return this.records[this.records.length - 1] ?? null;
	}
	lastOf(prefix) {
		const xs = this.byNs(prefix);
		return xs[xs.length - 1] ?? null;
	}
	clear() {
		this.records.length = 0;
	}
};
var EventBusImpl = class {
	listeners = new Map();
	wildcard = new Set();
	emit(name, payload) {
		for (const fn of this.listeners.get(name) ?? []) try {
			fn(payload);
		} catch {}
		for (const fn of this.wildcard) try {
			fn(name, payload);
		} catch {}
	}
	on(name, fn) {
		let set = this.listeners.get(name);
		if (!set) {
			set = new Set();
			this.listeners.set(name, set);
		}
		set.add(fn);
		return () => set.delete(fn);
	}
	/** Internal: tap all events for the auth.refresh handler etc. */
	onAny(fn) {
		this.wildcard.add(fn);
		return () => this.wildcard.delete(fn);
	}
};
/**
* Default mock implementations for namespaces that have safe in-memory
* behavior. Bundle authors can override any of these via `mocks`.
*/
function makeDefaultMocks(state) {
	return {
		"storage.get": (args) => {
			const key = args.key ?? "";
			return state.storage.get(key) ?? null;
		},
		"storage.set": (args) => {
			const { key, value } = args;
			state.storage.set(key, value);
			return null;
		},
		"storage.delete": (args) => {
			const key = args.key ?? "";
			state.storage.delete(key);
			return null;
		},
		"storage.list": (args) => {
			const { prefix = "", cursor, limit = 100 } = args ?? {};
			const keys = [...state.storage.keys()].filter((k) => k.startsWith(prefix)).sort();
			const startIdx = cursor ? keys.findIndex((k) => k > cursor) : 0;
			const slice = startIdx >= 0 ? keys.slice(startIdx, startIdx + limit) : [];
			const next_cursor = startIdx >= 0 && startIdx + limit < keys.length ? slice[slice.length - 1] : null;
			return {
				items: slice.map((k) => ({
					key: k,
					etag: null,
					size_bytes: JSON.stringify(state.storage.get(k) ?? null).length,
					metadata: null,
					tags: null,
					updated_at: null
				})),
				next_cursor
			};
		},
		"prefs.get": (args) => {
			const key = args.key ?? "";
			return state.prefs.get(key) ?? null;
		},
		"prefs.set": (args) => {
			const { key, value } = args;
			state.prefs.set(key, value);
			return null;
		},
		"chat.write_message": () => null,
		"window.set_title": () => null,
		"llm.complete": (args) => {
			const messages = args.messages ?? [];
			const firstUser = messages.find((m) => m?.role === "user");
			const content = firstUser?.content;
			let text = "";
			if (typeof content === "string") text = content;
			else if (content && typeof content === "object" && "text" in content) text = String(content.text ?? "");
			const truncated = text.length > 60 ? text.slice(0, 60) + "…" : text;
			const out = `<MOCK LLM: ${truncated}>`;
			const inputTokens = Math.max(1, Math.ceil(text.length / 4));
			const outputTokens = Math.max(1, Math.ceil(out.length / 4));
			return {
				role: "assistant",
				content: {
					type: "text",
					text: out
				},
				model: "mock-model",
				stopReason: "endTurn",
				usage: {
					inputTokens,
					outputTokens,
					totalTokens: inputTokens + outputTokens
				}
			};
		}
	};
}
async function mountBundle(opts) {
	const acl = deriveAcl(opts.manifest);
	const calls = new CallLogImpl();
	const events = new EventBusImpl();
	const startedAt = Date.now();
	const ttl = opts.tokenTtlMs ?? DEFAULT_TOKEN_TTL_MS;
	const wid = opts.wid ?? `harness-${randHex(8)}`;
	let token = opts.token ?? randHex(24);
	const state = {
		storage: new Map(),
		prefs: new Map()
	};
	const mocks = {
		...makeDefaultMocks(state),
		...opts.mocks ?? {}
	};
	events.onAny((name, payload) => {
		if (name === "auth.refresh") {
			const next = payload?.token;
			if (typeof next === "string" && next) token = next;
		}
	});
	let seq = 0;
	async function dispatch(ns, method, args) {
		seq += 1;
		const rec = {
			seq,
			t: Date.now() - startedAt,
			ns,
			method,
			args,
			outcome: null
		};
		calls.push(rec);
		const key = `${ns}.${method}`;
		if (!acl.allowed.has(key)) {
			rec.outcome = "denied";
			rec.errorCode = "DENIED";
			rec.errorMessage = `host API '${key}' not declared in manifest.ui.host_api`;
			throw new HostApiError(rec.errorCode, rec.errorMessage);
		}
		if (ns === "tools" && method === "invoke") {
			const toolId = args.tool_id ?? "";
			if (!isToolAllowed(acl, toolId)) {
				rec.outcome = "denied";
				rec.errorCode = "DENIED";
				rec.errorMessage = `tool '${toolId}' not declared in manifest.ui.host_api.tools`;
				throw new HostApiError(rec.errorCode, rec.errorMessage);
			}
		}
		const handler = mocks[key];
		if (!handler) {
			rec.outcome = "error";
			rec.errorCode = "NO_MOCK";
			rec.errorMessage = `no mock registered for '${key}' (add via opts.mocks or harness.mock)`;
			throw new HostApiError(rec.errorCode, rec.errorMessage);
		}
		try {
			const result = await Promise.resolve(handler(args, {
				ns,
				method
			}));
			rec.outcome = "ok";
			rec.result = result;
			return result;
		} catch (e) {
			rec.outcome = "error";
			const err = e;
			rec.errorCode = err.code ?? "MOCK_ERROR";
			rec.errorMessage = err.message ?? String(e);
			throw e;
		}
	}
	const facet = (ns) => (method) => (args) => dispatch(ns, method, args ?? {});
	const runtime = {
		get hello() {
			return {
				wid,
				t: token,
				user_id: opts.userId ?? 1
			};
		},
		call: dispatch,
		on: (name, fn) => events.on(name, fn),
		tools: { invoke: (args) => dispatch("tools", "invoke", args) },
		storage: {
			get: (key) => dispatch("storage", "get", { key }),
			set: (key, value) => dispatch("storage", "set", {
				key,
				value
			})
		},
		chat: { write_message: (text, optsArg) => dispatch("chat", "write_message", {
			text,
			...optsArg ?? {}
		}) },
		artifact: { create: (args) => dispatch("artifact", "create", args) },
		llm: { complete: (args) => dispatch("llm", "complete", args) },
		fs: new Proxy({}, { get: (_t, prop) => facet("fs")(prop) }),
		prefs: {
			get: (key) => dispatch("prefs", "get", { key }),
			set: (key, value) => dispatch("prefs", "set", {
				key,
				value
			})
		},
		window: { set_title: (title) => dispatch("window", "set_title", { title }) }
	};
	return {
		runtime,
		calls,
		events,
		acl,
		mock(key, handler) {
			mocks[key] = handler;
		},
		wait: (ms) => new Promise((res) => setTimeout(res, ms))
	};
}
/** Error class thrown for ACL denials or missing mocks. */
var HostApiError = class extends Error {
	code;
	constructor(code, message) {
		super(message);
		this.name = "HostApiError";
		this.code = code;
	}
};

//#endregion
export { HostApiError, deriveAcl, isToolAllowed, mountBundle };
