// @anna-ai/app-runtime/dist/index.js
/**
 * Anna App Runtime SDK — v0.8.0
 *
 * v0.8.0 — Adds `anna.agent.session.refresh(args)` plus the same `.refresh()`
 *   sugar on the AgentSession handle, and surfaces session lifecycle metadata
 *   (`expires_at`, `max_lifetime_at`, `idle_ttl_seconds`, `session_expires_in`)
 *   on every AgentSession. `refresh` re-mints the short-lived capability token
 *   for an EXISTING session (by `app_session_uuid`) AND slides its idle window,
 *   so a session survives long idle gaps, iframe reloads and host restarts:
 *     await anna.agent.session.refresh({ app_session_uuid })   // raw RPC
 *     await sess.refresh()                                     // handle sugar
 *   `refresh` is identity-authed (it does NOT require a live token), which is
 *   what lets it recover a session whose token already lapsed
 *   (`APP_SESSION_TOKEN_EXPIRED`). It fails with `APP_SESSION_EXPIRED` /
 *   `APP_SESSION_REVOKED` only when the underlying session is genuinely gone.
 *   The handle's lifecycle fields are refreshed in place from the response.
 *   Coordinated bump: wire method `agent/session.refresh` —
 *   dispatcher_version 0.10.0 + schema 0.10.0 + anna-app-core 0.9.0.
 *   Additive: old hosts return `-32601 Method not found`.
 *
 * v0.7.0 — Adds `anna.agent.session.attach(app_session_uuid)` — returns an
 *   AgentSession handle bound to an EXISTING session you already know the
 *   uuid of (e.g. one returned by `anna.agent.session.list()` after an
 *   iframe reload / in another tab). Unlike `anna.agent.session(opts)` it
 *   does NOT mint a new session — it is a pure client-side wrapper, so
 *   there is no new wire method and no dispatcher/schema change. The
 *   returned handle exposes the same streaming `.run()`, `.cancel()`,
 *   `.history()`, `.delete()` sugar over the existing `agent.session.*`
 *   RPCs (all of which are keyed by `app_session_uuid`). Accepts either a
 *   bare uuid string or a `{ app_session_uuid, ... }` info object (e.g. a
 *   row from `list()`), which also seeds submode/label metadata.
 *
 * v0.6.0 — Adds `anna.agent.session.list(args)` — account/app-scoped
 *   enumeration of the current app's active agent sessions. Lives only on
 *   the `agent.session` namespace (raw RPC), NOT on the per-session
 *   AgentSession wrapper, since it returns a list across sessions. Args:
 *     { include_expired?: boolean, limit?: number (1-100, default 50) }
 *   Returns { sessions: [{ app_session_uuid, kind, submode,
 *     fixed_client_id, label, created_at, last_active_at, expires_at }] }.
 *   Lets an app re-attach to / clean up sessions after an iframe reload or
 *   crash. Coordinated bump: dispatcher_version 0.9.0 + schema 0.9.0 +
 *   anna-app-core 0.8.0. No breaking changes.
 *
 * v0.4.1 — Per-namespace default RPC timeouts and per-call override.
 *   image.* / upload.* / agent.* / llm.* used to share the 30s default,
 *   but real LLM / image-gen latency easily exceeds that on high-res
 *   prompts (Seedream 5.0 @ 2048x2048 can take 60-120s). New defaults:
 *     image:   180s   upload:  120s   agent:  300s   llm:  180s
 *     storage:  30s   chat/window/...: 30s (unchanged)
 *   Any call can override via a second arg, e.g.
 *     await anna.image.generate(args, { timeoutMs: 240000 })
 *
 * Loaded by every Anna App iframe as a native ES module via
 *   `import { AnnaAppRuntime } from "/static/anna-apps/_sdk/latest/index.js"`.
 * Provides a typed proxy over the postMessage RPC bridge to the host
 * (Anna dashboard). All host calls go through the host page (parent
 * window), which validates them against the app's manifest ACL before
 * forwarding to the backend.
 *
 * v0.4.0 — BREAKING: removed the cross-origin HTTP shim and the
 * `AnnaHostError` class. `anna.image.{generate,edit}` and
 * `anna.upload.{inline,negotiate,confirm}` now go through the same
 * postMessage → `/api/v1/anna-apps/runtime/rpc` dispatcher channel as
 * every other namespace (symmetric with `anna.storage.*` /
 * `anna.llm.complete`). The dispatcher mints the underlying
 * `app_session_token` server-side; the iframe never holds capability
 * credentials. Errors arrive as the standard RPC error shape:
 * `Error` with `.code` (e.g. `APP_ERR_NOT_GRANTED`,
 * `APP_ERR_QUOTA_EXCEEDED`, `not_implemented`) and `.details`. Apps
 * that previously branched on `err.status === 403/415/429` must rewrite
 * to branch on `err.code` (see VERSIONS.md changelog for the mapping).
 *
 * v0.3.1 — packaged as pure ESM (no IIFE / window-global). Consumers must
 * use `<script type="module">` + `import`. Legacy `<script src=...>`
 * loading is no longer supported.
 *
 * Envelope: { kind:"req"|"res"|"event", id, ns, method, args | result | error }
 *
 * Usage inside the iframe:
 *
 *     import { AnnaAppRuntime } from "/static/anna-apps/_sdk/latest/index.js";
 *     const anna = await AnnaAppRuntime.connect();
 *     // Mint-only tool_ids (tool-{handle}-{slug}-{uniq}) carry no separator,
 *     // so the target plugin method is supplied via `method`:
 *     const out  = await anna.tools.invoke({
 *       tool_id: "tool-twitter-handle-twitter-tool-abcd1234",
 *       method: "post_tweet",
 *       args: { text: "hi" },
 *     });
 *     await anna.window.set_title({ title: "My App" });
 *     anna.on("entry_payload", (p) => console.log(p));
 */

// ---- Internals ----
  function uuid() {
    return (
      Date.now().toString(36) + "-" + Math.random().toString(36).slice(2, 10)
    );
  }

  // Per-namespace default RPC timeout (ms). Picked to comfortably cover
  // upstream provider P99 latency (host bridge fetch is unbounded; backend
  // httpx is 180s for image). Apps can override per-call via the second
  // arg: `anna.image.generate(args, { timeoutMs: 240000 })`.
  const DEFAULT_TIMEOUTS_BY_NS = {
    image: 180000,   // image gen/edit; matches src/tools/image_generation_tool.py httpx=180s
    upload: 120000,  // large multipart / negotiate+confirm
    agent: 300000,   // session.run can loop through model+tools many turns
    llm: 180000,     // long completions / large context
    // `tools.invoke` host bound = 65s plugin + 2s grace + 3s SDK margin.
    // Apps requesting longer per call must pass { timeoutMs }; bounded
    // server-side by ANNA_APP_TOOLS_INVOKE_TIMEOUT_MAX_MS.
    tools: 70000,
    // `files.*` covers R2 presign + DB write (upload_init/finalize); the
    // browser→R2 PUT itself is direct (not via this SDK), but the host
    // round-trip for finalize needs DB+R2 head ~ up to a few seconds.
    files: 60000,
  };
  const DEFAULT_TIMEOUT_MS = 30000;

  // Build a target proxy that exposes namespace.method(args, opts?) calls.
  // The optional second arg accepts `{ timeoutMs }`; missing → namespace
  // default → 30000 fallback.
  function buildNamespace(rt, ns) {
    return new Proxy(
      {},
      {
        get(_, method) {
          if (typeof method !== "string") return undefined;
          return (args, opts) => rt.call(ns, method, args || {}, opts);
        },
      }
    );
  }

  // ---- Agent namespace (v0.2.0) ----
  //
  // Exposes both raw RPCs (`anna.agent.session.create({submode, ...})`,
  // `anna.agent.session.run({...})`, etc.) AND a higher-level helper
  // `anna.agent.session(opts)` that returns an AgentSession instance
  // for ergonomic use in apps:
  //
  //     const sess = await anna.agent.session({ submode: "auto" });
  //     for await (const ev of sess.run({ content: "hi" })) {
  //       if (ev.event === "model_token") render(ev.text);
  //     }
  function buildAgentNamespace(rt) {
    const sessionRpcs = new Proxy(
      {},
      {
        get(_, method) {
          if (typeof method !== "string") return undefined;
          return (args, opts) =>
            rt.call("agent", "session." + method, args || {}, opts);
        },
      }
    );
    const ns = {
      // Higher-level: mint and return an AgentSession wrapper.
      session: async (opts = {}) => {
        const result = await rt.call("agent", "session.create", opts);
        return new AgentSession(rt, result);
      },
    };
    // Also expose raw RPC sub-methods on the function — apps can do
    // `anna.agent.session.create(...)` if they want to manage state manually.
    // `list` is account/app-scoped (enumerate this app's sessions), so it
    // lives only here, not on the per-session AgentSession wrapper. `refresh`
    // is identity-authed (re-mints token + slides idle window by uuid) and is
    // also exposed as `.refresh()` sugar on the AgentSession handle.
    for (const m of ["create", "run", "cancel", "history", "delete", "list", "refresh"]) {
      ns.session[m] = sessionRpcs[m];
    }
    // `attach` re-wraps an EXISTING session (known uuid) as an AgentSession
    // handle WITHOUT minting a new one — pure client-side, no RPC. Accepts a
    // bare uuid string or an info object (e.g. a row from `list()`).
    ns.session.attach = (info) => {
      const norm =
        typeof info === "string" ? { app_session_uuid: info } : info || {};
      if (!norm.app_session_uuid) {
        throw new Error("agent.session.attach: app_session_uuid is required");
      }
      return new AgentSession(rt, norm);
    };
    return ns;
  }

  // ---- HTTP-backed namespaces removed in v0.4.0 ----
  //
  // `anna.image.*` and `anna.upload.*` are now plain postMessage RPCs
  // (see the dispatcher's `image.*` / `upload.*` handlers in
  // anna-app-core 0.3.0+). The dispatcher mints the underlying
  // app_session_token in-process; the iframe never holds capability
  // credentials. Errors arrive on the standard RPC envelope with
  // `.code` set to the facade's `APP_ERR_*` value.

  // AsyncIterable view over `rpc.stream` event chunks. Reassembles by `seq`
  // so out-of-order arrivals are normalised; back-pressure is bounded only
  // by the iframe's heap (events are small JSON blobs).
  class AgentRunStream {
    constructor(streamId, runId) {
      this.streamId = streamId;
      this.runId = runId;
      this._queue = [];     // ordered by seq
      this._waiters = [];   // resolvers awaiting next event
      this._buffer = new Map(); // seq -> payload (out-of-order holding)
      this._nextSeq = 1;
      this._done = false;
    }
    _push(frame) {
      const seq = Number(frame.seq || 0);
      const payload = frame.payload;
      const done = !!frame.done;
      if (seq && seq !== this._nextSeq) {
        this._buffer.set(seq, { payload, done });
        return;
      }
      this._emit(payload, done);
      this._nextSeq += 1;
      // Drain consecutive buffered frames.
      while (this._buffer.has(this._nextSeq)) {
        const next = this._buffer.get(this._nextSeq);
        this._buffer.delete(this._nextSeq);
        this._emit(next.payload, next.done);
        this._nextSeq += 1;
      }
    }
    _emit(payload, done) {
      if (payload != null) this._queue.push(payload);
      if (done) this._done = true;
      while (this._waiters.length && (this._queue.length || this._done)) {
        const w = this._waiters.shift();
        if (this._queue.length) w({ value: this._queue.shift(), done: false });
        else w({ value: undefined, done: true });
      }
    }
    [Symbol.asyncIterator]() {
      const self = this;
      return {
        next() {
          if (self._queue.length) {
            return Promise.resolve({ value: self._queue.shift(), done: false });
          }
          if (self._done) return Promise.resolve({ value: undefined, done: true });
          return new Promise((resolve) => self._waiters.push(resolve));
        },
        return() {
          self._done = true;
          while (self._waiters.length) self._waiters.shift()({ value: undefined, done: true });
          return Promise.resolve({ value: undefined, done: true });
        },
      };
    }
  }

  // Higher-level agent session wrapper. Holds the dispatcher-managed
  // app_session_uuid; its methods are sugar around the raw `agent.session.*`
  // RPCs.
  class AgentSession {
    constructor(rt, info) {
      this._rt = rt;
      this.app_session_uuid = info.app_session_uuid;
      this.expires_in = info.expires_in;
      this.submode = info.submode;
      this.fixed_client_id = info.fixed_client_id || null;
      this.granted_tools = info.granted_tools || [];
      // Lifecycle metadata (v0.8.0). `expires_at` is the authoritative idle
      // deadline; `max_lifetime_at` is the absolute cap; `idle_ttl_seconds` is
      // how far each activity slides the window; `session_expires_in` is the
      // seconds-until-`expires_at` snapshot at mint/refresh time. All are
      // refreshed in place by `.refresh()`.
      this._applyLifecycle(info);
    }
    // Copy lifecycle fields from a create/refresh response onto the handle.
    _applyLifecycle(info) {
      if (info == null) return;
      if ("expires_in" in info) this.expires_in = info.expires_in;
      this.expires_at = info.expires_at ?? this.expires_at ?? null;
      this.max_lifetime_at =
        info.max_lifetime_at ?? this.max_lifetime_at ?? null;
      this.idle_ttl_seconds =
        info.idle_ttl_seconds ?? this.idle_ttl_seconds ?? null;
      this.session_expires_in =
        info.session_expires_in ?? this.session_expires_in ?? null;
    }
    run(opts = {}) {
      const args = Object.assign(
        { app_session_uuid: this.app_session_uuid },
        opts
      );
      const stream = new AgentRunStream(null, args.run_id || null);
      // Kick off the RPC; once the host returns {stream_id, run_id} we
      // register the stream so subsequent rpc.stream events flow into it.
      this._rt.call("agent", "session.run", args).then(
        (res) => {
          stream.streamId = res.stream_id || stream.streamId;
          stream.runId = res.run_id || stream.runId;
          if (stream.streamId) this._rt._registerStream(stream.streamId, stream);
          // If host returns done immediately (e.g. cached error), close.
          if (!stream.streamId) stream._emit(null, true);
        },
        (err) => {
          stream._emit({ event: "error", code: err.code, message: err.message }, true);
        }
      );
      return stream;
    }
    cancel(runId) {
      return this._rt.call("agent", "session.cancel", {
        app_session_uuid: this.app_session_uuid,
        run_id: runId,
      });
    }
    history(opts = {}) {
      return this._rt.call("agent", "session.history", Object.assign(
        { app_session_uuid: this.app_session_uuid }, opts
      ));
    }
    // Re-mint the capability token AND slide the idle window for this session.
    // Identity-authed — works even if the previous token already lapsed
    // (recovers from APP_SESSION_TOKEN_EXPIRED). Updates the handle's
    // lifecycle fields in place from the response and returns it.
    async refresh(opts = {}) {
      const res = await this._rt.call("agent", "session.refresh", Object.assign(
        { app_session_uuid: this.app_session_uuid }, opts
      ));
      this._applyLifecycle(res || {});
      return res;
    }
    delete() {
      return this._rt.call("agent", "session.delete", {
        app_session_uuid: this.app_session_uuid,
      });
    }
  }

  class AnnaAppRuntime {
    constructor(opts) {
      this.windowUuid = opts.windowUuid;
      this.token = opts.token;
      this.targetOrigin = opts.targetOrigin || "*"; // host enforces sandbox
      this.parent = opts.parent || window.parent;
      this._pending = new Map();   // id -> {resolve, reject, retries}
      this._handlers = new Map();  // event kind -> Set<fn>
      this._closed = false;
      this._handshake = null;
      this._heartbeatTimer = null;
      this.capabilities = null;
      this.viewMeta = null;
      this.entryPayload = null;
      this.runtimeState = null;
      this.geometry = null;

      // namespace proxies
      this.tools = buildNamespace(this, "tools");
      this.chat = buildNamespace(this, "chat");
      this.artifact = buildNamespace(this, "artifact");
      this.llm = buildNamespace(this, "llm");
      this.fs = buildNamespace(this, "fs");
      this.storage = buildNamespace(this, "storage");
      this.prefs = buildNamespace(this, "prefs");
      this.window = buildNamespace(this, "window");

      // ---- agent namespace (v0.2.0) ----
      // Exposes session.{create,run,cancel,history,delete}; SDK also offers
      // a higher-level `anna.agent.session(opts)` helper that returns an
      // AgentSession bound to a fresh AnnaAppSession.
      this.agent = buildAgentNamespace(this);

      // ---- image / upload namespaces (v0.4.0; postMessage) ----
      // anna.image.generate({prompt, n?, size?, reference_image_urls?, metadata?})
      // anna.image.edit({image_url, prompt, n?, size?, mask_url?, metadata?})
      // anna.upload.inline({filename, mime_type, content_b64, purpose?, metadata?})
      // anna.upload.negotiate({filename, mime_type, size_bytes, purpose?, metadata?})
      // anna.upload.confirm({r2_key})
      this.image = buildNamespace(this, "image");
      this.upload = buildNamespace(this, "upload");

      // ---- files namespace (APS object/blob API; v0.5.x) ----
      // anna.files.upload_init({path, content_type, size, ttl_seconds?, metadata?, tags?, if_match?})
      // anna.files.upload_finalize({path, etag, size_bytes, metadata?})
      // anna.files.download_url({path, ttl_seconds?, disposition?})
      // anna.files.list({prefix?, cursor?, limit?})
      // anna.files.delete({path, if_match?})
      this.files = buildNamespace(this, "files");

      // Reassembly buffers for `rpc.stream` events. Map<stream_id, AgentRunStream>.
      this._streams = new Map();
      this.on("rpc.stream", (payload) => this._onStreamFrame(payload || {}));

      window.addEventListener("message", (ev) => this._onMessage(ev));
      window.addEventListener("error", (ev) => {
        // best-effort error reporting
        this._post({
          kind: "req",
          id: uuid(),
          ns: "window",
          method: "report_error",
          args: {
            error: { message: String(ev.message), source: ev.filename, line: ev.lineno },
          },
        });
      });
      // Final flush on unload (best-effort)
      window.addEventListener("pagehide", () => {
        this._closed = true;
      });
    }

    _post(envelope) {
      try {
        this.parent.postMessage(
          Object.assign({ wid: this.windowUuid, t: this.token }, envelope),
          this.targetOrigin
        );
      } catch (e) {
        console.error("[anna-sdk] postMessage failed", e);
      }
    }

    _onMessage(ev) {
      const msg = ev.data;
      if (!msg || typeof msg !== "object") return;
      if (msg.wid && msg.wid !== this.windowUuid) return;

      if (msg.kind === "res") {
        const pending = this._pending.get(msg.id);
        if (!pending) return;
        this._pending.delete(msg.id);
        if (msg.error) {
          const err = new Error(msg.error.message || "RPC error");
          err.code = msg.error.code;
          err.details = msg.error.details;
          pending.reject(err);
        } else {
          pending.resolve(msg.result);
        }
        return;
      }

      if (msg.kind === "event") {
        if (msg.event === "auth.refresh" && msg.token) {
          this.token = msg.token;
          return;
        }
        const set = this._handlers.get(msg.event);
        if (set) for (const fn of set) try { fn(msg.payload); } catch (_) {}
      }
    }

    on(kind, fn) {
      let set = this._handlers.get(kind);
      if (!set) { set = new Set(); this._handlers.set(kind, set); }
      set.add(fn);
      return () => set.delete(fn);
    }

    call(ns, method, args, opts) {
      if (this._closed) return Promise.reject(new Error("runtime closed"));
      const timeoutMs =
        (opts && Number.isFinite(opts.timeoutMs) ? opts.timeoutMs : null) ??
        DEFAULT_TIMEOUTS_BY_NS[ns] ??
        DEFAULT_TIMEOUT_MS;
      const id = uuid();
      return new Promise((resolve, reject) => {
        const timer = setTimeout(() => {
          this._pending.delete(id);
          reject(
            new Error(
              `RPC ${ns}.${method} timed out after ${timeoutMs}ms` +
                ` (pass { timeoutMs } as the 2nd arg to override)`
            )
          );
        }, timeoutMs);
        this._pending.set(id, {
          resolve: (v) => { clearTimeout(timer); resolve(v); },
          reject: (e) => { clearTimeout(timer); reject(e); },
        });
        this._post({ kind: "req", id, ns, method, args: args || {} });
      });
    }

    // (HTTP shim removed in v0.4.0 — image/upload now use postMessage.)

    _startHeartbeat() {
      this._heartbeatTimer = setInterval(() => {
        if (this._closed) return;
        this.call("window", "ready", {}).catch(() => {});
      }, 10000);
    }

    _onStreamFrame(payload) {
      const sid = payload.stream_id;
      if (!sid) return;
      const stream = this._streams.get(sid);
      if (!stream) {
        // Frame may have arrived before `agent.session.run` resolved with
        // its stream_id. Buffer for a short window — the registration
        // path drains this on `_registerStream`.
        if (!this._earlyFrames) this._earlyFrames = new Map();
        let q = this._earlyFrames.get(sid);
        if (!q) { q = []; this._earlyFrames.set(sid, q); }
        q.push(payload);
        // Cap to avoid unbounded growth on truly orphaned streams.
        if (q.length > 256) q.shift();
        return;
      }
      stream._push(payload);
      if (payload.done) this._streams.delete(sid);
    }

    _registerStream(streamId, stream) {
      this._streams.set(streamId, stream);
      // Drain any frames that arrived before the RPC reply.
      const queued = this._earlyFrames && this._earlyFrames.get(streamId);
      if (queued && queued.length) {
        this._earlyFrames.delete(streamId);
        for (const p of queued) {
          stream._push(p);
          if (p.done) {
            this._streams.delete(streamId);
            break;
          }
        }
      }
    }

    static async connect(options = {}) {
      // Discover credentials from URL
      const params = new URLSearchParams(location.search);
      const wid = options.windowUuid || params.get("wid");
      const token = options.token || params.get("t");
      if (!wid || !token) {
        throw new Error("[anna-sdk] missing wid/t URL parameters");
      }
      const rt = new AnnaAppRuntime({ windowUuid: wid, token, parent: window.parent });
      const hello = await rt.call("window", "hello", {});
      rt.capabilities = hello.capabilities;
      rt.viewMeta = hello.view_meta;
      rt.entryPayload = hello.entry_payload || {};
      rt.runtimeState = hello.runtime_state || {};
      rt.geometry = hello.geometry || {};
      rt._startHeartbeat();
      // signal ready (host may show iframe at this point)
      rt.call("window", "ready", {}).catch(() => {});
      return rt;
    }
  }

// Convenience static references (mirrors pre-0.3.1 layout where these
// hung off the AnnaAppRuntime global). `AnnaHostError` was removed in
// v0.4.0; apps should now branch on the standard RPC error `.code`.
AnnaAppRuntime.AgentSession = AgentSession;
AnnaAppRuntime.AgentRunStream = AgentRunStream;

export { AnnaAppRuntime, AgentSession, AgentRunStream };
export default AnnaAppRuntime;
