// @anna-ai/cli/dist/_lifecycle-shared-CUK2CE76.js
import { canonicalHost, getAccount } from "./credentials-BTv2IfUZ.js";
import { CliError, NexusClient, ScopeError } from "./client-D-_z1ALk.js";
import { dim, red, yellow } from "kleur/colors";

//#region src/commands/_lifecycle-shared.ts
function resolveClient(opts = {}) {
	const envHost = process.env.ANNA_APP_HOST;
	const envPat = process.env.ANNA_APP_PAT;
	if (envHost && envPat) {
		const host = canonicalHost(opts.host ?? envHost);
		return {
			client: new NexusClient({
				host,
				pat: envPat
			}),
			host,
			source: "env"
		};
	}
	const acc = getAccount(opts.account ?? opts.host);
	if (!acc) throw new CliError("no PAT on disk and ANNA_APP_PAT not set — run `anna-app login --host <url>` first.", 2);
	return {
		client: new NexusClient({
			host: acc.host,
			pat: acc.pat
		}),
		host: acc.host,
		source: "credentials"
	};
}
/**
* Print an error from a CLI command and return its exit code.
*
* `ScopeError` gets the friendly multi-line hint from design §6.2.
* `CliError` prints its message + carries a numeric exit code.
* Everything else maps to exit 1 with the bare message.
*/
function reportError(e) {
	if (e instanceof ScopeError) {
		const need = e.needed.length > 0 ? e.needed.join(", ") : "(unknown)";
		const have = e.have.length > 0 ? e.have.join(", ") : "(none)";
		const missing = e.missing.length > 0 ? e.missing.join(", ") : need;
		console.error(red(`✗ this command needs PAT scope: ${missing}`) + "\n" + dim(`  current PAT scopes: [${have}]`) + "\n" + yellow(`  Mint a new token with:
    anna-app login --host <nexus-url> --scopes ${need}`) + "\n" + dim("  Existing tokens cannot be upgraded — old token continues to work for its current scopes."));
		return e.exitCode;
	}
	if (e instanceof CliError) {
		console.error(red(`✗ ${e.message}`));
		return e.exitCode;
	}
	const msg = e?.message ?? String(e);
	console.error(red(`✗ ${msg}`));
	return 1;
}
/** Wrap an async command body so unexpected throws become exit codes. */
async function withErrorHandling(fn) {
	try {
		return await fn();
	} catch (e) {
		return reportError(e);
	}
}

//#endregion
export { reportError, resolveClient, withErrorHandling };
