// @anna-ai/cli/dist/agent-J2bELWfv.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { hostOf, mintAppSession, requireAccount, withCode } from "./dev-account-WXEjIE16.js";
import { resolve } from "node:path";
import { existsSync, readFileSync } from "node:fs";

//#region src/executa/agent.ts
/** JSON-RPC error codes from matrix/src/executa/protocol.py. */
const AGENT_ERR_NOT_GRANTED = -32041;
const AGENT_ERR_SESSION_NOT_FOUND = -32042;
const AGENT_ERR_INVALID_REQUEST = -32043;
const AGENT_ERR_PROVIDER_ERROR = -32046;
const AGENT_ERR_RUN_TOO_LARGE = -32047;
const MAX_FRAMES_PER_RUN = 4096;
var AgentBridge = class {
	mocks = [];
	/** uuid → minted session token (one entry per `session.create`). */
	sessions = new Map();
	constructor(opts) {
		this.opts = opts;
		if (opts.mode === "mock" && opts.mockFile) {
			const path = resolve(opts.mockFile);
			if (existsSync(path)) for (const line of readFileSync(path, "utf8").split(/\r?\n/)) {
				if (!line.trim() || line.startsWith("#")) continue;
				try {
					this.mocks.push(JSON.parse(line));
				} catch {}
			}
		}
	}
	async call(method, params) {
		if (this.opts.mode === "off") throw withCode(new Error("agent reverse-RPC disabled — rerun without `--no-agent`, or pass `--mock-agent <fixture>` / `--agent <real>`"), AGENT_ERR_NOT_GRANTED);
		if (this.opts.mode === "mock") return this.mockCall(method, params);
		return this.realCall(method, params);
	}
	mockCall(method, params) {
		const shortName = method.startsWith("agent/") ? method.slice(6) : method;
		const content = collectContent(params);
		const matched = this.mocks.find((m) => m.ns === "agent" && m.method === shortName && (!m.match?.contentIncludes || content.includes(m.match.contentIncludes))) ?? this.mocks.find((m) => m.ns === "agent" && m.method === shortName);
		if (matched && matched.result !== void 0) return matched.result;
		if (shortName === "session.create") return {
			app_session_uuid: `aps_mock_${Math.random().toString(16).slice(2, 10)}`,
			expires_in: 600,
			agent_submode: "auto",
			granted_tools: []
		};
		if (shortName === "session.run") {
			const text = String(params.content ?? "");
			return {
				run_id: "mock_run",
				stream_id: "strm_mock",
				frames: [{
					event: "final",
					text: `(mock) echoing: ${text}`,
					seq: 0,
					stream_id: "strm_mock"
				}],
				final: {
					event: "final",
					text: `(mock) echoing: ${text}`
				},
				frame_count: 1
			};
		}
		if (shortName === "session.cancel") return { cancelled: true };
		if (shortName === "session.history") return {
			messages: [],
			cursor: null
		};
		if (shortName === "session.delete") return { deleted: true };
		if (shortName === "complete") return {
			role: "assistant",
			content: {
				type: "text",
				text: "(mock) no fixture matched"
			},
			model: "mock-model",
			stopReason: "endTurn"
		};
		throw withCode(new Error(`unknown agent method: ${method}`), AGENT_ERR_INVALID_REQUEST);
	}
	requireAppSlug() {
		if (!this.opts.appSlug) throw withCode(new Error("agent bridge has no app_slug — pass `--app-slug <slug>` so nexus can attribute the spend"), -32602);
		return this.opts.appSlug;
	}
	async realCall(method, params) {
		const shortName = method.startsWith("agent/") ? method.slice(6) : method;
		switch (shortName) {
			case "session.create": return this.sessionCreate(params);
			case "session.run": return this.sessionRun(params);
			case "session.cancel": return this.sessionCancel(params);
			case "session.history": return this.sessionHistory(params);
			case "session.delete": return this.sessionDelete(params);
			case "complete": return this.complete(params);
			default: throw withCode(new Error(`unknown agent method: ${method}`), AGENT_ERR_INVALID_REQUEST);
		}
	}
	async sessionCreate(params) {
		const acc = requireAccount(this.opts.account);
		const submode = params.agent_submode ?? "auto";
		if (submode !== "auto" && submode !== "fixed") throw withCode(new Error("agent_submode must be 'auto' or 'fixed'"), AGENT_ERR_INVALID_REQUEST);
		const fixedClientId = params.fixed_client_id ?? params.fixedClientId ?? null;
		if (submode === "fixed" && !fixedClientId) throw withCode(new Error("fixed_client_id required when agent_submode='fixed'"), AGENT_ERR_INVALID_REQUEST);
		const minted = await mintAppSession(acc.host, {
			pat: acc.pat,
			kind: "agent",
			app_slug: this.requireAppSlug(),
			submode,
			fixed_client_id: fixedClientId,
			label: params.label ?? "anna-app executa dev",
			quota_caps: params.quota_caps ?? null,
			ttl_seconds: params.ttl_seconds ?? 600
		});
		this.sessions.set(minted.app_session_uuid, {
			appSessionToken: minted.app_session_token,
			appSessionUuid: minted.app_session_uuid,
			expiresAt: Math.floor(Date.now() / 1e3) + (minted.expires_in || 600) - 30,
			submode
		});
		return {
			app_session_uuid: minted.app_session_uuid,
			thread_id: minted.app_session_uuid,
			agent_submode: submode,
			fixed_client_id: fixedClientId,
			expires_in: minted.expires_in,
			granted_tools: []
		};
	}
	getSession(params) {
		const uuid = String(params.app_session_uuid ?? "");
		if (!uuid.startsWith("aps_")) throw withCode(new Error("'app_session_uuid' must start with 'aps_'"), AGENT_ERR_INVALID_REQUEST);
		const ms = this.sessions.get(uuid);
		if (!ms) throw withCode(new Error(`app_session_uuid=${uuid} not in dev cache (expired or never created)`), AGENT_ERR_SESSION_NOT_FOUND);
		if (ms.expiresAt < Math.floor(Date.now() / 1e3)) {
			this.sessions.delete(uuid);
			throw withCode(new Error(`app_session_uuid=${uuid} expired — call session.create again`), AGENT_ERR_SESSION_NOT_FOUND);
		}
		return ms;
	}
	async sessionRun(params) {
		const acc = requireAccount(this.opts.account);
		const ms = this.getSession(params);
		const content = params.content;
		if (typeof content !== "string" || !content.trim()) throw withCode(new Error("'content' must be a non-empty string"), AGENT_ERR_INVALID_REQUEST);
		const runId = params.run_id ?? `run_${Math.random().toString(16).slice(2, 18)}`;
		const streamId = `strm_${Math.random().toString(16).slice(2, 18)}`;
		const body = {
			app_session_uuid: ms.appSessionUuid,
			content,
			run_id: runId,
			stream: true
		};
		if (params.attachments !== void 0) body.attachments = params.attachments;
		if (params.recursion_limit !== void 0) body.recursion_limit = params.recursion_limit;
		if (params.allowed_tools !== void 0) body.allowed_tools = params.allowed_tools;
		if (params.system !== void 0) body.system = params.system;
		const mp = params.modelPreferences ?? params.model_preferences;
		if (mp !== void 0) body.model_preferences = mp;
		const url = `${hostOf(acc)}/api/v1/copilot/app/agent`;
		const res = await fetch(url, {
			method: "POST",
			headers: {
				"content-type": "application/json",
				authorization: `Bearer ${ms.appSessionToken}`,
				accept: "text/event-stream"
			},
			body: JSON.stringify(body)
		});
		if (!res.ok || !res.body) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`agent/session.run failed: HTTP ${res.status}: ${text}`), AGENT_ERR_PROVIDER_ERROR);
		}
		const frames = [];
		let final = null;
		const textAcc = [];
		const reader = res.body.getReader();
		const decoder = new TextDecoder();
		let buf = "";
		while (true) {
			const { value, done } = await reader.read();
			if (done) break;
			buf += decoder.decode(value, { stream: true });
			let idx;
			while ((idx = buf.indexOf("\n\n")) >= 0) {
				const frameRaw = buf.slice(0, idx);
				buf = buf.slice(idx + 2);
				const payload = parseSseFrame(frameRaw);
				if (payload && typeof payload === "object") {
					const frame = payload;
					if (frame.seq === void 0) frame.seq = frames.length;
					if (frame.stream_id === void 0) frame.stream_id = streamId;
					frames.push(frame);
					if (frame.event === "delta" && typeof frame.text === "string") textAcc.push(frame.text);
					else if (frame.event === "final") final = frame;
					if (frames.length > MAX_FRAMES_PER_RUN) throw withCode(new Error(`agent run exceeded ${MAX_FRAMES_PER_RUN} frames`), AGENT_ERR_RUN_TOO_LARGE);
				}
			}
		}
		if (buf.trim().length > 0) {
			const payload = parseSseFrame(buf);
			if (payload && typeof payload === "object") {
				const frame = payload;
				if (frame.seq === void 0) frame.seq = frames.length;
				if (frame.stream_id === void 0) frame.stream_id = streamId;
				frames.push(frame);
				if (frame.event === "final") final = frame;
			}
		}
		if (final === null && textAcc.length > 0) {
			final = {
				event: "final",
				text: textAcc.join(""),
				stream_id: streamId,
				seq: frames.length,
				synthesized: true
			};
			frames.push(final);
		}
		return {
			run_id: runId,
			stream_id: streamId,
			frames,
			final,
			frame_count: frames.length
		};
	}
	async sessionCancel(params) {
		const acc = requireAccount(this.opts.account);
		const ms = this.getSession(params);
		const runId = params.run_id;
		if (typeof runId !== "string" || !runId.trim()) throw withCode(new Error("'run_id' required"), AGENT_ERR_INVALID_REQUEST);
		const url = `${hostOf(acc)}/api/v1/copilot/app/agent/cancel`;
		const res = await fetch(url, {
			method: "POST",
			headers: {
				"content-type": "application/json",
				authorization: `Bearer ${ms.appSessionToken}`
			},
			body: JSON.stringify({
				app_session_uuid: ms.appSessionUuid,
				run_id: runId
			})
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`agent/session.cancel failed: HTTP ${res.status}: ${text}`), AGENT_ERR_PROVIDER_ERROR);
		}
		return await res.json().catch(() => ({}));
	}
	async sessionHistory(params) {
		const acc = requireAccount(this.opts.account);
		const ms = this.getSession(params);
		const limit = Math.min(200, Math.max(1, Number(params.limit ?? 50)));
		const cursor = params.cursor ?? null;
		const url = new URL(`${hostOf(acc)}/api/v1/copilot/app/sessions/${encodeURIComponent(ms.appSessionUuid)}/history`);
		url.searchParams.set("limit", String(limit));
		if (cursor) url.searchParams.set("cursor", String(cursor));
		const res = await fetch(url, {
			method: "GET",
			headers: { authorization: `Bearer ${ms.appSessionToken}` }
		});
		if (res.status === 404) return {
			messages: [],
			cursor: null
		};
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`agent/session.history failed: HTTP ${res.status}: ${text}`), AGENT_ERR_PROVIDER_ERROR);
		}
		return await res.json();
	}
	async sessionDelete(params) {
		const acc = requireAccount(this.opts.account);
		const uuid = String(params.app_session_uuid ?? "");
		if (!uuid.startsWith("aps_")) throw withCode(new Error("'app_session_uuid' must start with 'aps_'"), AGENT_ERR_INVALID_REQUEST);
		const ms = this.sessions.get(uuid);
		this.sessions.delete(uuid);
		if (!ms) return {
			deleted: true,
			app_session_uuid: uuid,
			already_gone: true
		};
		const url = `${hostOf(acc)}/api/v1/copilot/app/sessions/${encodeURIComponent(uuid)}`;
		const res = await fetch(url, {
			method: "DELETE",
			headers: { authorization: `Bearer ${ms.appSessionToken}` }
		});
		if (!res.ok && res.status !== 404) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`agent/session.delete failed: HTTP ${res.status}: ${text}`), AGENT_ERR_PROVIDER_ERROR);
		}
		return {
			deleted: true,
			app_session_uuid: uuid
		};
	}
	/**
	* L1 `agent/complete` — single-turn completion convenience. Mints a
	* transient kind=complete session, forwards to `/copilot/app/complete`,
	* returns the assistant message.
	*/
	async complete(params) {
		const acc = requireAccount(this.opts.account);
		const messages = params.messages;
		if (!Array.isArray(messages) || messages.length === 0) throw withCode(new Error("'messages' must be a non-empty array"), AGENT_ERR_INVALID_REQUEST);
		const minted = await mintAppSession(acc.host, {
			pat: acc.pat,
			kind: "complete",
			app_slug: this.requireAppSlug()
		});
		const body = { messages };
		if (params.system) body.system = params.system;
		if (params.max_tokens != null) body.max_tokens = params.max_tokens;
		if (params.maxTokens != null) body.max_tokens = params.maxTokens;
		if (params.temperature != null) body.temperature = params.temperature;
		if (params.stop) body.stop = params.stop;
		if (params.model_hint) body.model_hint = params.model_hint;
		if (params.metadata) body.metadata = params.metadata;
		const res = await fetch(`${hostOf(acc)}/api/v1/copilot/app/complete`, {
			method: "POST",
			headers: {
				"content-type": "application/json",
				authorization: `Bearer ${minted.app_session_token}`
			},
			body: JSON.stringify(body)
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`agent/complete failed: HTTP ${res.status}: ${text}`), AGENT_ERR_PROVIDER_ERROR);
		}
		return await res.json();
	}
};
function collectContent(params) {
	const parts = [];
	if (typeof params.content === "string") parts.push(params.content);
	if (typeof params.system === "string") parts.push(params.system);
	const msgs = params.messages;
	if (Array.isArray(msgs)) for (const m of msgs) {
		const c = m.content;
		if (typeof c === "string") parts.push(c);
		else if (c && typeof c === "object" && "text" in c) parts.push(String(c.text ?? ""));
	}
	return parts.join("\n");
}
function parseSseFrame(frame) {
	const lines = frame.split(/\r?\n/);
	const dataLines = [];
	let eventName = null;
	for (const raw of lines) {
		if (!raw || raw.startsWith(":")) continue;
		if (raw.startsWith("data:")) dataLines.push(raw.slice(5).trimStart());
		else if (raw.startsWith("event:")) eventName = raw.slice(6).trim();
	}
	if (dataLines.length === 0) return null;
	const data = dataLines.join("\n");
	try {
		const obj = JSON.parse(data);
		if (eventName && obj && typeof obj === "object" && !("event" in obj)) return {
			...obj,
			event: eventName
		};
		return obj;
	} catch {
		return {
			event: eventName ?? "raw",
			text: data
		};
	}
}

//#endregion
export { AgentBridge };
