// @anna-ai/cli/dist/app-bundle-upload-CPf_nGou.js
import { finalizeBundle, finalizeWorkingBundle, getBundle, initBundle, initWorkingBundle, uploadBundleFile, uploadWorkingBundleFile } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { canonicalize } from "./executa-publish-Xmm4DMO9.js";
import { join, relative, sep } from "node:path";
import { readFileSync, readdirSync, statSync } from "node:fs";
import { createHash } from "node:crypto";

//#region src/publish/app-bundle-upload.ts
const CONTENT_TYPES = {
	html: "text/html",
	htm: "text/html",
	css: "text/css",
	js: "text/javascript",
	mjs: "text/javascript",
	json: "application/json",
	map: "application/json",
	webmanifest: "application/manifest+json",
	wasm: "application/wasm",
	png: "image/png",
	jpg: "image/jpeg",
	jpeg: "image/jpeg",
	svg: "image/svg+xml",
	webp: "image/webp",
	avif: "image/avif",
	ico: "image/x-icon",
	woff: "font/woff",
	woff2: "font/woff2",
	txt: "text/plain",
	md: "text/markdown"
};
const MAX_TOTAL_BYTES = 50 * 1024 * 1024;
const MAX_FILE_BYTES = 10 * 1024 * 1024;
const MAX_FILES = 2e3;
const UPLOAD_PARALLEL = 4;
function guessContentType(name) {
	const m = /\.([a-z0-9]+)$/i.exec(name);
	if (!m) return "application/octet-stream";
	return CONTENT_TYPES[m[1].toLowerCase()] ?? "application/octet-stream";
}
/** Recursively collect files under `bundleDir` with stable POSIX paths. */
function collectFiles(bundleDir) {
	const out = [];
	const walk = (base) => {
		for (const ent of readdirSync(base, { withFileTypes: true })) {
			const abs = join(base, ent.name);
			if (ent.isDirectory()) walk(abs);
			else if (ent.isFile()) {
				const rel = relative(bundleDir, abs).split(sep).join("/");
				const bytes = readFileSync(abs);
				out.push({
					relativePath: rel,
					bytes,
					entry: {
						sha256: createHash("sha256").update(bytes).digest("hex"),
						byte_size: bytes.byteLength,
						content_type: guessContentType(rel)
					}
				});
			}
		}
	};
	walk(bundleDir);
	out.sort((a, b) => a.relativePath.localeCompare(b.relativePath));
	return out;
}
/**
* Compute the same manifest fingerprint the server stores as
* `sha256_manifest` (sha256 of canonical JSON of the sorted file_map with
* each entry reduced to {sha256, byte_size, content_type}).
*/
function manifestFingerprint(files) {
	const obj = {};
	for (const f of [...files].sort((a, b) => a.relativePath.localeCompare(b.relativePath))) obj[f.relativePath] = {
		sha256: f.entry.sha256,
		byte_size: f.entry.byte_size,
		content_type: f.entry.content_type
	};
	return createHash("sha256").update(canonicalize(obj)).digest("hex");
}
/**
* Upload + finalize the UI bundle for one app version. Returns
* `{ status: "skipped" }` when the version's bundle is already finalized
* with identical content.
*/
async function uploadAppBundle(params) {
	const { client, appId, versionId, bundleDir, entryPath, bundleFormat = "static-spa", externalOrigins = [], log } = params;
	const files = collectFiles(bundleDir);
	if (files.length === 0) throw new CliError(`bundle directory is empty: ${bundleDir}`, 4);
	if (files.length > MAX_FILES) throw new CliError(`too many bundle files (${files.length} > ${MAX_FILES})`, 4);
	const totalBytes = files.reduce((s, f) => s + f.entry.byte_size, 0);
	if (totalBytes > MAX_TOTAL_BYTES) throw new CliError(`bundle is ${(totalBytes / 1024 / 1024).toFixed(2)} MB — exceeds 50 MB limit`, 4);
	const tooBig = files.find((f) => f.entry.byte_size > MAX_FILE_BYTES);
	if (tooBig) throw new CliError(`bundle file exceeds 10 MB: ${tooBig.relativePath} (${(tooBig.entry.byte_size / 1024 / 1024).toFixed(2)} MB)`, 4);
	if (!files.some((f) => f.relativePath === entryPath)) throw new CliError(`manifest entry "${entryPath}" not found in bundle dir ${bundleDir}`, 4);
	const fingerprint = manifestFingerprint(files);
	const existing = await getBundle(client, appId, versionId);
	if (existing && existing.status === "bundle_ready") {
		if (existing.sha256_manifest === fingerprint) {
			log?.(`  bundle unchanged (${files.length} files, sha ${fingerprint.slice(0, 12)}…) — skipped`);
			return {
				status: "skipped",
				reason: "unchanged",
				detail: existing
			};
		}
		throw new CliError("version already has a finalized bundle with different content; bump the app version (e.g. --bump patch) to upload a new bundle", 4);
	}
	log?.(`  uploading UI bundle: ${files.length} files, ${(totalBytes / 1024).toFixed(1)} KB`);
	const fileMap = {};
	for (const f of files) fileMap[f.relativePath] = f.entry;
	const initResp = await initBundle(client, appId, versionId, {
		file_map: fileMap,
		total_size_bytes: totalBytes,
		file_count: files.length,
		bundle_format: bundleFormat,
		entry_path: entryPath,
		external_origins: externalOrigins
	});
	const proxyByPath = new Map(initResp.files.map((p) => [p.relative_path, p.proxy_upload_url]));
	const queue = [...files];
	const uploadOne = async () => {
		for (;;) {
			const f = queue.shift();
			if (!f) return;
			const proxyUrl = proxyByPath.get(f.relativePath);
			if (!proxyUrl) throw new CliError(`server did not return an upload URL for ${f.relativePath}`, 6);
			const ab = f.bytes.buffer.slice(f.bytes.byteOffset, f.bytes.byteOffset + f.bytes.byteLength);
			const blob = new Blob([ab], { type: f.entry.content_type });
			await uploadBundleFile(client, appId, versionId, f.relativePath, blob);
		}
	};
	await Promise.all(Array.from({ length: Math.min(UPLOAD_PARALLEL, files.length) }, uploadOne));
	const detail = await finalizeBundle(client, appId, versionId, false);
	log?.(`  ✓ bundle finalized (${detail.file_count} files, status=${detail.status})`);
	return {
		status: "uploaded",
		detail,
		fileCount: files.length,
		totalBytes
	};
}
/** True when `path` is an existing directory. */
function isExistingDir(path) {
	try {
		return statSync(path).isDirectory();
	} catch {
		return false;
	}
}
/**
* Stage + finalize the working (mutable) UI bundle. When `knownFingerprint`
* matches the freshly-computed manifest fingerprint, skips the upload
* (the working bundle already reflects these bytes).
*/
async function uploadWorkingBundle(params) {
	const { client, appId, bundleDir, entryPath, bundleFormat = "static-spa", externalOrigins = [], knownFingerprint, log } = params;
	const files = collectFiles(bundleDir);
	if (files.length === 0) throw new CliError(`bundle directory is empty: ${bundleDir}`, 4);
	if (files.length > MAX_FILES) throw new CliError(`too many bundle files (${files.length} > ${MAX_FILES})`, 4);
	const totalBytes = files.reduce((s, f) => s + f.entry.byte_size, 0);
	if (totalBytes > MAX_TOTAL_BYTES) throw new CliError(`bundle is ${(totalBytes / 1024 / 1024).toFixed(2)} MB — exceeds 50 MB limit`, 4);
	const tooBig = files.find((f) => f.entry.byte_size > MAX_FILE_BYTES);
	if (tooBig) throw new CliError(`bundle file exceeds 10 MB: ${tooBig.relativePath} (${(tooBig.entry.byte_size / 1024 / 1024).toFixed(2)} MB)`, 4);
	if (!files.some((f) => f.relativePath === entryPath)) throw new CliError(`manifest entry "${entryPath}" not found in bundle dir ${bundleDir}`, 4);
	const fingerprint = manifestFingerprint(files);
	if (knownFingerprint && knownFingerprint === fingerprint) {
		log?.(`  working bundle unchanged (${files.length} files, sha ${fingerprint.slice(0, 12)}…) — skipped`);
		return {
			status: "skipped",
			fileCount: files.length,
			totalBytes,
			fingerprint,
			bundleStatus: "ready"
		};
	}
	log?.(`  staging working bundle: ${files.length} files, ${(totalBytes / 1024).toFixed(1)} KB`);
	const fileMap = {};
	for (const f of files) fileMap[f.relativePath] = f.entry;
	const initResp = await initWorkingBundle(client, appId, {
		file_map: fileMap,
		total_size_bytes: totalBytes,
		file_count: files.length,
		bundle_format: bundleFormat,
		entry_path: entryPath,
		external_origins: externalOrigins
	});
	const proxyByPath = new Map(initResp.files.map((p) => [p.relative_path, p.proxy_upload_url]));
	const queue = [...files];
	const uploadOne = async () => {
		for (;;) {
			const f = queue.shift();
			if (!f) return;
			const proxyUrl = proxyByPath.get(f.relativePath);
			if (!proxyUrl) throw new CliError(`server did not return an upload URL for ${f.relativePath}`, 6);
			const ab = f.bytes.buffer.slice(f.bytes.byteOffset, f.bytes.byteOffset + f.bytes.byteLength);
			const blob = new Blob([ab], { type: f.entry.content_type });
			await uploadWorkingBundleFile(client, appId, f.relativePath, blob);
		}
	};
	await Promise.all(Array.from({ length: Math.min(UPLOAD_PARALLEL, files.length) }, uploadOne));
	const detail = await finalizeWorkingBundle(client, appId, false);
	log?.(`  ✓ working bundle staged (${detail.file_count} files, status=${detail.bundle_status})`);
	return {
		status: "uploaded",
		fileCount: files.length,
		totalBytes,
		fingerprint,
		bundleStatus: detail.bundle_status
	};
}

//#endregion
export { isExistingDir, uploadAppBundle, uploadWorkingBundle };
