// @anna-ai/cli/dist/app-cache-BqnTkxUw.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { dirname, join, resolve } from "node:path";
import { existsSync, mkdirSync, readFileSync, renameSync, unlinkSync, writeFileSync } from "node:fs";

//#region src/identity/app-cache.ts
const CACHE_DIR = ".anna";
const CACHE_FILE = "app.json";
function appCachePath(cwd) {
	return resolve(cwd, CACHE_DIR, CACHE_FILE);
}
function readAppIdentity(cwd) {
	const p = appCachePath(cwd);
	if (!existsSync(p)) return null;
	try {
		const raw = JSON.parse(readFileSync(p, "utf-8"));
		if (typeof raw.host === "string" && typeof raw.slug === "string" && typeof raw.app_id === "number") return {
			$schema: "anna-app-identity/v1",
			host: canonicalHost(raw.host),
			app_id: raw.app_id,
			slug: raw.slug,
			handle: typeof raw.handle === "string" ? raw.handle : void 0,
			first_published_at: raw.first_published_at
		};
	} catch {}
	return null;
}
function writeAppIdentity(cwd, identity) {
	const p = appCachePath(cwd);
	mkdirSync(dirname(p), { recursive: true });
	const tmp = `${p}.tmp.${process.pid}`;
	writeFileSync(tmp, JSON.stringify({
		...identity,
		host: canonicalHost(identity.host)
	}, null, 2) + "\n", "utf-8");
	try {
		renameSync(tmp, p);
	} catch (e) {
		try {
			unlinkSync(tmp);
		} catch {}
		throw e;
	}
}
function appCacheMatches(identity, host, slug) {
	if (!identity) return false;
	return identity.host === canonicalHost(host) && identity.slug === slug;
}
const _internal = {
	CACHE_DIR,
	CACHE_FILE: join(CACHE_DIR, CACHE_FILE)
};

//#endregion
export { appCacheMatches, readAppIdentity, writeAppIdentity };
