// @anna-ai/cli/dist/apps-sync-meta-DdPnXuIU.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { findAppBySlug, getApp, patchApp } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { loadAppManifest } from "./manifest-CMc7RM57.js";
import { readAppIdentity } from "./app-cache-BqnTkxUw.js";
import { resolve } from "node:path";
import { bold, cyan, dim, green } from "kleur/colors";

//#region src/commands/apps-sync-meta.ts
async function runAppsSyncMeta(opts) {
	return withErrorHandling(async () => {
		const cwd = resolve(opts.cwd ?? process.cwd());
		const manifest = loadAppManifest(cwd, opts.manifest);
		const { client, host } = resolveClient({ account: opts.account });
		let appId = null;
		let serverSlug = manifest.slug;
		const cached = readAppIdentity(cwd);
		if (cached && cached.host === canonicalHost(host) && cached.slug === manifest.slug) {
			appId = cached.app_id;
			serverSlug = cached.slug;
		} else {
			const found = await findAppBySlug(client, manifest.slug);
			if (found) {
				appId = found.id;
				serverSlug = found.slug;
			}
		}
		if (appId === null) throw new CliError(`no app with slug "${manifest.slug}" on ${host} — run 'anna-app apps publish' first`, 5);
		const meta = {
			name: manifest.name,
			category: manifest.category,
			tagline: manifest.tagline,
			description: manifest.description
		};
		if (opts.dryRun) {
			if (opts.json) {
				console.log(JSON.stringify({
					host,
					app_id: appId,
					slug: serverSlug,
					meta,
					dry_run: true
				}, null, 2));
				return 0;
			}
			console.log(`[dry-run] would PATCH apps/${serverSlug} metadata: ` + JSON.stringify(meta));
			return 0;
		}
		await patchApp(client, appId, meta);
		const finalApp = await getApp(client, appId);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				app_id: appId,
				slug: serverSlug,
				meta,
				status: finalApp.status
			}, null, 2));
			return 0;
		}
		console.log(green(`✓ apps/${serverSlug}: metadata synced`));
		console.log(`  ${dim("name     :")} ${manifest.name}`);
		if (manifest.category) console.log(`  ${dim("category :")} ${manifest.category}`);
		if (manifest.tagline) console.log(`  ${dim("tagline  :")} ${manifest.tagline}`);
		console.log(bold(cyan(`status: ${finalApp.status}`)));
		return 0;
	});
}

//#endregion
export { runAppsSyncMeta };
