// @anna-ai/cli/dist/apps-publish-B-ZVNmgn.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { createApp, createVersion, findAppBySlug, getApp } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { parseExecutaIdOverrides, readExecutasLock, substituteBundledRefs, validateBundledHandles, writeBundleToolIdSidecar, writeExecutasLock } from "./bundled-executas-_H1A9M50.js";
import { bumpVersion, bundleHash, manifestHash, rewriteVersion, runExecutaPublish } from "./executa-publish-Xmm4DMO9.js";
import { loadAppManifest } from "./manifest-CMc7RM57.js";
import { appCacheMatches, readAppIdentity, writeAppIdentity } from "./app-cache-BqnTkxUw.js";
import { isExistingDir, uploadAppBundle } from "./app-bundle-upload-CPf_nGou.js";
import { resolve } from "node:path";
import { bold, cyan, dim, green, yellow } from "kleur/colors";

//#region src/commands/apps-publish.ts
async function runAppsPublish(opts) {
	return withErrorHandling(async () => {
		const cwd = resolve(opts.cwd ?? process.cwd());
		const manifest = loadAppManifest(cwd, opts.manifest);
		const { client, host, source } = resolveClient({ account: opts.account });
		if (!opts.json) {
			console.log(dim(`using PAT from ${source} (host=${host})`));
			console.log(yellow("⚠ `apps publish` mints an immutable version directly. The Option-C working-draft flow is now preferred:"));
			console.log(dim("    anna-app apps push            ") + dim("# upsert the mutable working draft (safe every commit)"));
			console.log(dim("    anna-app apps cut <version>   ") + dim("# snapshot + freeze executa deps"));
		}
		if (opts.bump) {
			const next = bumpVersion(manifest.version, opts.bump);
			if (!opts.json) console.log(yellow(`↑ bumping manifest.version: ${manifest.version} → ${next}`));
			if (!opts.noWrite && !opts.dryRun) rewriteVersion(manifest.metadataPath, manifest.metadataRaw, next);
			manifest.version = next;
			manifest.metadataRaw.version = next;
			if (manifest.metadataRaw !== manifest.raw && "version" in manifest.raw) manifest.raw.version = next;
		}
		const overrides = parseExecutaIdOverrides(opts.executaId);
		const bundledActive = !opts.noBundledExecutas && (manifest.bundledExecutas.length > 0 || overrides.length > 0);
		let publishManifestRaw = manifest.raw;
		if (bundledActive) {
			const { referenced } = validateBundledHandles(manifest.raw, manifest.bundledExecutas, overrides);
			const overrideMap = new Map(overrides.map((o) => [o.handle, o.toolId]));
			const lock = readExecutasLock(cwd, host);
			const resolved = {};
			const lockEntries = {};
			for (const decl of manifest.bundledExecutas) {
				if (overrideMap.has(decl.handle)) {
					const id = overrideMap.get(decl.handle);
					resolved[decl.handle] = id;
					lockEntries[decl.handle] = {
						tool_id: id,
						path: decl.path,
						source: "override"
					};
					if (!opts.json) console.log(dim(`  bundled:${decl.handle} → ${id} (override)`));
					continue;
				}
				if (opts.skipExecutaPublish) {
					const cachedId = lock?.executas[decl.handle]?.tool_id;
					if (!cachedId) throw new CliError(`--skip-executa-publish set but no cached tool_id for "${decl.handle}" in .anna/executas.lock.json (run without --skip-executa-publish first, or pass --executa-id ${decl.handle}=<tool-id>)`, 4);
					resolved[decl.handle] = cachedId;
					lockEntries[decl.handle] = {
						tool_id: cachedId,
						path: decl.path,
						source: "cache"
					};
					if (!opts.json) console.log(dim(`  bundled:${decl.handle} → ${cachedId} (cache)`));
					continue;
				}
				const execCwd = resolve(cwd, decl.path);
				if (opts.dryRun) {
					if (!opts.json) console.log(yellow(`  [dry-run] would publish bundled executa "${decl.handle}" from ${decl.path}`));
					continue;
				}
				if (!opts.json) console.log(dim(`  publishing bundled executa "${decl.handle}" (${decl.path})…`));
				let captured;
				const code = await runExecutaPublish({
					cwd: execCwd,
					account: opts.account,
					quiet: true,
					capture: (r) => {
						captured = r;
					}
				});
				if (code !== 0 || !captured) throw new CliError(`failed to publish bundled executa "${decl.handle}" (${decl.path})`, code === 0 ? 4 : code);
				resolved[decl.handle] = captured.tool_id;
				lockEntries[decl.handle] = {
					tool_id: captured.tool_id,
					path: decl.path,
					source: "published"
				};
				if (!opts.json) console.log(green(`  ✓ bundled:${decl.handle} → ${captured.tool_id}`) + dim(` (v${captured.version}${captured.idempotent_hit ? ", unchanged" : ""})`));
			}
			for (const o of overrides) if (!(o.handle in resolved)) {
				resolved[o.handle] = o.toolId;
				lockEntries[o.handle] = {
					tool_id: o.toolId,
					source: "override"
				};
				if (!opts.json) console.log(dim(`  bundled:${o.handle} → ${o.toolId} (override)`));
			}
			const missing = [...referenced].filter((h) => !(h in resolved));
			if (missing.length > 0 && !opts.dryRun) throw new CliError(`could not resolve bundled handle(s): ${missing.sort().join(", ")}`, 4);
			if (!opts.dryRun) {
				const merged = {
					$schema: "anna-executas-lock/v1",
					host: canonicalHost(host),
					executas: {
						...lock?.executas ?? {},
						...lockEntries
					}
				};
				writeExecutasLock(cwd, merged);
				publishManifestRaw = substituteBundledRefs(manifest.raw, resolved);
				const bundleEntry = manifest.raw.ui?.bundle?.entry;
				if (bundleEntry) {
					const bundleDir$1 = resolve(cwd, "bundle");
					if (Object.keys(resolved).length > 0) {
						const sidecar = writeBundleToolIdSidecar(bundleDir$1, resolved);
						if (!opts.json) console.log(dim(`  wrote ${sidecar.replace(cwd + "/", "")}`));
					}
				}
			}
		}
		let cached = readAppIdentity(cwd);
		if (cached && cached.host !== canonicalHost(host)) cached = null;
		let appId;
		let serverSlug;
		let firstPublish = false;
		let appIdFromCache = false;
		if (cached && appCacheMatches(cached, host, manifest.slug)) {
			appId = cached.app_id;
			serverSlug = cached.slug;
			appIdFromCache = true;
		} else {
			const found = await findAppBySlug(client, manifest.slug);
			if (found) {
				appId = found.id;
				serverSlug = found.slug;
			} else {
				if (opts.dryRun) {
					if (!opts.json) console.log(yellow(`[dry-run] would create AnnaApp slug=${manifest.slug}, then upload version ${manifest.version}`));
					return 0;
				}
				const created = await createApp(client, {
					slug: manifest.slug,
					name: manifest.name,
					category: manifest.category,
					tagline: manifest.tagline,
					description: manifest.description
				});
				appId = created.id;
				serverSlug = created.slug;
				firstPublish = true;
			}
			writeAppIdentity(cwd, {
				$schema: "anna-app-identity/v1",
				host: canonicalHost(host),
				app_id: appId,
				slug: serverSlug,
				first_published_at: cached?.first_published_at ?? new Date().toISOString()
			});
		}
		if (opts.dryRun) {
			if (!opts.json) {
				console.log(yellow(`[dry-run] resolved app_id=${appId} slug=${serverSlug}; would POST version ${manifest.version}`));
				const uiEntry = publishManifestRaw.ui?.bundle?.entry;
				const dryBundleDir = resolve(cwd, opts.bundleDir ?? "bundle");
				if (uiEntry && !opts.noBundle && isExistingDir(dryBundleDir)) console.log(yellow(`[dry-run] would upload UI bundle from ${dryBundleDir} (entry=${uiEntry})`));
			}
			return 0;
		}
		const hashInput = {
			version: manifest.version,
			manifest: publishManifestRaw
		};
		const hashBundleEntry = publishManifestRaw.ui?.bundle?.entry;
		const hashBundleDir = resolve(cwd, opts.bundleDir ?? "bundle");
		const hash = hashBundleEntry && !opts.noBundle && isExistingDir(hashBundleDir) ? bundleHash(hashInput, hashBundleDir) : manifestHash(hashInput);
		let verResp;
		try {
			verResp = await createVersion(client, appId, {
				version: manifest.version,
				changelog: manifest.changelog,
				manifest: publishManifestRaw
			}, hash);
		} catch (e) {
			if (appIdFromCache && e instanceof CliError && e.status === 404) {
				if (!opts.json) console.log(yellow(`\u26a0 cached app_id=${appId} not found on server (deleted?); re-resolving identity and retrying\u2026`));
				const found = await findAppBySlug(client, manifest.slug);
				if (found) {
					appId = found.id;
					serverSlug = found.slug;
				} else {
					const created = await createApp(client, {
						slug: manifest.slug,
						name: manifest.name,
						category: manifest.category,
						tagline: manifest.tagline,
						description: manifest.description
					});
					appId = created.id;
					serverSlug = created.slug;
					firstPublish = true;
				}
				writeAppIdentity(cwd, {
					$schema: "anna-app-identity/v1",
					host: canonicalHost(host),
					app_id: appId,
					slug: serverSlug,
					first_published_at: cached?.first_published_at ?? new Date().toISOString()
				});
				verResp = await createVersion(client, appId, {
					version: manifest.version,
					changelog: manifest.changelog,
					manifest: publishManifestRaw
				}, hash);
			} else throw e;
		}
		let bundleResult = null;
		const uiBundle = publishManifestRaw.ui?.bundle;
		const bundleDir = resolve(cwd, opts.bundleDir ?? "bundle");
		if (uiBundle?.entry && !opts.noBundle) if (!isExistingDir(bundleDir)) {
			if (!opts.json) console.log(yellow(`⚠ manifest declares ui.bundle.entry="${uiBundle.entry}" but bundle dir not found (${bundleDir}); skipping UI bundle upload`));
		} else bundleResult = await uploadAppBundle({
			client,
			appId,
			versionId: verResp.version.id,
			bundleDir,
			entryPath: uiBundle.entry,
			bundleFormat: uiBundle.format,
			externalOrigins: Array.isArray(uiBundle.external_origins) ? uiBundle.external_origins : [],
			log: opts.json ? void 0 : (m) => console.log(dim(m))
		});
		const finalApp = await getApp(client, appId);
		if (opts.json) {
			console.log(JSON.stringify({
				app_id: appId,
				slug: serverSlug,
				host,
				version: verResp.version.version,
				version_id: verResp.version.id,
				content_hash: hash,
				idempotent_hit: verResp.idempotentHit,
				first_publish: firstPublish,
				status: finalApp.status,
				bundle: bundleResult ? {
					status: bundleResult.status,
					bundle_status: bundleResult.detail.status,
					file_count: bundleResult.detail.file_count,
					total_size_bytes: bundleResult.detail.total_size_bytes,
					sha256_manifest: bundleResult.detail.sha256_manifest
				} : null
			}, null, 2));
			return 0;
		}
		if (verResp.idempotentHit) console.log(green(`✓ apps/${serverSlug}: version ${manifest.version} already live (content hash unchanged)`));
		else console.log(green(`✓ apps/${serverSlug}: version ${manifest.version} published (id=${verResp.version.id})`));
		if (firstPublish) console.log(dim(`  first publish — wrote .anna/app.json`));
		if (bundleResult) if (bundleResult.status === "skipped") console.log(dim(`  UI bundle unchanged — skipped`));
		else console.log(green(`  ✓ UI bundle: ${bundleResult.fileCount} files, ${(bundleResult.totalBytes / 1024).toFixed(1)} KB → ${bundleResult.detail.status}`));
		console.log(bold(cyan(`status: ${finalApp.status}`)));
		if (!verResp.idempotentHit) console.log(dim(`→ run `) + cyan(`anna-app apps release ${manifest.version}`) + dim(` to freeze & publish this version`));
		return 0;
	});
}

//#endregion
export { runAppsPublish };
