// @anna-ai/cli/dist/apps-73pdBAFE.js
//#region src/api/apps.ts
async function findAppBySlug(client, slug) {
	const r = await client.request({
		path: "/api/v1/developer/apps",
		query: { slug },
		allowStatuses: [404]
	});
	if (r.status === 404) return null;
	if (Array.isArray(r.data) && r.data.length > 0) return r.data[0] ?? null;
	return null;
}
async function listMyApps(client) {
	const r = await client.request({ path: "/api/v1/developer/apps" });
	return Array.isArray(r.data) ? r.data : [];
}
async function getApp(client, appId) {
	const r = await client.request({ path: `/api/v1/developer/apps/${appId}` });
	return r.data;
}
async function createApp(client, body) {
	const r = await client.request({
		method: "POST",
		path: "/api/v1/developer/apps",
		body
	});
	return r.data;
}
async function patchApp(client, appId, body) {
	const r = await client.request({
		method: "PATCH",
		path: `/api/v1/developer/apps/${appId}`,
		body
	});
	return r.data;
}
async function listVersions(client, appId) {
	const r = await client.request({ path: `/api/v1/developer/apps/${appId}/versions` });
	return Array.isArray(r.data) ? r.data : [];
}
async function createVersion(client, appId, body, contentHash) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/versions`,
		query: contentHash ? { content_hash: contentHash } : void 0,
		body
	});
	return {
		version: r.data,
		idempotentHit: r.idempotentHit
	};
}
async function submitForReview(client, appId) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/submit-review`
	});
	return r.data;
}
/**
* Freeze & publish an existing immutable version (`publish_version`):
* marks `is_latest`, sets `published_at`, freezes executa dependency
* snapshots and flips the app to PUBLISHED. Idempotent on the server
* (re-running on the already-latest version is harmless).
*/
async function publishVersion(client, appId, versionId) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/versions/${versionId}/publish`
	});
	return r.data;
}
async function unpublishApp(client, appId) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/unpublish`
	});
	return r.data;
}
async function archiveApp(client, appId) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/archive`
	});
	return r.data;
}
async function unarchiveApp(client, appId) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/unarchive`
	});
	return r.data;
}
async function deleteApp(client, appId) {
	await client.request({
		method: "DELETE",
		path: `/api/v1/developer/apps/${appId}`,
		allowStatuses: [204]
	});
}
async function getAppGrants(client, appId) {
	const r = await client.request({
		path: `/api/v1/apps/${appId}/grants`,
		allowStatuses: [404]
	});
	return r.status === 404 ? null : r.data;
}
/**
* Set (or rename) the caller's developer handle — the per-developer
* namespace for `@handle/slug` public app refs.
*
* `PUT /api/v1/developer/profile/handle`. Server validates the handle
* grammar/reserved words and uniqueness; renaming away from a handle that
* owns a published app is rejected (409 SLUG_FROZEN / manual flow).
*/
async function setDeveloperHandle(client, handle) {
	const r = await client.request({
		method: "PUT",
		path: "/api/v1/developer/profile/handle",
		body: { handle }
	});
	return r.data;
}
/** GET the bundle for a version, or `null` when none is initialized (404). */
async function getBundle(client, appId, versionId) {
	const r = await client.request({
		path: `/api/v1/developer/apps/${appId}/versions/${versionId}/bundle`,
		allowStatuses: [404]
	});
	return r.status === 404 ? null : r.data;
}
async function initBundle(client, appId, versionId, body) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/versions/${versionId}/bundle/init`,
		body
	});
	return r.data;
}
/** Upload one file via the same-origin proxy endpoint (multipart). */
async function uploadBundleFile(client, appId, versionId, relativePath, blob) {
	await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/versions/${versionId}/bundle/file`,
		multipart: { fields: {
			relative_path: relativePath,
			file: blob
		} },
		allowStatuses: [201]
	});
}
async function finalizeBundle(client, appId, versionId, skipRemoteCheck = false) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/versions/${versionId}/bundle/finalize`,
		body: { skip_remote_check: skipRemoteCheck }
	});
	return r.data;
}
/**
* Upsert the single mutable working draft for an app. `ifMatch` sends an
* `If-Match: <revision>` optimistic-lock header (stale → 409). Returns
* `unchanged: true` (server-side dedupe on `content_hash`) without bumping
* the revision.
*/
async function upsertWorkingDraft(client, appId, body, ifMatch) {
	const r = await client.request({
		method: "PUT",
		path: `/api/v1/developer/apps/${appId}/working`,
		body,
		headers: ifMatch !== void 0 ? { "If-Match": String(ifMatch) } : void 0
	});
	const unchanged = r.data.unchanged || r.idempotentHit;
	return {
		...r.data,
		unchanged
	};
}
/** GET the working draft, or `null` when none exists yet (404). */
async function getWorkingDraft(client, appId) {
	const r = await client.request({
		path: `/api/v1/developer/apps/${appId}/working`,
		allowStatuses: [404]
	});
	return r.status === 404 ? null : r.data;
}
/** Discard the working draft (idempotent). */
async function deleteWorkingDraft(client, appId) {
	await client.request({
		method: "DELETE",
		path: `/api/v1/developer/apps/${appId}/working`,
		allowStatuses: [204, 404]
	});
}
/**
* Snapshot the working draft into an immutable `AnnaAppVersion` and freeze
* executa deps. `allow409` lets the caller inspect a duplicate-version 409
* (`{ conflict: true }`) instead of throwing.
*/
async function cutWorkingDraft(client, appId, body) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/working/cut`,
		body
	});
	return r.data;
}
async function initWorkingBundle(client, appId, body) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/working/bundle/init`,
		body
	});
	return r.data;
}
/** Upload one working-bundle file via the same-origin proxy (multipart). */
async function uploadWorkingBundleFile(client, appId, relativePath, blob) {
	await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/working/bundle/file`,
		multipart: { fields: {
			relative_path: relativePath,
			file: blob
		} },
		allowStatuses: [201]
	});
}
async function finalizeWorkingBundle(client, appId, skipRemoteCheck = false) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/developer/apps/${appId}/working/bundle/finalize`,
		body: { skip_remote_check: skipRemoteCheck }
	});
	return r.data;
}

//#endregion
export { archiveApp, createApp, createVersion, cutWorkingDraft, deleteApp, deleteWorkingDraft, finalizeBundle, finalizeWorkingBundle, findAppBySlug, getApp, getAppGrants, getBundle, getWorkingDraft, initBundle, initWorkingBundle, listMyApps, listVersions, patchApp, publishVersion, setDeveloperHandle, submitForReview, unarchiveApp, unpublishApp, uploadBundleFile, uploadWorkingBundleFile, upsertWorkingDraft };
