// @anna-ai/cli/dist/apps-push-Bh8zwT--.js
import "./credentials-BTv2IfUZ.js";
import "./nexus-root-BlPwOusj.js";
import { getApp, getWorkingDraft, upsertWorkingDraft } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import "./bundled-executas-_H1A9M50.js";
import "./executas-BhFed1B-.js";
import "./executa-cache-Do1D7xC0.js";
import { bumpVersion, bundleHash, manifestHash, rewriteVersion } from "./executa-publish-Xmm4DMO9.js";
import { loadAppManifest } from "./manifest-CMc7RM57.js";
import "./dev-aWkfWGLD.js";
import "./executa-install-A-BgavhA.js";
import "./app-cache-BqnTkxUw.js";
import { installLocalBundledShims, resolveAppIdentity, resolveBundledExecutas } from "./working-orchestration-CX0ONKIq.js";
import { isExistingDir, uploadWorkingBundle } from "./app-bundle-upload-CPf_nGou.js";
import { resolve } from "node:path";
import { bold, cyan, dim, green, yellow } from "kleur/colors";

//#region src/commands/apps-push.ts
async function runAppsPush(opts) {
	return withErrorHandling(async () => {
		const cwd = resolve(opts.cwd ?? process.cwd());
		const manifest = loadAppManifest(cwd, opts.manifest);
		const { client, host, source } = resolveClient({ account: opts.account });
		if (!opts.json) console.log(dim(`using PAT from ${source} (host=${host})`));
		if (opts.bump) {
			const next = bumpVersion(manifest.version, opts.bump);
			if (!opts.json) console.log(yellow(`↑ bumping manifest.version: ${manifest.version} → ${next}`));
			if (!opts.noWrite && !opts.dryRun) rewriteVersion(manifest.metadataPath, manifest.metadataRaw, next);
			manifest.version = next;
			manifest.metadataRaw.version = next;
		}
		const { publishManifestRaw, resolved } = await resolveBundledExecutas({
			manifest,
			cwd,
			host,
			executaId: opts.executaId,
			skipExecutaPublish: opts.skipExecutaPublish,
			noBundledExecutas: opts.noBundledExecutas,
			dryRun: opts.dryRun,
			freeze: false,
			lockSource: "working",
			account: opts.account,
			json: opts.json
		});
		let installedShims = [];
		if (!opts.dryRun && !opts.noInstallLocal && Object.keys(resolved).length > 0) installedShims = await installLocalBundledShims({
			manifest,
			cwd,
			resolved,
			json: opts.json
		});
		const identity = await resolveAppIdentity({
			client,
			cwd,
			host,
			manifest,
			dryRun: opts.dryRun,
			json: opts.json
		});
		if (!identity) return 0;
		const { appId, serverSlug, firstPublish } = identity;
		const hashInput = {
			version: manifest.version,
			manifest: publishManifestRaw
		};
		const bundleEntry = publishManifestRaw.ui?.bundle?.entry;
		const bundleDir = resolve(cwd, opts.bundleDir ?? "bundle");
		const hasBundle = !!bundleEntry && !opts.noBundle && isExistingDir(bundleDir);
		const hash = hasBundle ? bundleHash(hashInput, bundleDir) : manifestHash(hashInput);
		if (opts.dryRun) {
			if (!opts.json) console.log(yellow(`[dry-run] app_id=${appId} slug=${serverSlug}; would PUT /working (content_hash=${hash.slice(0, 12)}…)` + (hasBundle ? ` + stage working bundle from ${bundleDir}` : "")));
			return 0;
		}
		let resp;
		try {
			resp = await upsertWorkingDraft(client, appId, {
				manifest: publishManifestRaw,
				changelog: manifest.changelog,
				content_hash: hash
			}, opts.ifMatch);
		} catch (e) {
			if (e instanceof CliError && e.status === 409) throw new CliError(`working draft was modified concurrently (If-Match=${opts.ifMatch} is stale); re-run \`apps push\` without --if-match for last-writer-wins, or fetch the current revision with \`apps status\``, 4);
			throw e;
		}
		let bundleResult = null;
		if (hasBundle) {
			const uiBundle = publishManifestRaw.ui.bundle;
			const skipStage = resp.unchanged && resp.bundle_status === "ready";
			if (!skipStage) bundleResult = await uploadWorkingBundle({
				client,
				appId,
				bundleDir,
				entryPath: uiBundle.entry,
				bundleFormat: uiBundle.format,
				externalOrigins: Array.isArray(uiBundle.external_origins) ? uiBundle.external_origins : [],
				log: opts.json ? void 0 : (m) => console.log(dim(m))
			});
		}
		const finalApp = await getApp(client, appId);
		const draft = await getWorkingDraft(client, appId);
		if (opts.json) {
			console.log(JSON.stringify({
				app_id: appId,
				slug: serverSlug,
				host,
				revision: resp.revision,
				content_hash: resp.content_hash,
				unchanged: resp.unchanged,
				bundle_status: resp.bundle_status,
				first_publish: firstPublish,
				status: finalApp.status,
				dirty_vs_latest_cut: draft?.dirty_vs_latest_cut ?? null,
				installed_shims: installedShims,
				bundle: bundleResult ? {
					status: bundleResult.status,
					file_count: bundleResult.fileCount,
					total_size_bytes: bundleResult.totalBytes,
					bundle_status: bundleResult.bundleStatus
				} : null
			}, null, 2));
			return 0;
		}
		if (resp.unchanged) console.log(green(`✓ apps/${serverSlug}: working draft unchanged (rev ${resp.revision})`));
		else console.log(green(`✓ apps/${serverSlug}: working draft updated (rev ${resp.revision})`));
		if (firstPublish) console.log(dim(`  first push — wrote .anna/app.json`));
		if (bundleResult) if (bundleResult.status === "skipped") console.log(dim(`  working bundle unchanged — skipped`));
		else console.log(green(`  ✓ working bundle: ${bundleResult.fileCount} files, ${(bundleResult.totalBytes / 1024).toFixed(1)} KB → ${bundleResult.bundleStatus}`));
		console.log(bold(cyan(`status: ${finalApp.status}`)));
		if (draft?.dirty_vs_latest_cut) console.log(dim(`  (uncommitted changes vs latest cut version)`));
		if (installedShims.length > 0) {
			for (const s of installedShims) console.log(green(`  ✓ installed local dev shim: `) + cyan(s.tool_id));
			console.log(dim(`→ click `) + cyan(`Rediscover Local`) + dim(` in the Agent's Plugin Details to load `) + (installedShims.length === 1 ? dim("it") : dim("them")));
		}
		console.log(dim(`→ install & test the draft (reserved `) + cyan(`0.0.0-draft`) + dim(`) from the Developer Console`));
		console.log(dim(`→ run `) + cyan(`anna-app apps cut ${manifest.version}`) + dim(` to snapshot an immutable version`));
		return 0;
	});
}

//#endregion
export { runAppsPush };
