// @anna-ai/cli/dist/apps-status-5DV9cY4A.js
import "./credentials-BTv2IfUZ.js";
import { findAppBySlug, getApp, listVersions } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { bold, cyan, dim, green, red, yellow } from "kleur/colors";

//#region src/commands/apps-status.ts
async function runAppsStatus(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const app = await findAppBySlug(client, opts.slug);
		if (!app) throw new CliError(`no app with slug "${opts.slug}" on ${host}`, 5);
		const full = await getApp(client, app.id);
		const versions = await listVersions(client, app.id);
		const latest = versions.find((v) => v.is_latest) ?? versions[0] ?? null;
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				app_id: full.id,
				slug: full.slug,
				name: full.name,
				status: full.status,
				visibility: full.visibility ?? null,
				is_published: full.is_published ?? false,
				review_message: full.review_message ?? null,
				latest_version: latest ? {
					id: latest.id,
					version: latest.version,
					published_at: latest.published_at,
					is_latest: latest.is_latest ?? null
				} : null,
				version_count: versions.length
			}, null, 2));
			return 0;
		}
		console.log(bold(cyan(`apps/${full.slug}`)) + " " + dim(`(${host})`));
		console.log(`  ${dim("name      :")} ${full.name}`);
		console.log(`  ${dim("app_id    :")} ${full.id}`);
		console.log(`  ${dim("status    :")} ${colorStatus(full.status)}` + (full.is_published ? " " + green("[published]") : ""));
		if (latest) console.log(`  ${dim("latest    :")} v${latest.version} (id=${latest.id}, published=${latest.published_at ?? "—"})`);
		else console.log(`  ${dim("latest    :")} ${dim("(no versions yet)")}`);
		if (full.review_message) console.log(`  ${dim("reviewer  :")} ${full.review_message}`);
		console.log(`  ${dim("versions  :")} ${versions.length} total`);
		return 0;
	});
}
function colorStatus(s) {
	switch (s) {
		case "PUBLISHED": return green(s);
		case "PENDING_REVIEW": return yellow(s);
		case "REJECTED": return red(s);
		case "ARCHIVED": return dim(s);
		default: return s;
	}
}

//#endregion
export { runAppsStatus };
