// @anna-ai/cli/dist/apps-destructive-PY0wKjqu.js
import "./credentials-BTv2IfUZ.js";
import { archiveApp, deleteApp, findAppBySlug, unarchiveApp, unpublishApp } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { ensureDestructiveAllowed } from "./confirm-BuT56E_B.js";
import { dim, green, yellow } from "kleur/colors";

//#region src/commands/apps-destructive.ts
async function resolveAppId(account, slug) {
	const { client, host } = resolveClient({ account });
	const app = await findAppBySlug(client, slug);
	if (!app) throw new CliError(`no app with slug "${slug}" on ${host}`, 5);
	return {
		client,
		host,
		appId: app.id,
		serverSlug: app.slug
	};
}
async function runAppsUnpublish(opts) {
	return withErrorHandling(async () => {
		await ensureDestructiveAllowed({
			verb: "apps unpublish",
			...opts
		});
		const { client, host, appId, serverSlug } = await resolveAppId(opts.account, opts.slug);
		const updated = await unpublishApp(client, appId);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				app_id: appId,
				slug: serverSlug,
				status: updated.status
			}, null, 2));
			return 0;
		}
		console.log(green(`✓ apps/${serverSlug}: unpublished (status=${updated.status})`));
		return 0;
	});
}
async function runAppsArchive(opts) {
	return withErrorHandling(async () => {
		await ensureDestructiveAllowed({
			verb: "apps archive",
			...opts
		});
		const { client, host, appId, serverSlug } = await resolveAppId(opts.account, opts.slug);
		const updated = await archiveApp(client, appId);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				app_id: appId,
				slug: serverSlug,
				status: updated.status
			}, null, 2));
			return 0;
		}
		console.log(yellow(`✓ apps/${serverSlug}: archived (status=${updated.status})`));
		return 0;
	});
}
async function runAppsUnarchive(opts) {
	return withErrorHandling(async () => {
		if (!opts.yes) throw new CliError(`\`apps unarchive\` refused: pass --yes to acknowledge`, 7);
		const { client, host, appId, serverSlug } = await resolveAppId(opts.account, opts.slug);
		const updated = await unarchiveApp(client, appId);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				app_id: appId,
				slug: serverSlug,
				status: updated.status
			}, null, 2));
			return 0;
		}
		console.log(green(`✓ apps/${serverSlug}: unarchived (status=${updated.status})`));
		return 0;
	});
}
async function runAppsDelete(opts) {
	return withErrorHandling(async () => {
		await ensureDestructiveAllowed({
			verb: "apps delete",
			...opts
		});
		const { client, host, appId, serverSlug } = await resolveAppId(opts.account, opts.slug);
		await deleteApp(client, appId);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				app_id: appId,
				slug: serverSlug,
				deleted: true
			}, null, 2));
			return 0;
		}
		console.log(yellow(`✓ apps/${serverSlug}: deleted`));
		console.log(dim(`  (local .anna/app.json may now be stale; remove it manually if you no longer need it)`));
		return 0;
	});
}

//#endregion
export { runAppsArchive, runAppsDelete, runAppsUnarchive, runAppsUnpublish };
