// @anna-ai/cli/dist/apps-versions-vGihRRje.js
import "./credentials-BTv2IfUZ.js";
import { findAppBySlug, listVersions } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { bold, cyan, dim, green } from "kleur/colors";

//#region src/commands/apps-versions.ts
async function runAppsVersions(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const app = await findAppBySlug(client, opts.slug);
		if (!app) throw new CliError(`no app with slug "${opts.slug}" on ${host}`, 5);
		const versions = await listVersions(client, app.id);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				app_id: app.id,
				slug: app.slug,
				versions: versions.map((v) => ({
					id: v.id,
					version: v.version,
					is_latest: v.is_latest ?? false,
					published_at: v.published_at ?? null,
					created_at: v.created_at ?? null
				}))
			}, null, 2));
			return 0;
		}
		console.log(bold(cyan(`apps/${app.slug}`)) + " " + dim(`versions (${versions.length})`));
		if (versions.length === 0) {
			console.log(dim("  (none)"));
			return 0;
		}
		for (const v of versions) {
			const tag = v.is_latest ? green("[latest]") : dim("        ");
			console.log(`  ${tag} v${v.version} ${dim(`(id=${v.id}, published=${v.published_at ?? "—"})`)}`);
		}
		return 0;
	});
}

//#endregion
export { runAppsVersions };
