// @anna-ai/cli/dist/apps-release-DNlN_8gz.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { findAppBySlug, getApp, getBundle, listVersions, publishVersion } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { readAppIdentity } from "./app-cache-BqnTkxUw.js";
import { resolve } from "node:path";
import { bold, cyan, dim, green, yellow } from "kleur/colors";

//#region src/commands/apps-release.ts
/** Resolve the target app via `--slug` or the local `.anna/app.json` cache. */
async function resolveApp(client, host, opts) {
	if (opts.slug) {
		const app = await findAppBySlug(client, opts.slug);
		if (!app) throw new CliError(`no app with slug "${opts.slug}" on ${host}`, 5);
		return getApp(client, app.id);
	}
	const cwd = resolve(opts.cwd ?? process.cwd());
	const cached = readAppIdentity(cwd);
	if (cached && cached.host === canonicalHost(host)) try {
		return await getApp(client, cached.app_id);
	} catch (e) {
		if (e instanceof CliError && e.status === 404) {
			const app = await findAppBySlug(client, cached.slug);
			if (app) return getApp(client, app.id);
		} else throw e;
	}
	throw new CliError("cannot resolve app — pass --slug <slug> or run from a project dir with a .anna/app.json cache (run 'anna-app apps publish' first)", 5);
}
/** Does this version declare a UI bundle in its frozen manifest? */
function declaresUiBundle(version) {
	const entry = version.manifest?.ui?.bundle?.entry;
	return typeof entry === "string" && entry.length > 0;
}
function statusGuard(app) {
	if (app.status === "APPROVED" || app.status === "PUBLISHED") return;
	let hint;
	switch (app.status) {
		case "DRAFT":
		case "REJECTED":
			hint = "submit for review first: `anna-app apps submit-review`";
			break;
		case "PENDING_REVIEW":
			hint = "awaiting admin approval — cannot release until APPROVED";
			break;
		case "ARCHIVED":
			hint = "app is archived; restore it first: `anna-app apps unarchive`";
			break;
		default: hint = "app must be APPROVED or PUBLISHED to release";
	}
	throw new CliError(`app status is ${app.status}; release not permitted — ${hint}`, 6);
}
async function runAppsRelease(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		let app = await resolveApp(client, host, opts);
		let versions = await listVersions(client, app.id);
		let target = versions.find((v) => v.version === opts.version);
		if (!target) {
			if (opts.allowCreate) {
				if (!opts.json) console.log(yellow(`version ${opts.version} not found remotely; running 'apps publish' first (--allow-create)…`));
				const { runAppsPublish } = await import("./apps-publish-3wbhXfAg.js");
				const code = await runAppsPublish({
					cwd: opts.cwd,
					account: opts.account,
					json: opts.json
				});
				if (code !== 0) return code;
				app = await getApp(client, app.id);
				versions = await listVersions(client, app.id);
				target = versions.find((v) => v.version === opts.version);
			}
			if (!target) throw new CliError(`no version ${opts.version} for app "${app.slug}" on ${host} — run 'anna-app apps publish' first`, 5);
		}
		if (declaresUiBundle(target)) {
			const bundle = await getBundle(client, app.id, target.id);
			if (!bundle || bundle.status !== "bundle_ready") throw new CliError(`version ${opts.version} declares a UI bundle but it is not finalized (status=${bundle?.status ?? "missing"}) — re-run 'anna-app apps publish'`, 4);
		}
		statusGuard(app);
		if (target.is_latest && app.status === "PUBLISHED") {
			if (opts.json) {
				console.log(JSON.stringify({
					host,
					app_id: app.id,
					slug: app.slug,
					version: target.version,
					version_id: target.id,
					is_latest: true,
					published_at: target.published_at ?? null,
					status: app.status,
					already_live: true
				}, null, 2));
				return 0;
			}
			console.log(green(`✓ apps/${app.slug}: version ${target.version} already live`) + dim(` (no-op)`));
			return 0;
		}
		if (opts.dryRun) {
			if (opts.json) {
				console.log(JSON.stringify({
					host,
					app_id: app.id,
					slug: app.slug,
					version: target.version,
					version_id: target.id,
					current_status: app.status,
					dry_run: true
				}, null, 2));
				return 0;
			}
			console.log(yellow(`[dry-run] would release apps/${app.slug} version ${target.version} (id=${target.id}, current status=${app.status})`));
			return 0;
		}
		const released = await publishVersion(client, app.id, target.id);
		const finalApp = await getApp(client, app.id);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				app_id: app.id,
				slug: app.slug,
				version: released.version,
				version_id: released.id,
				is_latest: released.is_latest ?? true,
				published_at: released.published_at ?? null,
				status: finalApp.status
			}, null, 2));
			return 0;
		}
		console.log(green(`✓ apps/${app.slug}: released version ${released.version}`) + dim(` (id=${released.id})`));
		console.log(bold(cyan(`status: ${finalApp.status}`)) + " — now live in the store");
		return 0;
	});
}

//#endregion
export { runAppsRelease };
