// @anna-ai/cli/dist/apps-cut-C5SH9qZN.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import "./nexus-root-BlPwOusj.js";
import { cutWorkingDraft } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { readExecutasLock, writeExecutasLock } from "./bundled-executas-_H1A9M50.js";
import "./executas-BhFed1B-.js";
import "./executa-cache-Do1D7xC0.js";
import "./executa-publish-Xmm4DMO9.js";
import "./manifest-CMc7RM57.js";
import "./dev-aWkfWGLD.js";
import "./executa-install-A-BgavhA.js";
import "./app-cache-BqnTkxUw.js";
import { resolveAppBySlugOrCache } from "./working-orchestration-CX0ONKIq.js";
import { resolve } from "node:path";
import { bold, cyan, dim, green, yellow } from "kleur/colors";

//#region src/commands/apps-cut.ts
async function runAppsCut(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const app = await resolveAppBySlugOrCache(client, host, {
			slug: opts.slug,
			cwd: opts.cwd
		});
		if (opts.dryRun) {
			if (!opts.json) console.log(yellow(`[dry-run] would cut version ${opts.version} for apps/${app.slug} (app_id=${app.id})`));
			return 0;
		}
		let result;
		try {
			result = await cutWorkingDraft(client, app.id, {
				version: opts.version,
				changelog: opts.changelog
			});
		} catch (e) {
			if (e instanceof CliError && e.status === 409) throw new CliError(`version ${opts.version} already cut for apps/${app.slug}; choose a higher SemVer (e.g. bump patch/minor) and retry`, 9);
			if (e instanceof CliError && (e.status === 400 || e.status === 422)) throw new CliError(`cannot cut: ${e.message} — fix the working draft and \`apps push\` again`, 4);
			throw e;
		}
		if (result.frozen_executas.length > 0) {
			const cwd = resolve(opts.cwd ?? process.cwd());
			const lock = readExecutasLock(cwd, host);
			if (lock) {
				const byToolId = new Map(Object.entries(lock.executas).map(([h, e]) => [e.tool_id, h]));
				for (const fr of result.frozen_executas) {
					const handle = byToolId.get(fr.tool_id);
					if (handle && lock.executas[handle]) lock.executas[handle] = {
						...lock.executas[handle],
						tool_id: fr.tool_id,
						source: "cut"
					};
				}
				writeExecutasLock(cwd, {
					$schema: "anna-executas-lock/v1",
					host: canonicalHost(host),
					executas: lock.executas
				});
			}
		}
		if (opts.json) {
			console.log(JSON.stringify({
				app_id: app.id,
				slug: app.slug,
				host,
				version_id: result.version_id,
				version: result.version,
				frozen_executas: result.frozen_executas,
				bundle: result.bundle
			}, null, 2));
			return 0;
		}
		console.log(green(`✓ apps/${app.slug}: cut immutable version ${result.version} `) + dim(`(version_id=${result.version_id})`));
		for (const fr of result.frozen_executas) console.log(dim(`  froze ${fr.tool_id} → executa_version=${fr.executa_version_id}`) + (fr.version ? dim(` (v${fr.version})`) : ""));
		if (result.bundle && typeof result.bundle === "object") {
			const prefix = result.bundle.object_prefix;
			if (prefix) console.log(dim(`  bundle frozen at ${prefix}`));
		}
		console.log(dim(`→ run `) + cyan(`anna-app apps release ${result.version}`) + dim(` to go live`));
		console.log(bold(cyan(`status: ${app.status}`)));
		return 0;
	});
}

//#endregion
export { runAppsCut };
