// @anna-ai/cli/dist/apps-BpJ-TvHP.js
import "./credentials-BTv2IfUZ.js";
import { listMyApps } from "./apps-73pdBAFE.js";
import "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { bold, cyan, dim, green, magenta, red, yellow } from "kleur/colors";

//#region src/commands/apps.ts
/** Colorize the lifecycle status so the next action is obvious at a glance. */
function statusTag(status) {
	switch (status.toUpperCase()) {
		case "PUBLISHED": return green("[PUBLISHED]");
		case "APPROVED": return cyan("[APPROVED]");
		case "PENDING_REVIEW": return yellow("[PENDING_REVIEW]");
		case "REJECTED": return red("[REJECTED]");
		case "ARCHIVED": return dim("[ARCHIVED]");
		case "DRAFT": return magenta("[DRAFT]");
		default: return dim(`[${status}]`);
	}
}
async function runAppsList(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const apps = await listMyApps(client);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				apps: apps.map((a) => ({
					id: a.id,
					slug: a.slug,
					name: a.name,
					status: a.status,
					is_dev: a.is_dev ?? false,
					updated_at: a.updated_at ?? null
				}))
			}, null, 2));
			return 0;
		}
		console.log(bold(cyan("apps you authored on")) + " " + cyan(host));
		if (apps.length === 0) {
			console.log(dim("  (none — run `anna-app apps publish` in a project to create one)"));
			return 0;
		}
		for (const a of apps) {
			const dev = a.is_dev ? " " + yellow("[dev]") : "";
			console.log(`  ${statusTag(a.status)}${dev} ${bold(a.slug)} ${dim(`(id=${a.id})`)}`);
			if (a.name && a.name !== a.slug) console.log(`        ${dim(a.name)}`);
		}
		return 0;
	});
}

//#endregion
export { runAppsList };
