// @anna-ai/cli/dist/bridge-DXH59CmP.js
import { delimiter, resolve } from "node:path";
import { spawn } from "node:child_process";
import { createInterface } from "node:readline";

//#region src/harness/bridge.ts
/**
* Pinned `anna-app-runtime-local` version for uvx mode. Bump in lock-step
* with `packages/VERSIONS.md` (matrix-nexus). The bridge passes this to
* `uvx <pkg>@<version>` so end users always run the dispatcher version
* the CLI was tested against.
*/
const PINNED_RUNTIME_VERSION = "0.2.0a9";
/**
* Throwable from a {@link RequestHandler} to send a structured JSON-RPC
* error back to the python bridge with a stable string ``code`` (e.g.
* ``"llm_disabled"``, ``"provider_error"``). The python side decodes the
* code from ``error.data.errorCode``.
*/
var BridgeRequestError = class extends Error {
	constructor(code, message, details) {
		super(message);
		this.code = code;
		this.details = details;
	}
};
var PythonBridge = class {
	proc = null;
	nextId = 1;
	pending = new Map();
	requestHandlers = new Map();
	readyPromise = null;
	closed = false;
	constructor(opts) {
		this.opts = opts;
	}
	/**
	* Register a handler for an outbound python→node RPC method (e.g.
	* ``host.llm.complete``). Subsequent registrations for the same method
	* replace the previous handler.
	*/
	onRequest(method, handler) {
		this.requestHandlers.set(method, handler);
	}
	/**
	* Build the default launch command from `opts.mode`. uvx by default; for
	* `nexus-source` mode, `matrixNexusRoot` must be set.
	*/
	buildDefaultCommand() {
		const mode = this.opts.mode ?? "uvx";
		if (mode === "nexus-source") {
			if (!this.opts.matrixNexusRoot) throw new Error("BridgeOptions.matrixNexusRoot is required when mode=\"nexus-source\"");
			return [
				"uv",
				"run",
				"--project",
				this.opts.matrixNexusRoot,
				"python",
				"-m",
				"anna_app_runtime_local.bridge"
			];
		}
		const version = this.opts.runtimeVersion ?? PINNED_RUNTIME_VERSION;
		return [
			"uvx",
			"--from",
			`anna-app-runtime-local@${version}`,
			"anna-app-bridge"
		];
	}
	/** Spawn the Python process and wait for the `_ready` notification. */
	async start() {
		if (this.proc) return;
		const cmd = this.opts.runtimeCommand ?? this.buildDefaultCommand();
		const [bin, ...args] = cmd;
		const useNexusPathInjection = !this.opts.runtimeCommand && (this.opts.mode ?? "uvx") === "nexus-source" && !!this.opts.matrixNexusRoot;
		const env = { ...process.env };
		if (useNexusPathInjection) {
			const wheelSrc = resolve(this.opts.matrixNexusRoot, "packages/anna-app-runtime-local/src");
			const existingPath = process.env.PYTHONPATH ?? "";
			env.PYTHONPATH = existingPath ? `${wheelSrc}${delimiter}${existingPath}` : wheelSrc;
		}
		const cwd = (this.opts.mode ?? "uvx") === "nexus-source" ? this.opts.matrixNexusRoot : process.cwd();
		if (!bin) throw new Error("PythonBridge: empty launch command (no executable)");
		this.proc = spawn(bin, args, {
			cwd,
			stdio: [
				"pipe",
				"pipe",
				"pipe"
			],
			env
		});
		this.proc.on("exit", (code, signal) => {
			this.closed = true;
			const err = new Error(`python bridge exited (code=${code} signal=${signal ?? "-"})`);
			for (const p of this.pending.values()) p.reject(err);
			this.pending.clear();
		});
		const stderrSink = this.opts.onStderr ?? ((l) => process.stderr.write(`[bridge] ${l}\n`));
		createInterface({ input: this.proc.stderr }).on("line", stderrSink);
		this.readyPromise = new Promise((resolve$1, reject) => {
			const rl = createInterface({ input: this.proc.stdout });
			let ready = false;
			rl.on("line", (line) => {
				if (!line.trim()) return;
				let env$1;
				try {
					env$1 = JSON.parse(line);
				} catch (e) {
					stderrSink(`non-json from bridge stdout: ${line}`);
					return;
				}
				if (!ready && env$1.method === "_ready") {
					ready = true;
					resolve$1();
					return;
				}
				if (typeof env$1.method === "string") {
					this.handleIncomingRequest(env$1);
					return;
				}
				this.handleResponse(env$1);
			});
			this.proc.on("error", (e) => reject(e));
			setTimeout(() => {
				if (!ready) reject(new Error("python bridge did not signal ready in 8s"));
			}, 8e3);
		});
		await this.readyPromise;
	}
	async handleIncomingRequest(env) {
		const method = env.method;
		const id = env.id;
		if (id == null) return;
		const params = env.params ?? {};
		const handler = this.requestHandlers.get(method);
		if (!handler) {
			this.writeFrame({
				jsonrpc: "2.0",
				id,
				error: {
					code: -32601,
					message: `method not found: ${method}`
				}
			});
			return;
		}
		try {
			const result = await handler(params);
			this.writeFrame({
				jsonrpc: "2.0",
				id,
				result: result ?? {}
			});
		} catch (e) {
			if (e instanceof BridgeRequestError) this.writeFrame({
				jsonrpc: "2.0",
				id,
				error: {
					code: -32e3,
					message: e.message,
					data: {
						errorCode: e.code,
						...e.details ?? {}
					}
				}
			});
			else this.writeFrame({
				jsonrpc: "2.0",
				id,
				error: {
					code: -32603,
					message: e.message,
					data: { errorCode: "internal_error" }
				}
			});
		}
	}
	writeFrame(env) {
		if (!this.proc) return;
		this.proc.stdin.write(`${JSON.stringify(env)}\n`);
	}
	handleResponse(env) {
		const id = env.id;
		if (id == null) return;
		const pending = this.pending.get(id);
		if (!pending) return;
		this.pending.delete(id);
		if (env.error) {
			const e = env.error;
			const err = new Error(`[${e.code}] ${e.message}`);
			err.rpc = e;
			pending.reject(err);
		} else pending.resolve(env.result);
	}
	/** Send one JSON-RPC request and await its response. */
	call(method, params = {}) {
		if (this.closed || !this.proc) return Promise.reject(new Error("python bridge not running"));
		const id = this.nextId++;
		const env = {
			jsonrpc: "2.0",
			id,
			method,
			params
		};
		return new Promise((resolve$1, reject) => {
			this.pending.set(id, {
				resolve: (v) => resolve$1(v),
				reject
			});
			this.proc.stdin.write(`${JSON.stringify(env)}\n`);
		});
	}
	async stop() {
		if (!this.proc) return;
		this.closed = true;
		this.proc.stdin.end();
		this.proc.kill("SIGTERM");
		this.proc = null;
	}
};

//#endregion
export { BridgeRequestError, PINNED_RUNTIME_VERSION, PythonBridge };
