// @anna-ai/cli/dist/bundled-executas-_H1A9M50.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { CliError } from "./client-D-_z1ALk.js";
import { dirname, resolve } from "node:path";
import { existsSync, mkdirSync, readFileSync, renameSync, unlinkSync, writeFileSync } from "node:fs";

//#region src/publish/bundled-executas.ts
const BUNDLED_PREFIX = "bundled:";
const LOCK_DIR = ".anna";
const LOCK_FILE = "executas.lock.json";
/** Parse repeated `--executa-id handle=tool-id` flags. */
function parseExecutaIdOverrides(raw) {
	if (!raw || raw.length === 0) return [];
	return raw.map((entry) => {
		const eq = entry.indexOf("=");
		if (eq <= 0) throw new CliError(`--executa-id must be in the form handle=tool-id (got "${entry}")`, 2);
		const handle = entry.slice(0, eq).trim();
		const toolId = entry.slice(eq + 1).trim();
		if (!handle || !toolId) throw new CliError(`--executa-id must be in the form handle=tool-id (got "${entry}")`, 2);
		return {
			handle,
			toolId
		};
	});
}
function lockPath(cwd) {
	return resolve(cwd, LOCK_DIR, LOCK_FILE);
}
function readExecutasLock(cwd, host) {
	const p = lockPath(cwd);
	if (!existsSync(p)) return null;
	try {
		const raw = JSON.parse(readFileSync(p, "utf-8"));
		if (raw.executas && typeof raw.executas === "object" && typeof raw.host === "string" && canonicalHost(raw.host) === canonicalHost(host)) return {
			$schema: "anna-executas-lock/v1",
			host: canonicalHost(raw.host),
			executas: raw.executas
		};
	} catch {}
	return null;
}
function writeExecutasLock(cwd, lock) {
	const p = lockPath(cwd);
	mkdirSync(dirname(p), { recursive: true });
	const tmp = `${p}.tmp.${process.pid}`;
	writeFileSync(tmp, JSON.stringify({
		...lock,
		host: canonicalHost(lock.host)
	}, null, 2) + "\n", "utf-8");
	try {
		renameSync(tmp, p);
	} catch (e) {
		try {
			unlinkSync(tmp);
		} catch {}
		throw e;
	}
}
/**
* Walk a JSON value and collect every `bundled:<handle>` token found in
* a string. Tokens can be bare (`bundled:foo`) or prefixed
* (`required:bundled:foo`, `optional:bundled:foo`).
*/
function collectBundledHandles(value, out) {
	if (typeof value === "string") {
		const h = extractHandle(value);
		if (h) out.add(h);
		return;
	}
	if (Array.isArray(value)) {
		for (const v of value) collectBundledHandles(v, out);
		return;
	}
	if (value && typeof value === "object") for (const v of Object.values(value)) collectBundledHandles(v, out);
}
function extractHandle(s) {
	let body = s;
	if (body.startsWith("required:")) body = body.slice(9);
	else if (body.startsWith("optional:")) body = body.slice(9);
	if (body.startsWith(BUNDLED_PREFIX)) return body.slice(BUNDLED_PREFIX.length);
	return null;
}
/**
* Return a deep copy of `value` with every `bundled:<handle>` token
* replaced by the resolved tool_id (preserving any required:/optional:
* prefix).
*/
function substituteBundledRefs(value, resolved) {
	if (typeof value === "string") return substituteString(value, resolved);
	if (Array.isArray(value)) return value.map((v) => substituteBundledRefs(v, resolved));
	if (value && typeof value === "object") {
		const out = {};
		for (const [k, v] of Object.entries(value)) out[k] = substituteBundledRefs(v, resolved);
		return out;
	}
	return value;
}
function substituteString(s, resolved) {
	let prefix = "";
	let body = s;
	if (body.startsWith("required:")) {
		prefix = "required:";
		body = body.slice(prefix.length);
	} else if (body.startsWith("optional:")) {
		prefix = "optional:";
		body = body.slice(prefix.length);
	}
	if (body.startsWith(BUNDLED_PREFIX)) {
		const handle = body.slice(BUNDLED_PREFIX.length);
		const id = resolved[handle];
		if (!id) throw new CliError(`unresolved bundled handle "${handle}"`, 4);
		return prefix + id;
	}
	return s;
}
/**
* Write the Option-A sidecar `<bundleDir>/anna-tool-ids.js`.
*
* The bundle can read it via:
*
*     <script src="anna-tool-ids.js"></script>
*     const TOOL_ID = window.__ANNA_TOOL_IDS__["focus-session"];
*/
function writeBundleToolIdSidecar(bundleDir, resolved) {
	const file = resolve(bundleDir, "anna-tool-ids.js");
	const body = "/* AUTO-GENERATED by `anna-app apps publish`. Do not edit. */\n/* Maps local bundled_executas handles → server-minted tool_ids. */\nwindow.__ANNA_TOOL_IDS__ = " + JSON.stringify(resolved, null, 2) + ";\n";
	mkdirSync(bundleDir, { recursive: true });
	const tmp = `${file}.tmp.${process.pid}`;
	writeFileSync(tmp, body, "utf-8");
	try {
		renameSync(tmp, file);
	} catch (e) {
		try {
			unlinkSync(tmp);
		} catch {}
		throw e;
	}
	return file;
}
/**
* Validate that the declared `bundled_executas` map, the `bundled:`
* tokens used in the manifest, and any `--executa-id` overrides are
* mutually consistent.
*
* - Every `bundled:<handle>` token referenced in the manifest must have
*   a matching entry in `bundled_executas` OR a `--executa-id` override.
* - Every override handle must be referenced or declared (typo guard).
*/
function validateBundledHandles(manifestRaw, declared, overrides) {
	const referenced = new Set();
	collectBundledHandles(manifestRaw, referenced);
	const declaredHandles = new Set(declared.map((d) => d.handle));
	const overrideHandles = new Set(overrides.map((o) => o.handle));
	const unresolvable = [];
	for (const h of referenced) if (!declaredHandles.has(h) && !overrideHandles.has(h)) unresolvable.push(h);
	if (unresolvable.length > 0) throw new CliError(`manifest references bundled:${unresolvable.sort().join(", bundled:")} but no matching app.json bundled_executas entry or --executa-id override was found`, 4);
	for (const h of overrideHandles) if (!referenced.has(h) && !declaredHandles.has(h)) throw new CliError(`--executa-id "${h}=…" does not match any bundled:${h} reference or app.json bundled_executas entry (typo?)`, 2);
	return { referenced };
}

//#endregion
export { collectBundledHandles, lockPath, parseExecutaIdOverrides, readExecutasLock, substituteBundledRefs, validateBundledHandles, writeBundleToolIdSidecar, writeExecutasLock };
