// @anna-ai/cli/dist/client-D-_z1ALk.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";

//#region src/api/client.ts
var CliError = class extends Error {
	exitCode;
	status;
	body;
	constructor(message, exitCode, opts) {
		super(message);
		this.exitCode = exitCode;
		this.status = opts?.status;
		this.body = opts?.body;
	}
};
var ScopeError = class extends CliError {
	needed;
	missing;
	have;
	constructor(opts) {
		super(`missing PAT scope(s): ${opts.missing.join(", ") || "(unknown)"}`, 3, {
			status: 403,
			body: opts.body
		});
		this.needed = opts.needed;
		this.missing = opts.missing;
		this.have = opts.have;
	}
};
var NexusClient = class {
	host;
	pat;
	fetchImpl;
	userAgent;
	constructor(opts) {
		this.host = canonicalHost(opts.host);
		this.pat = opts.pat;
		this.fetchImpl = opts.fetchImpl ?? fetch;
		this.userAgent = opts.userAgent ?? "anna-app-cli";
	}
	async request(opts) {
		const url = new URL(this.host + opts.path);
		if (opts.query) for (const [k, v] of Object.entries(opts.query)) {
			if (v === void 0 || v === null) continue;
			url.searchParams.set(k, String(v));
		}
		const headers = {
			authorization: `Bearer ${this.pat}`,
			"user-agent": this.userAgent,
			accept: "application/json",
			...opts.headers ?? {}
		};
		let body;
		if (opts.multipart) {
			const fd = new FormData();
			for (const [k, v] of Object.entries(opts.multipart.fields)) fd.append(k, v);
			body = fd;
		} else if (opts.body !== void 0) {
			headers["content-type"] = "application/json";
			body = JSON.stringify(opts.body);
		}
		const method = opts.method ?? "GET";
		const allowed = new Set(opts.allowStatuses ?? []);
		let res;
		try {
			res = await this.fetchImpl(url, {
				method,
				headers,
				body
			});
		} catch (e) {
			throw new CliError(`network error contacting ${this.host}: ${e.message}`, 6);
		}
		if (res.status >= 500 && !opts.noRetry && !allowed.has(res.status) && method === "GET") try {
			res = await this.fetchImpl(url, {
				method,
				headers,
				body
			});
		} catch (e) {
			throw new CliError(`network error contacting ${this.host}: ${e.message}`, 6);
		}
		const idempotentHit = res.headers.get("x-anna-idempotent") === "hit";
		let parsed = null;
		const text = await res.text();
		if (text.length > 0) try {
			parsed = JSON.parse(text);
		} catch {
			parsed = text;
		}
		if (!res.ok && !allowed.has(res.status)) this.throwForStatus(res.status, parsed);
		return {
			data: parsed,
			status: res.status,
			headers: res.headers,
			idempotentHit
		};
	}
	throwForStatus(status, body) {
		if (status === 401) throw new CliError("PAT rejected by server (401). Run `anna-app login --host …` to refresh.", 2, {
			status,
			body
		});
		if (status === 403) {
			const det = extractDetail(body);
			if (det && typeof det === "object" && det.detail === "missing_scope") {
				const dd = det;
				throw new ScopeError({
					needed: dd.needed ?? [],
					missing: dd.missing ?? [],
					have: dd.have ?? [],
					body
				});
			}
			throw new CliError(`forbidden (403): ${describeBody(body)}`, 3, {
				status,
				body
			});
		}
		if (status >= 500) throw new CliError(`server error (${status}): ${describeBody(body)}`, 6, {
			status,
			body
		});
		throw new CliError(`request failed (${status}): ${describeBody(body)}`, 5, {
			status,
			body
		});
	}
};
function extractDetail(body) {
	if (body && typeof body === "object" && "detail" in body) {
		const d = body.detail;
		if (typeof d === "string") return { detail: d };
		if (d && typeof d === "object") return d;
	}
	return null;
}
function describeBody(body) {
	if (typeof body === "string") return body.slice(0, 240);
	if (body && typeof body === "object") {
		const det = body.detail;
		if (typeof det === "string") return det;
		try {
			return JSON.stringify(body).slice(0, 240);
		} catch {}
	}
	return String(body ?? "(empty body)");
}

//#endregion
export { CliError, NexusClient, ScopeError };
