// @anna-ai/cli/dist/credentials-BTv2IfUZ.js
import { createRequire } from "module";
import { dirname, join } from "node:path";
import { chmodSync, existsSync, mkdirSync, readFileSync, statSync, unlinkSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";

//#region rolldown:runtime
var __require = /* @__PURE__ */ createRequire(import.meta.url);

//#endregion
//#region src/credentials.ts
/** Resolve credentials file path. Honours $XDG_CONFIG_HOME. */
function credentialsPath() {
	const xdg = process.env.XDG_CONFIG_HOME;
	const base = xdg && xdg.length > 0 ? xdg : join(homedir(), ".config");
	return join(base, "anna", "credentials.json");
}
/** Normalise a host URL to its canonical key (https://x.example.com). */
function canonicalHost(input) {
	let url;
	try {
		url = new URL(input);
	} catch {
		url = new URL(`https://${input}`);
	}
	const port = url.port ? `:${url.port}` : "";
	return `${url.protocol}//${url.hostname}${port}`;
}
/** Read credentials file (returns empty container if missing or unreadable). */
function readCredentials() {
	const path = credentialsPath();
	if (!existsSync(path)) return {
		version: 1,
		current: null,
		accounts: {}
	};
	try {
		const raw = readFileSync(path, "utf8");
		const parsed = JSON.parse(raw);
		return {
			version: 1,
			current: parsed.current ?? null,
			accounts: parsed.accounts ?? {}
		};
	} catch {
		return {
			version: 1,
			current: null,
			accounts: {}
		};
	}
}
/** Atomically write credentials file (chmod 600). */
function writeCredentials(data) {
	const path = credentialsPath();
	mkdirSync(dirname(path), {
		recursive: true,
		mode: 448
	});
	const tmp = `${path}.tmp.${process.pid}`;
	writeFileSync(tmp, JSON.stringify(data, null, 2), { mode: 384 });
	try {
		__require("node:fs").renameSync(tmp, path);
	} catch (e) {
		try {
			unlinkSync(tmp);
		} catch {}
		throw e;
	}
	try {
		chmodSync(path, 384);
	} catch {}
}
/** Set or replace the entry for `host`; mark as `current`. */
function saveAccount(rec) {
	const data = readCredentials();
	const key = canonicalHost(rec.host);
	data.accounts[key] = {
		...rec,
		host: key
	};
	data.current = key;
	writeCredentials(data);
}
/** Remove one account; if it was current, switch to any remaining one. */
function removeAccount(host) {
	const data = readCredentials();
	const key = canonicalHost(host);
	if (!(key in data.accounts)) return false;
	delete data.accounts[key];
	if (data.current === key) {
		const next = Object.keys(data.accounts);
		data.current = next.length > 0 ? next[0] ?? null : null;
	}
	writeCredentials(data);
	return true;
}
/** Look up an account by host (or current account if `host` omitted). */
function getAccount(host) {
	const data = readCredentials();
	const key = host ? canonicalHost(host) : data.current;
	if (!key) return null;
	return data.accounts[key] ?? null;
}
/** Mask a PAT for safe display: ``eyJh…aZQp (90d, scopes=aps:dev)``. */
function maskPat(pat) {
	if (pat.length <= 12) return "***";
	return `${pat.slice(0, 4)}…${pat.slice(-4)}`;
}
/** Permission check — warn callers if file mode is loose. */
function credentialsAreLooselyPermissioned() {
	const path = credentialsPath();
	if (!existsSync(path)) return false;
	try {
		const st = statSync(path);
		return (st.mode & 63) !== 0;
	} catch {
		return false;
	}
}

//#endregion
export { canonicalHost, credentialsAreLooselyPermissioned, credentialsPath, getAccount, maskPat, readCredentials, removeAccount, saveAccount, writeCredentials };
