// @anna-ai/cli/dist/dev-aWkfWGLD.js
import { findMatrixNexusRoot, nexusSchemaDir } from "./nexus-root-BlPwOusj.js";
import { dirname, isAbsolute, resolve } from "node:path";
import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { bold, cyan, dim, green, red, yellow } from "kleur/colors";
import { parse } from "smol-toml";

//#region src/commands/dev.ts
async function runDev(opts) {
	const cwd = resolve(opts.cwd);
	const manifestPath = isAbsolute(opts.manifestPath) ? opts.manifestPath : resolve(cwd, opts.manifestPath);
	if (!existsSync(manifestPath)) {
		console.error(red(`✗ manifest not found: ${manifestPath}`));
		return 2;
	}
	let manifest;
	try {
		manifest = JSON.parse(readFileSync(manifestPath, "utf-8"));
	} catch (e) {
		console.error(red(`✗ manifest is not valid json: ${e.message}`));
		return 2;
	}
	const slug = opts.slug ?? deriveSlug(manifest, manifestPath);
	const bundleDir = opts.bundleDir ? resolve(cwd, opts.bundleDir) : resolve(dirname(manifestPath), "bundle");
	const bundleEntry = manifest.ui?.bundle?.entry ?? "index.html";
	if (!existsSync(bundleDir)) {
		console.error(red(`✗ bundle dir not found: ${bundleDir}`));
		return 2;
	}
	if (!existsSync(resolve(bundleDir, bundleEntry))) {
		console.error(red(`✗ bundle entry "${bundleEntry}" not found under ${bundleDir}`));
		return 2;
	}
	const matrixNexusRoot = await resolveMatrixNexusRoot(opts.matrixNexusRoot, cwd);
	const mode = matrixNexusRoot ? "nexus-source" : "uvx";
	let schemaSource = null;
	if (matrixNexusRoot && !process.env.ANNA_APP_SCHEMA_DIR) {
		schemaSource = nexusSchemaDir(matrixNexusRoot);
		if (schemaSource) process.env.ANNA_APP_SCHEMA_DIR = schemaSource;
	}
	const storageMode = opts.storageMode === "aps" ? "aps" : "legacy";
	if (opts.storageMode && opts.storageMode !== "legacy" && opts.storageMode !== "aps") {
		console.error(red(`✗ --storage must be "legacy" or "aps", got "${opts.storageMode}"`));
		return 2;
	}
	if (storageMode === "aps") {
		if (opts.noLlm || opts.mockLlm) {
			console.error(red("✗ --storage aps requires a real LLM bridge (PAT on disk); drop --no-llm / --mock-llm or switch back to --storage legacy."));
			return 2;
		}
		process.env.ANNA_APP_RUNTIME_STORAGE_MODE = "aps";
	}
	const { PythonBridge, PINNED_RUNTIME_VERSION } = await import("./bridge-CmoKMMcG.js");
	const { HarnessServer } = await import("./server-DtnMLKQS.js");
	const bridge = new PythonBridge({
		mode,
		matrixNexusRoot: matrixNexusRoot ?? void 0,
		onStderr: (line) => process.stderr.write(dim(`[bridge] ${line}\n`))
	});
	console.log(bold(cyan("anna-app dev")));
	console.log(`  manifest          ${dim(manifestPath)}`);
	console.log(`  bundle            ${dim(`${bundleDir}/${bundleEntry}`)}`);
	if (mode === "nexus-source") {
		console.log(`  matrix-nexus root ${dim(matrixNexusRoot)}`);
		console.log(`  runtime           ${dim("nexus-source (uv run)")}`);
		console.log(`  browser sdk       ${dim("in-tree packages/anna-app-runtime/dist")}`);
		console.log(`  schema            ${dim(schemaSource ? "in-tree packages/anna-app-schema" : process.env.ANNA_APP_SCHEMA_DIR ? `override ${process.env.ANNA_APP_SCHEMA_DIR}` : "npm @anna-ai/app-schema")}`);
	} else {
		console.log(`  runtime           ${dim(`uvx anna-app-runtime-local@${PINNED_RUNTIME_VERSION}`)}`);
		console.log(`  browser sdk       ${dim("npm @anna-ai/app-runtime")}`);
	}
	console.log(`  spawning python bridge…`);
	try {
		await bridge.start();
	} catch (e) {
		console.error(red(`✗ bridge failed to start: ${e.message}`));
		if (mode === "uvx") console.error(dim("  (try: install uv from https://docs.astral.sh/uv/, or pass --matrix-nexus-root to use a checkout)"));
		else console.error(dim("  (try: cd matrix-nexus && uv sync; ensure `uv` is on PATH; or unset --matrix-nexus-root to use uvx)"));
		return 2;
	}
	try {
		const pong = await bridge.call("ping");
		if (!pong.pong) throw new Error("bridge ping returned non-pong");
	} catch (e) {
		console.error(red(`✗ bridge ping failed: ${e.message}`));
		await bridge.stop();
		return 2;
	}
	console.log(green("  ✓ bridge ready"));
	const executas = opts.executas ?? autoDiscoverExecutas(dirname(manifestPath));
	if (executas.length > 0) console.log(`  executas          ${dim(executas.map((e) => e.tool_id).join(", "))}`);
	const bundledResolved = resolveBundledHandles(dirname(manifestPath), executas);
	if (Object.keys(bundledResolved).length > 0) {
		const { substituteBundledRefs, writeBundleToolIdSidecar } = await import("./bundled-executas-06keEPP_.js");
		try {
			manifest = substituteBundledRefs(manifest, bundledResolved);
			writeBundleToolIdSidecar(bundleDir, bundledResolved);
			console.log(`  bundled handles   ${dim(Object.entries(bundledResolved).map(([h, id]) => `${h}→${id}`).join(", "))}`);
		} catch (e) {
			console.error(red(`✗ bundled handle resolution failed: ${e.message}`));
			await bridge.stop();
			return 2;
		}
	}
	const llm = opts.noLlm ? { mode: "off" } : opts.mockLlm ? {
		mode: "mock",
		mockFile: opts.mockLlm
	} : await resolveRealLlm({
		cwd,
		account: opts.llmAccount,
		appSlug: opts.llmAppSlug ?? slug,
		manifest
	});
	if (llm === null) return 2;
	llm.storageMode = storageMode;
	console.log(`  llm bridge        ${dim(llm.mode === "off" ? "disabled (--no-llm)" : llm.mode === "mock" ? `mock (${opts.mockLlm})` : `real${opts.llmAccount ? ` [${opts.llmAccount}]` : ""} → app_slug=${llm.appSlug}`)}`);
	console.log(`  storage backend   ${dim(storageMode === "aps" ? "aps (real nexus APS via /api/v1/storage/*)" : "legacy (in-memory runtime_state)")}`);
	const server = new HarnessServer({
		slug,
		manifest,
		bundleDir,
		bundleEntry,
		view: opts.view,
		matrixNexusRoot,
		userId: opts.userId,
		port: opts.port,
		executas,
		watch: !opts.noWatch,
		llm
	}, bridge);
	try {
		await server.listen();
	} catch (e) {
		const msg = e.code === "EADDRINUSE" ? `port ${opts.port} already in use` : e.message;
		console.error(red(`✗ harness failed to listen: ${msg}`));
		await bridge.stop();
		return 2;
	}
	console.log(`  ${green("✓")} dashboard ${cyan(`http://localhost:${opts.port}/`)}`);
	console.log(yellow("  press Ctrl+C to stop"));
	const shutdown = async () => {
		console.log(dim("\n  shutting down…"));
		try {
			await server.close();
			await bridge.stop();
		} finally {
			process.exit(0);
		}
	};
	process.once("SIGINT", shutdown);
	process.once("SIGTERM", shutdown);
	await new Promise(() => {});
	return 0;
}
/**
* Resolve the `real` LlmBridgeOptions: load the dev PAT, register
* (or re-use) an `AnnaApp(is_dev=True)` on nexus, cache the result.
*
* Returns `null` (after printing) when registration fails so the caller
* can exit cleanly. Otherwise returns a bridge config carrying the
* registered `appSlug`.
*/
async function resolveRealLlm(args) {
	const { getAccount } = await import("./credentials-Chkoidh5.js");
	const { ensureDevAppRegistered } = await import("./dev-app-cache-CP2P8SCZ.js");
	const acc = getAccount(args.account);
	if (!acc) {
		console.error(red("✗ no developer PAT on disk — run `anna-app login --host <nexus-url>` first.\n  (or use `--no-llm` / `--mock-llm <fixture>` to develop offline.)"));
		return null;
	}
	if (acc.expires_at && acc.expires_at < Math.floor(Date.now() / 1e3)) {
		console.error(red("✗ PAT expired — run `anna-app login` again."));
		return null;
	}
	const manifest = args.manifest;
	const registerInput = {
		slug: args.appSlug,
		name: manifest.name,
		category: manifest.category,
		tagline: manifest.tagline
	};
	try {
		const entry = await ensureDevAppRegistered({
			cwd: args.cwd,
			host: acc.host,
			pat: acc.pat,
			input: registerInput
		});
		return {
			mode: "real",
			account: args.account,
			appSlug: entry.slug,
			onAppSlugNotFound: async () => {
				const { invalidateDevAppCache } = await import("./dev-app-cache-CP2P8SCZ.js");
				invalidateDevAppCache(args.cwd);
				const fresh = getAccount(args.account);
				if (!fresh) throw new Error("PAT not found while re-registering dev app — run `anna-app login` again");
				await ensureDevAppRegistered({
					cwd: args.cwd,
					host: fresh.host,
					pat: fresh.pat,
					input: registerInput
				});
			}
		};
	} catch (e) {
		console.error(red(`✗ failed to register dev app on nexus: ${e.message}`));
		return null;
	}
}
function deriveSlug(manifest, path) {
	const fromMan = manifest.slug ?? manifest.name;
	if (typeof fromMan === "string" && fromMan) return fromMan;
	return dirname(path).split(/[\\/]/).pop() || "anna-app-dev";
}
/**
* Map each `app.json` `bundled_executas` handle to the concrete dev
* `tool_id` of the executa discovered at its `path`. Mirrors PUBLISH's
* handle→minted-id resolution so `anna-app dev` can rewrite
* `bundled:<handle>` refs in the manifest. Returns `{}` when there is no
* `app.json` or no `bundled_executas`. Unresolvable handles are warned and
* skipped (the leftover `bundled:` ref will then fail the whitelist, which
* surfaces the misconfiguration loudly).
*/
function resolveBundledHandles(manifestDir, executas) {
	const appJsonPath = resolve(manifestDir, "app.json");
	if (!existsSync(appJsonPath)) return {};
	let appJson;
	try {
		appJson = JSON.parse(readFileSync(appJsonPath, "utf-8"));
	} catch (e) {
		console.warn(yellow(`⚠ app.json is not valid json: ${e.message}`));
		return {};
	}
	const bundled = appJson.bundled_executas;
	if (!bundled || typeof bundled !== "object") return {};
	const byDir = new Map();
	for (const ex of executas) byDir.set(resolve(ex.project_dir), ex.tool_id);
	const resolved = {};
	for (const [handle, spec] of Object.entries(bundled)) {
		const p = spec?.path;
		if (!p) {
			console.warn(yellow(`⚠ bundled_executas["${handle}"] missing "path"`));
			continue;
		}
		const dir = isAbsolute(p) ? p : resolve(manifestDir, p);
		const toolId = byDir.get(resolve(dir));
		if (!toolId) {
			console.warn(yellow(`⚠ bundled handle "${handle}" → ${p}: no executa discovered at that path (check executas/<dir>/executa.json "tool_id")`));
			continue;
		}
		resolved[handle] = toolId;
	}
	return resolved;
}
function autoDiscoverExecutas(manifestDir) {
	const root = resolve(manifestDir, "executas");
	if (!existsSync(root)) return [];
	const found = [];
	const seen = new Map();
	for (const name of readdirSync(root)) {
		if (name.startsWith(".")) continue;
		const dir = resolve(root, name);
		let st;
		try {
			st = statSync(dir);
		} catch {
			continue;
		}
		if (!st.isDirectory()) continue;
		const result = discoverOne(dir, name);
		if (result === "skip" || result === "disabled") continue;
		if (result instanceof Error) {
			console.warn(yellow(`⚠ skipping executa ${name}: ${result.message}`));
			continue;
		}
		const prior = seen.get(result.tool_id);
		if (prior) {
			console.warn(yellow(`⚠ skipping executa ${name}: tool_id "${result.tool_id}" already provided by ${prior} (toggle "enabled": false in executa.json to silence)`));
			continue;
		}
		seen.set(result.tool_id, name);
		found.push(result);
	}
	return found;
}
function discoverOne(dir, name, opts = { respectEnabled: true }) {
	const cfgPath = resolve(dir, "executa.json");
	if (existsSync(cfgPath)) {
		let cfg;
		try {
			cfg = JSON.parse(readFileSync(cfgPath, "utf-8"));
		} catch (err) {
			return new Error(`failed to parse executa.json (${err.message})`);
		}
		if (cfg.enabled === false && opts.respectEnabled !== false) return "disabled";
		if (!cfg.tool_id) return new Error("executa.json missing required `tool_id`");
		if (!cfg.type) return new Error("executa.json missing required `type`");
		if (cfg.command && cfg.command.length > 0) return {
			tool_id: cfg.tool_id,
			project_dir: dir,
			command: cfg.command
		};
		const cmd = defaultCommand(cfg.type, dir, cfg.tool_id);
		if (cmd instanceof Error) return cmd;
		return {
			tool_id: cfg.tool_id,
			project_dir: dir,
			command: cmd
		};
	}
	if (existsSync(resolve(dir, "pyproject.toml"))) return discoverPython(dir, name);
	if (existsSync(resolve(dir, "package.json"))) return discoverNode(dir, name);
	if (existsSync(resolve(dir, "go.mod"))) return new Error("Go executa requires an explicit executa.json with `tool_id` and `type: \"go\"` (go.mod has no script-name field)");
	const binCandidate = resolve(dir, "bin", name);
	if (existsSync(binCandidate)) return {
		tool_id: name,
		project_dir: dir,
		command: [binCandidate]
	};
	return "skip";
}
function defaultCommand(type, dir, toolId) {
	switch (type) {
		case "python": {
			const script = pythonScriptName(dir);
			if (script instanceof Error) return script;
			return [
				"uv",
				"run",
				"--project",
				dir,
				script ?? toolId
			];
		}
		case "node": {
			const pkg = readPackageJson(dir);
			if (!pkg) return new Error("type=node but no package.json found");
			const entry = nodeEntry(pkg, toolId, dir);
			if (entry instanceof Error) return entry;
			return ["node", entry];
		}
		case "go": return [
			"go",
			"run",
			"."
		];
		case "binary": {
			const guess = resolve(dir, "bin", toolId);
			if (!existsSync(guess)) return new Error(`type=binary but no executable at ${guess}; specify "command" in executa.json explicitly`);
			return [guess];
		}
		default: return new Error(`unknown type "${type}" — expected python | node | go | binary`);
	}
}
function discoverPython(dir, name) {
	const script = pythonScriptName(dir);
	if (script instanceof Error) return script;
	if (script === null) return new Error("no [project.scripts] entry in pyproject.toml");
	return {
		tool_id: script,
		project_dir: dir,
		command: [
			"uv",
			"run",
			"--project",
			dir,
			script
		]
	};
}
/**
* First `[project.scripts]` entry name from a dir's pyproject.toml — the
* console-script `uv run` should launch. Returns `null` when there's no
* pyproject / no scripts table (caller decides the fallback), or an Error
* when pyproject exists but can't be parsed.
*/
function pythonScriptName(dir) {
	const py = resolve(dir, "pyproject.toml");
	if (!existsSync(py)) return null;
	let body;
	try {
		body = readFileSync(py, "utf-8");
	} catch {
		return new Error("could not read pyproject.toml");
	}
	let parsed;
	try {
		parsed = parse(body);
	} catch (err) {
		return new Error(`failed to parse pyproject.toml (${err.message})`);
	}
	const scripts = parsed?.project?.scripts;
	if (!scripts || typeof scripts !== "object") return null;
	const keys = Object.keys(scripts);
	if (keys.length === 0) return null;
	return keys[0];
}
function discoverNode(dir, name) {
	const pkg = readPackageJson(dir);
	if (!pkg) return new Error("could not read package.json");
	const toolId = (typeof pkg.executa === "object" && pkg.executa ? pkg.executa.tool_id : void 0) ?? firstBinKey(pkg) ?? (typeof pkg.name === "string" ? pkg.name : void 0);
	if (!toolId) return new Error("package.json missing tool_id — set `executa.tool_id`, `bin`, or `name`");
	const entry = nodeEntry(pkg, toolId, dir);
	if (entry instanceof Error) return entry;
	return {
		tool_id: toolId,
		project_dir: dir,
		command: ["node", entry]
	};
}
function readPackageJson(dir) {
	const p = resolve(dir, "package.json");
	if (!existsSync(p)) return null;
	try {
		return JSON.parse(readFileSync(p, "utf-8"));
	} catch {
		return null;
	}
}
function firstBinKey(pkg) {
	if (typeof pkg.bin === "string") return typeof pkg.name === "string" ? pkg.name : void 0;
	if (pkg.bin && typeof pkg.bin === "object") {
		const keys = Object.keys(pkg.bin);
		return keys[0];
	}
	return void 0;
}
function nodeEntry(pkg, toolId, dir) {
	let rel;
	if (typeof pkg.bin === "string") rel = pkg.bin;
	else if (pkg.bin && typeof pkg.bin === "object") {
		const obj = pkg.bin;
		const v = obj[toolId];
		if (typeof v === "string") rel = v;
		else {
			const first = Object.values(obj).find((x) => typeof x === "string");
			if (typeof first === "string") rel = first;
		}
	}
	if (!rel && typeof pkg.main === "string") rel = pkg.main;
	if (!rel && typeof pkg.module === "string") rel = pkg.module;
	if (!rel) return new Error("package.json has no usable entry — set `bin`, `main`, or override `command` in executa.json");
	const abs = isAbsolute(rel) ? rel : resolve(dir, rel);
	if (!existsSync(abs)) return new Error(`node entry not found: ${abs}`);
	return abs;
}
/**
* Parse `--executa <spec>` flag values from CLI. Spec syntax (comma-separated
* key=value pairs, all keys lowercase):
*
*   --executa dir=./executas/foo,type=node
*   --executa dir=/abs/path,tool_id=tool-x,command="node plugin.js"
*
* Required: `dir`. Optional: `tool_id`, `type`, `command` (space-split unless
* already an array via JSON), `cwd`. When `tool_id`/`type`/`command` are
* omitted, falls back to the same rules as `executa.json` discovery.
*/
function parseExecutaSpec(spec, cwd) {
	const parts = spec.split(",").map((s) => s.trim()).filter(Boolean);
	const kv = {};
	for (const p of parts) {
		const eq = p.indexOf("=");
		if (eq === -1) return new Error(`bad --executa spec part: "${p}"`);
		kv[p.slice(0, eq).trim().toLowerCase()] = p.slice(eq + 1).trim();
	}
	const dirRaw = kv.dir;
	if (!dirRaw) return new Error("--executa spec missing `dir=`");
	const dir = isAbsolute(dirRaw) ? dirRaw : resolve(cwd, dirRaw);
	if (!existsSync(dir)) return new Error(`--executa dir not found: ${dir}`);
	if (!kv.tool_id && !kv.type && !kv.command) {
		const result = discoverOne(dir, dir.split(/[\\/]/).pop() ?? dir, { respectEnabled: false });
		if (result === "skip") return new Error(`nothing discoverable at ${dir} (no executa.json / pyproject.toml / package.json / bin/)`);
		if (result === "disabled") return new Error(`executa.json at ${dir} has "enabled": false`);
		if (result instanceof Error) return result;
		return result;
	}
	const tool_id = kv.tool_id;
	if (!tool_id) return new Error("--executa spec missing `tool_id=`");
	let command;
	if (kv.command) if (kv.command.startsWith("[")) try {
		const arr = JSON.parse(kv.command);
		if (!Array.isArray(arr) || !arr.every((x) => typeof x === "string")) return new Error("--executa command JSON must be a string array");
		command = arr;
	} catch (err) {
		return new Error(`--executa command JSON invalid: ${err.message}`);
	}
	else command = kv.command.split(/\s+/);
	else if (kv.type) {
		const c = defaultCommand(kv.type, dir, tool_id);
		if (c instanceof Error) return c;
		command = c;
	}
	return {
		tool_id,
		project_dir: dir,
		command
	};
}
async function resolveMatrixNexusRoot(explicit, cwd) {
	return findMatrixNexusRoot(explicit, cwd);
}

//#endregion
export { parseExecutaSpec, runDev };
