// @anna-ai/cli/dist/dev-account-WXEjIE16.js
import { canonicalHost, getAccount } from "./credentials-BTv2IfUZ.js";

//#region src/executa/dev-account.ts
/** Embed a JSON-RPC error code on a thrown Error so the runner can forward it. */
function withCode(err, code) {
	err.rpcCode = code;
	return err;
}
/**
* Resolve a saved developer PAT for the given account host (or the
* "current" one when omitted).
*
* @throws JSON-RPC `-32001` when no PAT is on disk or it has expired.
*/
function requireAccount(account) {
	const acc = getAccount(account);
	if (!acc) throw withCode(new Error("no PAT on disk — run `anna-app login --host <nexus-url>` first (or use a `mock`/`off`/`memory` mode for the affected bridge)"), -32001);
	if (acc.expires_at && acc.expires_at < Math.floor(Date.now() / 1e3)) throw withCode(new Error("PAT expired — run `anna-app login` again"), -32001);
	return acc;
}
/** Return the canonical `https://host[:port]` form of a saved account. */
function hostOf(acc) {
	return canonicalHost(acc.host);
}
/**
* Call `POST /api/v1/anna-apps/dev/session/mint` and parse the response.
* Thin wrapper used by both sampling and agent bridges.
*/
async function mintAppSession(host, body) {
	const url = `${canonicalHost(host)}/api/v1/anna-apps/dev/session/mint`;
	const res = await fetch(url, {
		method: "POST",
		headers: { "content-type": "application/json" },
		body: JSON.stringify(body)
	});
	if (!res.ok) {
		const text = await res.text().catch(() => "");
		throw withCode(new Error(`session.mint failed: HTTP ${res.status} ${text}`), -32e3);
	}
	return await res.json();
}

//#endregion
export { hostOf, mintAppSession, requireAccount, withCode };
