// @anna-ai/cli/dist/dev-app-cache-CP2P8SCZ.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { dirname, join, resolve } from "node:path";
import { existsSync, mkdirSync, readFileSync, unlinkSync, writeFileSync } from "node:fs";

//#region src/dev-app-cache.ts
const CACHE_DIR = ".anna";
const CACHE_FILE = "dev-app.json";
function cachePath(cwd) {
	return resolve(cwd, CACHE_DIR, CACHE_FILE);
}
function readDevAppCache(cwd) {
	const p = cachePath(cwd);
	if (!existsSync(p)) return null;
	try {
		const raw = JSON.parse(readFileSync(p, "utf-8"));
		if (typeof raw.host === "string" && typeof raw.slug === "string" && typeof raw.app_id === "number") return {
			host: raw.host,
			slug: raw.slug,
			app_id: raw.app_id,
			name: raw.name ?? raw.slug,
			registered_at: raw.registered_at ?? ""
		};
	} catch {}
	return null;
}
function writeDevAppCache(cwd, entry) {
	const p = cachePath(cwd);
	mkdirSync(dirname(p), { recursive: true });
	writeFileSync(p, JSON.stringify(entry, null, 2) + "\n", "utf-8");
}
/** Delete the local dev-app cache file.
*
*  Used by the LLM bridge's recovery flow: when `/dev/session/mint`
*  returns 404 with "app slug ... not found", the cache is stale (the
*  nexus DB no longer has the row this cache asserts is registered).
*  Invalidating forces the next `ensureDevAppRegistered` call to
*  re-POST `/dev/apps/register`. Idempotent: a missing file is a no-op.
*/
function invalidateDevAppCache(cwd) {
	const p = cachePath(cwd);
	if (existsSync(p)) try {
		unlinkSync(p);
	} catch {}
}
/** Call POST /api/v1/anna-apps/dev/apps/register. Idempotent server-side. */
async function registerDevApp(args) {
	const url = `${canonicalHost(args.host)}/api/v1/anna-apps/dev/apps/register`;
	const res = await fetch(url, {
		method: "POST",
		headers: { "content-type": "application/json" },
		body: JSON.stringify({
			pat: args.pat,
			slug: args.input.slug,
			name: args.input.name,
			category: args.input.category,
			tagline: args.input.tagline
		})
	});
	if (!res.ok) {
		const text = await res.text().catch(() => "");
		if (res.status === 409 && text.includes("HANDLE_REQUIRED")) throw new Error("developer handle required before registering a dev app — run `anna-app account set-handle <handle>` first.");
		throw new Error(`/dev/apps/register failed: HTTP ${res.status} ${text}`);
	}
	return await res.json();
}
/** Call GET /api/v1/anna-apps/dev/apps. */
async function listDevApps(args) {
	const url = new URL(`${canonicalHost(args.host)}/api/v1/anna-apps/dev/apps`);
	url.searchParams.set("pat", args.pat);
	const res = await fetch(url, { method: "GET" });
	if (!res.ok) {
		const text = await res.text().catch(() => "");
		throw new Error(`/dev/apps failed: HTTP ${res.status} ${text}`);
	}
	const body = await res.json();
	return body.apps;
}
/**
* Convenience helper for `anna-app dev`: returns a valid cached entry if
* it still matches the manifest slug, otherwise hits the server to
* (re-)register. Throws if no PAT is available on disk.
*/
async function ensureDevAppRegistered(args) {
	const canonical = canonicalHost(args.host);
	const cached = readDevAppCache(args.cwd);
	if (cached && cached.host === canonical && cached.slug === args.input.slug) try {
		const apps = await listDevApps({
			host: args.host,
			pat: args.pat
		});
		if (apps.some((a) => a.slug === args.input.slug)) return cached;
		invalidateDevAppCache(args.cwd);
	} catch {
		return cached;
	}
	const r = await registerDevApp({
		host: args.host,
		pat: args.pat,
		input: args.input
	});
	const entry = {
		host: canonical,
		slug: r.slug,
		app_id: r.app_id,
		name: r.name,
		registered_at: new Date().toISOString()
	};
	writeDevAppCache(args.cwd, entry);
	return entry;
}
const _internal = {
	cachePath,
	CACHE_DIR: join(CACHE_DIR, CACHE_FILE)
};
const EXECUTA_CACHE_FILE = "dev-executa.json";
/**
* Bump this whenever the server-side dev-executa grant shape changes in
* a way that requires a re-register to pick up. Cached entries written
* with a lower `cache_version` are treated as stale and re-POSTed.
*
*   1 — initial (llm_grant only)
*   2 — adds image_grant + upload_grant; fixes `allowedPurposes` to
*       protocol whitelist {image_input,image_reference,user_artifact}
*/
const EXECUTA_CACHE_VERSION = 2;
function executaCachePath(cwd) {
	return resolve(cwd, CACHE_DIR, EXECUTA_CACHE_FILE);
}
function readDevExecutaCache(cwd) {
	const p = executaCachePath(cwd);
	if (!existsSync(p)) return null;
	try {
		const raw = JSON.parse(readFileSync(p, "utf-8"));
		if (typeof raw.host === "string" && typeof raw.slug === "string" && typeof raw.app_id === "number" && typeof raw.tool_id === "string") return {
			host: raw.host,
			slug: raw.slug,
			app_id: raw.app_id,
			tool_id: raw.tool_id,
			name: raw.name ?? raw.tool_id,
			registered_at: raw.registered_at ?? "",
			cache_version: typeof raw.cache_version === "number" ? raw.cache_version : 0
		};
	} catch {}
	return null;
}
function writeDevExecutaCache(cwd, entry) {
	const p = executaCachePath(cwd);
	mkdirSync(dirname(p), { recursive: true });
	writeFileSync(p, JSON.stringify(entry, null, 2) + "\n", "utf-8");
}
/** Call `POST /api/v1/anna-apps/dev/executas/register`. Idempotent. */
async function registerDevExecuta(args) {
	const url = `${canonicalHost(args.host)}/api/v1/anna-apps/dev/executas/register`;
	const res = await fetch(url, {
		method: "POST",
		headers: { "content-type": "application/json" },
		body: JSON.stringify({
			pat: args.pat,
			tool_id: args.toolId,
			slug: args.slug,
			name: args.name
		})
	});
	if (!res.ok) {
		const text = await res.text().catch(() => "");
		throw new Error(`/dev/executas/register failed: HTTP ${res.status} ${text}`);
	}
	return await res.json();
}
/**
* Convenience: return a cached executa registration if still valid for
* this host+tool_id, otherwise hit the server. Throws if no PAT.
*/
async function ensureDevExecutaRegistered(args) {
	const canonical = canonicalHost(args.host);
	const cached = readDevExecutaCache(args.cwd);
	if (cached && cached.host === canonical && cached.tool_id === args.toolId && cached.cache_version >= EXECUTA_CACHE_VERSION && (!args.slug || cached.slug === args.slug)) return cached;
	const r = await registerDevExecuta({
		host: args.host,
		pat: args.pat,
		toolId: args.toolId,
		slug: args.slug,
		name: args.name
	});
	const entry = {
		host: canonical,
		slug: r.slug,
		app_id: r.app_id,
		tool_id: r.tool_id,
		name: r.name,
		registered_at: new Date().toISOString(),
		cache_version: EXECUTA_CACHE_VERSION
	};
	writeDevExecutaCache(args.cwd, entry);
	return entry;
}

//#endregion
export { ensureDevAppRegistered, ensureDevExecutaRegistered, invalidateDevAppCache };
