// @anna-ai/cli/dist/executa-destructive-8iSnfmcn.js
import "./credentials-BTv2IfUZ.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { deleteMyTool, listMyTools, listToolVersions, unpublishTool, yankToolVersion } from "./executas-BhFed1B-.js";
import { ensureDestructiveAllowed } from "./confirm-BuT56E_B.js";
import { dim, green, yellow } from "kleur/colors";

//#region src/commands/executa-destructive.ts
async function findBySlug(client, ref) {
	const rows = await listMyTools(client);
	const exact = rows.find((r) => r.tool_id === ref);
	if (exact) return exact;
	const matches = rows.filter((r) => r.tool_id.includes(ref));
	if (matches.length === 0) return null;
	if (matches.length === 1) return matches[0] ?? null;
	const list = matches.map((r) => `  - ${r.tool_id}`).join("\n");
	throw new CliError(`ref "${ref}" is ambiguous (${matches.length} matches):\n${list}\nPlease pass the full tool_id.`, 5);
}
/**
* Confirm-token UX: accept the slug the user typed OR the real server
* tool_id. Validating against tool_id is the safer match because the
* user has just been told what it is by the resolver.
*/
async function checkConfirm(verb, refTyped, toolId, opts) {
	await ensureDestructiveAllowed({
		verb,
		slug: toolId,
		yes: opts.yes,
		confirm: opts.confirm === refTyped ? toolId : opts.confirm
	});
}
async function runExecutaUnpublish(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const ex = await findBySlug(client, opts.slug);
		if (!ex) throw new CliError(`no executa matching "${opts.slug}" on ${host}`, 5);
		await checkConfirm("executa unpublish", opts.slug, ex.tool_id, opts);
		const updated = await unpublishTool(client, ex.id);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				executa_id: ex.id,
				tool_id: ex.tool_id,
				visibility: updated.visibility ?? null
			}, null, 2));
			return 0;
		}
		console.log(green(`✓ executas/${ex.tool_id}: unpublished (visibility=${updated.visibility})`));
		return 0;
	});
}
async function runExecutaYank(opts) {
	return withErrorHandling(async () => {
		const at = opts.ref.lastIndexOf("@");
		if (at < 1) throw new CliError(`\`executa yank\` expects <slug>@<version> (got "${opts.ref}")`, 4);
		const slug = opts.ref.slice(0, at);
		const version = opts.ref.slice(at + 1);
		const { client, host } = resolveClient({ account: opts.account });
		const ex = await findBySlug(client, slug);
		if (!ex) throw new CliError(`no executa matching "${slug}" on ${host}`, 5);
		await checkConfirm("executa yank", slug, ex.tool_id, opts);
		const versions = await listToolVersions(client, ex.id);
		const ver = versions.find((v) => v.version === version);
		if (!ver) throw new CliError(`no version "${version}" for executa "${slug}" (have: ${versions.map((v) => v.version).join(", ") || "(none)"})`, 5);
		await yankToolVersion(client, ex.id, ver.id);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				executa_id: ex.id,
				tool_id: ex.tool_id,
				version,
				version_id: ver.id,
				yanked: true
			}, null, 2));
			return 0;
		}
		console.log(yellow(`✓ executas/${ex.tool_id}@${version}: yanked (version_id=${ver.id})`));
		return 0;
	});
}
async function runExecutaDelete(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const ex = await findBySlug(client, opts.slug);
		if (!ex) throw new CliError(`no executa matching "${opts.slug}" on ${host}`, 5);
		await checkConfirm("executa delete", opts.slug, ex.tool_id, opts);
		await deleteMyTool(client, ex.id);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				executa_id: ex.id,
				tool_id: ex.tool_id,
				deleted: true
			}, null, 2));
			return 0;
		}
		console.log(yellow(`✓ executas/${ex.tool_id}: deleted`));
		console.log(dim(`  (local .anna/executa.json is now stale; remove it manually if you no longer need it)`));
		return 0;
	});
}

//#endregion
export { runExecutaDelete, runExecutaUnpublish, runExecutaYank };
