// @anna-ai/cli/dist/executas-BhFed1B-.js
//#region src/api/executas.ts
async function createDraft(client, body) {
	const r = await client.request({
		method: "POST",
		path: "/api/v1/executas/my/drafts",
		body
	});
	const committed = r.idempotentHit && r.data.id < 0 && r.data.payload && typeof r.data.payload === "object" && typeof r.data.payload.committed_executa_id === "number" ? r.data.payload.committed_executa_id : null;
	return {
		draft: r.data,
		idempotentHit: r.idempotentHit,
		committedExecutaId: committed
	};
}
async function commitDraft(client, draftId, finalPayload) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/executas/my/drafts/${draftId}/commit`,
		body: finalPayload ? { payload: finalPayload } : void 0
	});
	return r.data;
}
async function listMyTools(client) {
	const r = await client.request({ path: "/api/v1/executas/my/tools" });
	return Array.isArray(r.data) ? r.data : [];
}
async function getMyTool(client, executaId) {
	const r = await client.request({ path: `/api/v1/executas/my/tools/${executaId}/detail` });
	return r.data;
}
async function updateMyTool(client, executaId, body) {
	const r = await client.request({
		method: "PUT",
		path: `/api/v1/executas/my/tools/${executaId}`,
		body
	});
	return r.data;
}
async function listToolVersions(client, executaId) {
	const r = await client.request({
		path: `/api/v1/executas/my/tools/${executaId}/versions`,
		allowStatuses: [404]
	});
	if (r.status === 404) return [];
	return Array.isArray(r.data) ? r.data : [];
}
async function publishToolVersion(client, executaId, body, contentHash) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/executas/my/tools/${executaId}/versions`,
		query: contentHash ? { content_hash: contentHash } : void 0,
		body: body ?? void 0
	});
	return {
		executa: r.data,
		idempotentHit: r.idempotentHit
	};
}
async function setToolVisibility(client, executaId, visibility) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/executas/my/tools/${executaId}/visibility`,
		body: { visibility }
	});
	return r.data;
}
async function unpublishTool(client, executaId) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/executas/my/tools/${executaId}/unpublish`
	});
	return r.data;
}
async function getMySkill(client, executaId) {
	const r = await client.request({ path: `/api/v1/executas/my/skills/${executaId}/detail` });
	return r.data;
}
async function updateMySkill(client, executaId, body) {
	const r = await client.request({
		method: "PUT",
		path: `/api/v1/executas/my/skills/${executaId}`,
		body
	});
	return r.data;
}
async function setSkillVisibility(client, executaId, visibility) {
	const r = await client.request({
		method: "POST",
		path: `/api/v1/executas/my/skills/${executaId}/visibility`,
		body: { visibility }
	});
	return r.data;
}
async function deleteMyTool(client, executaId) {
	await client.request({
		method: "DELETE",
		path: `/api/v1/executas/my/tools/${executaId}`,
		allowStatuses: [200, 204]
	});
}
async function yankToolVersion(client, executaId, versionId) {
	await client.request({
		method: "DELETE",
		path: `/api/v1/executas/my/tools/${executaId}/versions/${versionId}`
	});
}

//#endregion
export { commitDraft, createDraft, deleteMyTool, getMySkill, getMyTool, listMyTools, listToolVersions, publishToolVersion, setSkillVisibility, setToolVisibility, unpublishTool, updateMySkill, updateMyTool, yankToolVersion };
