// @anna-ai/cli/dist/executa-install-A-BgavhA.js
import { readExecutaIdentity } from "./executa-cache-Do1D7xC0.js";
import { parseExecutaSpec } from "./dev-aWkfWGLD.js";
import { isAbsolute, join, resolve } from "node:path";
import { chmodSync, existsSync, mkdirSync, writeFileSync } from "node:fs";
import { bold, cyan, dim, green, red, yellow } from "kleur/colors";
import { homedir } from "node:os";

//#region src/commands/executa-install.ts
/** POSIX single-quote one argument for safe embedding in the shim. */
function shQuote(arg) {
	return `'${arg.replace(/'/g, "'\\''")}'`;
}
function buildShim(projectDir, command) {
	const cd = `cd ${shQuote(projectDir)} || exit 1`;
	const exec = `exec ${command.map(shQuote).join(" ")} "$@"`;
	return [
		"#!/bin/sh",
		"# Auto-generated by `anna-app executa install`.",
		"# Local-dev launcher shim — runs the Executa from source so the Agent",
		"# can discover it via 'Rediscover Local'. Safe to delete; regenerate",
		"# with `anna-app executa install`.",
		cd,
		exec,
		""
	].join("\n");
}
async function runExecutaInstall(opts) {
	const json = opts.json ?? false;
	const cwd = opts.dir ? isAbsolute(opts.dir) ? opts.dir : resolve(process.cwd(), opts.dir) : process.cwd();
	if (!existsSync(cwd)) {
		console.error(red(`✗ executa dir not found: ${cwd}`));
		return 2;
	}
	const parsed = parseExecutaSpec(`dir=${cwd}`, process.cwd());
	if (parsed instanceof Error) {
		console.error(red(`✗ ${parsed.message}`));
		return 2;
	}
	if (!parsed.command || parsed.command.length === 0) {
		console.error(red(`✗ could not derive a launch command for ${cwd}`));
		return 2;
	}
	const identity = readExecutaIdentity(cwd);
	const targetId = opts.toolId ?? identity?.tool_id ?? parsed.tool_id;
	if (!targetId) {
		console.error(red("✗ could not resolve a tool_id — pass --tool-id <id>, or run `anna-app executa publish` / `anna-app apps push` first to mint one."));
		return 2;
	}
	if (/[/\\]/.test(targetId) || targetId.startsWith(".")) {
		console.error(red(`✗ invalid tool_id "${targetId}" (no path separators or leading dot)`));
		return 2;
	}
	const binDir = opts.binDir ? isAbsolute(opts.binDir) ? opts.binDir : resolve(process.cwd(), opts.binDir) : join(homedir(), ".anna", "executa", "bin");
	const shimPath = join(binDir, targetId);
	if (existsSync(shimPath) && !opts.force) {
		console.error(red(`✗ shim already exists: ${shimPath}`) + dim(" (pass --force to overwrite)"));
		return 1;
	}
	try {
		mkdirSync(binDir, { recursive: true });
		writeFileSync(shimPath, buildShim(parsed.project_dir, parsed.command), "utf-8");
		chmodSync(shimPath, 493);
	} catch (e) {
		console.error(red(`✗ failed to write shim: ${e.message}`));
		return 1;
	}
	if (json) {
		console.log(JSON.stringify({
			tool_id: targetId,
			shim_path: shimPath,
			project_dir: parsed.project_dir,
			command: parsed.command,
			source: opts.toolId ? "flag" : identity?.tool_id ? "identity-cache" : "discovered"
		}));
		return 0;
	}
	if (opts.quiet) return 0;
	console.log(bold(cyan("anna-app executa install")));
	console.log(`  tool_id   ${green(targetId)}`);
	console.log(`  shim      ${dim(shimPath)}`);
	console.log(`  dir       ${dim(parsed.project_dir)}`);
	console.log(`  command   ${dim(parsed.command.join(" "))}`);
	if (!opts.toolId && identity?.tool_id) console.log(`  source    ${dim("minted tool_id from .anna/executa.json")}`);
	else if (!opts.toolId) console.log(yellow("  ! using the placeholder tool_id from executa.json — run `executa publish`/`apps push` to mint the real one, then re-run with --force (or pass --tool-id)."));
	console.log(dim("\n→ open the Agent's Plugin Details modal and click " + bold("Rediscover Local") + dim(" to load it under this tool_id.")));
	return 0;
}

//#endregion
export { runExecutaInstall };
