// @anna-ai/cli/dist/executa-publish-Xmm4DMO9.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { commitDraft, createDraft, getMySkill, getMyTool, listToolVersions, publishToolVersion, setSkillVisibility, setToolVisibility, updateMySkill, updateMyTool } from "./executas-BhFed1B-.js";
import { executaCacheMatches, invalidateExecutaCache, mintIdempotencyKey, readExecutaIdentity, writeExecutaIdentity } from "./executa-cache-Do1D7xC0.js";
import { loadExecutaManifest } from "./manifest-CMc7RM57.js";
import { join, relative, resolve, sep } from "node:path";
import { readFileSync, readdirSync, statSync, writeFileSync } from "node:fs";
import { bold, cyan, dim, green, yellow } from "kleur/colors";
import { createHash } from "node:crypto";

//#region src/publish/bundle.ts
function manifestHash(manifest) {
	const canon = canonicalize(manifest);
	return createHash("sha256").update(canon).digest("hex");
}
function bundleHash(manifest, bundleDir) {
	const h = createHash("sha256");
	h.update(canonicalize(manifest));
	if (bundleDir) {
		const files = walk(bundleDir);
		files.sort();
		for (const rel of files) {
			const abs = join(bundleDir, rel);
			const data = readFileSync(abs);
			h.update("\n");
			h.update(rel.split(sep).join("/"));
			h.update("\0");
			h.update(createHash("sha256").update(data).digest("hex"));
		}
	}
	return h.digest("hex");
}
/** RFC-8785-ish JSON canonicalization, sufficient for our hash use. */
function canonicalize(value) {
	if (value === null || typeof value !== "object") return JSON.stringify(value);
	if (Array.isArray(value)) return "[" + value.map(canonicalize).join(",") + "]";
	const obj = value;
	const keys = Object.keys(obj).sort();
	return "{" + keys.map((k) => JSON.stringify(k) + ":" + canonicalize(obj[k])).join(",") + "}";
}
function walk(root, base = root) {
	const out = [];
	for (const ent of readdirSync(base, { withFileTypes: true })) {
		const abs = join(base, ent.name);
		if (ent.isDirectory()) out.push(...walk(root, abs));
		else if (ent.isFile()) out.push(relative(root, abs));
	}
	return out;
}

//#endregion
//#region src/publish/bump.ts
function bumpVersion(current, kind) {
	const m = current.match(/^(\d+)\.(\d+)\.(\d+)(?:[-+][0-9A-Za-z.-]+)?$/);
	if (!m || !m[1] || !m[2] || !m[3]) throw new CliError(`cannot bump non-semver version "${current}"`, 4);
	let [, maj, min, pat] = m;
	if (kind === "major") {
		maj = String(Number(maj) + 1);
		min = "0";
		pat = "0";
	} else if (kind === "minor") {
		min = String(Number(min) + 1);
		pat = "0";
	} else pat = String(Number(pat) + 1);
	return `${maj}.${min}.${pat}`;
}
/**
* Rewrite a manifest's `version` field in place.
*
* Preserves on-disk formatting as much as possible (re-serialising
* with 2-space JSON.stringify) — `manifest.json` and `executa.json`
* are both expected to be machine-managed, so cosmetic formatting
* differences are acceptable.
*/
function rewriteVersion(manifestPath, manifest, newVersion) {
	const next = {
		...manifest,
		version: newVersion
	};
	writeFileSync(manifestPath, JSON.stringify(next, null, 2) + "\n", "utf-8");
}

//#endregion
//#region src/commands/executa-publish.ts
/**
* Flatten the parsed `executa.json` `distribution` block into the snake_case
* field set the server expects on both the draft-commit payload and the
* `PUT /my/tools/{id}` metadata sync. Returns an empty object when the
* Executa declares no distribution (dev-only) so the spread is a no-op.
*
* Note: per-asset `sha256` is folded into each `binary_urls` asset dict
* (`{ url, sha256, … }`) rather than sent as a separate top-level
* `binary_sha256` map — the `PUT /my/tools/{id}` schema
* (`UserToolUpdateRequest`) has no `binary_sha256` field, whereas
* `binary_urls` accepts the asset-dict form on both the create and update
* paths.
*/
function distributionPayload(d) {
	if (!d) return {};
	const out = { distribution_type: d.type };
	if (d.packageName !== void 0) out.package_name = d.packageName;
	if (d.executableName !== void 0) out.executable_name = d.executableName;
	if (d.supportsProtocol !== void 0) out.supports_protocol = d.supportsProtocol;
	if (d.binaryUrls !== void 0) out.binary_urls = d.binaryUrls;
	if (d.capabilities !== void 0) out.capabilities = d.capabilities;
	return out;
}
async function runExecutaPublish(opts) {
	return withErrorHandling(async () => {
		const cwd = resolve(opts.cwd ?? process.cwd());
		const manifest = loadExecutaManifest(cwd, opts.manifest);
		const isSkill = manifest.executa_type === "skill";
		const { client, host, source } = resolveClient({ account: opts.account });
		if (!opts.json && !opts.quiet) console.log(dim(`using PAT from ${source} (host=${host})`));
		if (opts.bump) {
			const next = bumpVersion(manifest.version, opts.bump);
			if (!opts.json && !opts.quiet) console.log(yellow(`↑ bumping executa.json version: ${manifest.version} → ${next}`));
			if (!opts.noWrite && !opts.dryRun) rewriteVersion(manifest.path, manifest.raw, next);
			manifest.version = next;
			manifest.raw.version = next;
		}
		let cached = readExecutaIdentity(cwd);
		if (cached && cached.host !== canonicalHost(host)) cached = null;
		let executaId;
		let toolId;
		let serverSlug;
		let firstPublish = false;
		let cacheValid = false;
		if (cached && executaCacheMatches(cached, host, manifest.slug)) try {
			if (isSkill) await getMySkill(client, cached.executa_id);
			else await getMyTool(client, cached.executa_id);
			cacheValid = true;
		} catch (e) {
			if (e instanceof CliError && e.status === 404) {
				if (!opts.json && !opts.quiet) console.log(yellow(`! cached executa_id=${cached.executa_id} no longer exists on ${host}; clearing .anna/executa.json and re-minting via draft+commit`));
				invalidateExecutaCache(cwd);
				cached = null;
			} else throw e;
		}
		if (cacheValid && cached) {
			executaId = cached.executa_id;
			toolId = cached.tool_id;
			serverSlug = cached.slug;
		} else {
			if (opts.dryRun) {
				if (!opts.json) console.log(yellow(`[dry-run] would POST /my/drafts (slug=${manifest.slug}, type=${manifest.executa_type}) and commit`));
				return 0;
			}
			const idemKey = mintIdempotencyKey(host, manifest.slug, manifest.executa_type);
			const draftRes = await createDraft(client, {
				name: manifest.name,
				executa_type: manifest.executa_type,
				slug_override: manifest.slug,
				idempotency_key: idemKey
			});
			if (draftRes.committedExecutaId !== null) {
				executaId = draftRes.committedExecutaId;
				toolId = draftRes.draft.tool_id;
				serverSlug = manifest.slug;
				if (!opts.json && !opts.quiet) console.log(dim(`(server idempotent hit on already-committed Executa id=${executaId})`));
			} else {
				const draft = draftRes.draft;
				const finalPayload = {
					...typeof draft.payload === "object" && draft.payload !== null ? draft.payload : {},
					name: manifest.name,
					description: manifest.description,
					executa_type: manifest.executa_type,
					slug: manifest.slug,
					version: manifest.version,
					...distributionPayload(manifest.distribution)
				};
				if (manifest.executa_type === "skill") {
					finalPayload.skill_content = manifest.skillContent;
					if (manifest.executionMode) finalPayload.execution_mode = manifest.executionMode;
				}
				const committed = draftRes.idempotentHit ? await commitDraft(client, draft.id, finalPayload) : await commitDraft(client, draft.id, finalPayload);
				executaId = committed.executa_id;
				toolId = committed.tool_id;
				serverSlug = manifest.slug;
				firstPublish = true;
			}
			writeExecutaIdentity(cwd, {
				$schema: "anna-executa-identity/v1",
				host: canonicalHost(host),
				executa_id: executaId,
				tool_id: toolId,
				slug: serverSlug,
				executa_type: manifest.executa_type,
				first_published_at: cached?.first_published_at ?? new Date().toISOString()
			});
		}
		if (opts.dryRun) {
			if (!opts.json) console.log(yellow(`[dry-run] resolved executa_id=${executaId} tool_id=${toolId}; would PUT metadata + POST version ${manifest.version}`));
			return 0;
		}
		if (isSkill) await updateMySkill(client, executaId, {
			name: manifest.name,
			description: manifest.description,
			skill_content: manifest.skillContent,
			version: manifest.version,
			...manifest.executionMode ? { skill_execution_mode: manifest.executionMode } : {}
		});
		else await updateMyTool(client, executaId, {
			name: manifest.name,
			description: manifest.description,
			...distributionPayload(manifest.distribution)
		});
		const preFreeze = isSkill ? await getMySkill(client, executaId) : await getMyTool(client, executaId);
		const currentVis = preFreeze.visibility ?? "private";
		const targetVis = opts.publish ? "public" : currentVis === "public" ? "public" : "app_bundled";
		const visibilityChanged = currentVis !== targetVis;
		if (visibilityChanged) if (isSkill) await setSkillVisibility(client, executaId, targetVis);
		else await setToolVisibility(client, executaId, targetVis);
		const hash = manifestHash(manifest.raw);
		const freeze = opts.freeze !== false;
		let frozenIdempotentHit;
		if (!freeze) frozenIdempotentHit = true;
		else if (isSkill) frozenIdempotentHit = !visibilityChanged;
		else {
			const verRes = await publishToolVersion(client, executaId, { version: manifest.version }, hash);
			frozenIdempotentHit = verRes.idempotentHit;
		}
		let publishedToPublic = false;
		if (opts.publish && targetVis === "public") publishedToPublic = true;
		const finalExecuta = isSkill ? await getMySkill(client, executaId) : await getMyTool(client, executaId);
		const versions = await listToolVersions(client, executaId);
		const latest = versions[0] ?? null;
		const result = {
			host,
			executa_id: executaId,
			tool_id: toolId,
			slug: serverSlug,
			version: manifest.version,
			content_hash: hash,
			idempotent_hit: frozenIdempotentHit,
			first_publish: firstPublish,
			published_public: publishedToPublic,
			visibility: finalExecuta.visibility ?? null,
			latest_version_id: latest?.id ?? null
		};
		opts.capture?.(result);
		if (opts.quiet) return 0;
		if (opts.json) {
			console.log(JSON.stringify(result, null, 2));
			return 0;
		}
		if (frozenIdempotentHit) console.log(green(`✓ executas/${serverSlug}: version ${manifest.version} already frozen (content hash unchanged)`));
		else console.log(green(`✓ executas/${serverSlug}: version ${manifest.version} frozen`));
		if (firstPublish) console.log(dim(`  first publish — wrote .anna/executa.json (tool_id=${toolId})`));
		if (publishedToPublic) console.log(green(`  visibility set to PUBLIC`));
		console.log(bold(cyan(`tool_id: ${toolId}`)));
		return 0;
	});
}

//#endregion
export { bumpVersion, bundleHash, canonicalize, manifestHash, rewriteVersion, runExecutaPublish };
