// @anna-ai/cli/dist/executa-register-CMmlKyKB.js
import { getAccount } from "./credentials-BTv2IfUZ.js";
import { bold, cyan, dim, green, red } from "kleur/colors";

//#region src/commands/executa-register.ts
async function runExecutaRegister(opts) {
	const acc = getAccount(opts.account);
	if (!acc) {
		console.error(red("✗ no PAT on disk — run `anna-app login --host <nexus-url>` first."));
		return 2;
	}
	const body = {
		pat: acc.pat,
		tool_id: opts.toolId,
		slug: opts.slug,
		name: opts.name
	};
	const url = `${acc.host.replace(/\/$/, "")}/api/v1/anna-apps/dev/executas/register`;
	let res;
	try {
		res = await fetch(url, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify(body)
		});
	} catch (e) {
		console.error(red(`✗ network error: ${e.message}`));
		return 2;
	}
	if (res.status === 404) {
		console.error(red("✗ your nexus does not expose POST /api/v1/anna-apps/dev/executas/register — upgrade matrix-nexus to a version that ships this endpoint."));
		return 2;
	}
	if (!res.ok) {
		const text = await res.text().catch(() => "");
		console.error(red(`✗ HTTP ${res.status}: ${text}`));
		return 1;
	}
	const out = await res.json();
	console.log(`${green(out.created ? "✓ registered" : "✓ exists")} ${bold(out.slug)} ${dim(`(app_id=${out.app_id}, kind=${out.kind})`)}`);
	console.log(`  tool_id   ${dim(out.tool_id)}`);
	console.log(`  name      ${dim(out.name)}`);
	console.log(`  host      ${cyan(acc.host)}`);
	return 0;
}

//#endregion
export { runExecutaRegister };
