// @anna-ai/cli/dist/executa-dev-0XdBPTDy.js
import "./nexus-root-BlPwOusj.js";
import { parseExecutaSpec } from "./dev-aWkfWGLD.js";
import { isAbsolute, resolve } from "node:path";
import { existsSync } from "node:fs";
import { bold, cyan, dim, green, red, yellow } from "kleur/colors";
import { createInterface } from "node:readline/promises";
import * as readline from "node:readline";

//#region src/commands/executa-dev.ts
/**
* Mutable hook the REPL installs while it's waiting at a prompt, so
* out-of-band stderr lines from the executa subprocess don't garble
* the prompt. When set, the stderr sink clears the current readline
* line on stdout, prints the message, then asks readline to redraw
* the prompt + any pending user input.
*/
let activeRepl = null;
async function runExecutaDev(opts) {
	const cwd = process.cwd();
	const dir = opts.dir ? isAbsolute(opts.dir) ? opts.dir : resolve(cwd, opts.dir) : cwd;
	if (!existsSync(dir)) {
		console.error(red(`✗ dir not found: ${dir}`));
		return 2;
	}
	const specStr = opts.spec ? prependDirIfMissing(opts.spec, dir) : `dir=${dir}`;
	const parsed = parseExecutaSpec(specStr, cwd);
	if (parsed instanceof Error) {
		console.error(red(`✗ ${parsed.message}`));
		return 2;
	}
	if (!parsed.command || parsed.command.length === 0) {
		console.error(red(`✗ could not derive a launch command for ${dir}`));
		return 2;
	}
	const oneShot = !!(opts.describe || opts.health || opts.invoke);
	const quiet = oneShot && (opts.json ?? false);
	const { getAccount } = await import("./credentials-Chkoidh5.js");
	const { ensureDevExecutaRegistered } = await import("./dev-app-cache-CP2P8SCZ.js");
	const needsRealMint = !opts.noSampling && !opts.mockSampling || !opts.noAgent && !opts.mockAgent || !opts.noImage && !opts.mockImage || !opts.noUpload && !opts.mockUpload || opts.storage === "real";
	let effectiveAppSlug = opts.appSlug;
	let autoRegistered = false;
	if (!effectiveAppSlug && needsRealMint) {
		const acc = getAccount(opts.samplingAccount);
		if (acc) try {
			const reg = await ensureDevExecutaRegistered({
				cwd: parsed.project_dir,
				host: acc.host,
				pat: acc.pat,
				toolId: parsed.tool_id
			});
			effectiveAppSlug = reg.slug;
			autoRegistered = true;
		} catch (e) {
			if (!quiet) {
				console.warn(yellow(`! executa auto-register failed: ${e.message}`));
				console.warn(yellow("  reverse-RPC bridges (sampling/agent/image/upload) will run in disabled mode. Pass `--app-slug <slug>` to override, or `--no-sampling --no-agent --no-image --no-upload` to silence."));
			}
		}
	}
	const { SamplingBridge } = await import("./sampling-Cc0qaj8a.js");
	const simulateUnsupportedResponseFormat = opts.samplingUnsupportedFormat;
	const sampling = opts.noSampling ? new SamplingBridge({ mode: "off" }) : opts.mockSampling ? new SamplingBridge({
		mode: "mock",
		mockFile: opts.mockSampling,
		simulateUnsupportedResponseFormat
	}) : effectiveAppSlug ? new SamplingBridge({
		mode: "real",
		account: opts.samplingAccount,
		appSlug: effectiveAppSlug,
		simulateUnsupportedResponseFormat
	}) : new SamplingBridge({ mode: "off" });
	const { AgentBridge } = await import("./agent-J2bELWfv.js");
	const agent = opts.noAgent ? new AgentBridge({ mode: "off" }) : opts.mockAgent ? new AgentBridge({
		mode: "mock",
		mockFile: opts.mockAgent
	}) : effectiveAppSlug ? new AgentBridge({
		mode: "real",
		account: opts.agentAccount ?? opts.samplingAccount,
		appSlug: effectiveAppSlug
	}) : new AgentBridge({ mode: "off" });
	const { StorageBridge } = await import("./storage-C95wApr2.js");
	const storageMode = opts.storage ?? (opts.mockStorage ? "mock" : "memory");
	const storage = new StorageBridge({
		mode: storageMode,
		mockFile: opts.mockStorage,
		account: opts.storageAccount ?? opts.samplingAccount,
		appSlug: effectiveAppSlug,
		scopes: opts.storageScopes ? opts.storageScopes.split(",").map((s) => s.trim()).filter(Boolean) : void 0,
		pluginName: parsed.tool_id
	});
	const { ExecutaRunner } = await import("./runner-DVu-o_U8.js");
	const { ImageBridge } = await import("./image-Bdi4Hu0m.js");
	const image = opts.noImage ? new ImageBridge({ mode: "off" }) : opts.mockImage ? new ImageBridge({
		mode: "mock",
		mockFile: opts.mockImage
	}) : effectiveAppSlug ? new ImageBridge({
		mode: "real",
		account: opts.imageAccount ?? opts.samplingAccount,
		appSlug: effectiveAppSlug
	}) : new ImageBridge({ mode: "off" });
	const { HostUploadBridge } = await import("./host_upload-sfo4UHlR.js");
	const hostUpload = opts.noUpload ? new HostUploadBridge({ mode: "off" }) : opts.mockUpload ? new HostUploadBridge({
		mode: "mock",
		mockFile: opts.mockUpload
	}) : effectiveAppSlug ? new HostUploadBridge({
		mode: "real",
		account: opts.uploadAccount ?? opts.samplingAccount,
		appSlug: effectiveAppSlug
	}) : new HostUploadBridge({ mode: "off" });
	const runner = new ExecutaRunner({
		command: parsed.command,
		cwd: parsed.project_dir,
		sampling,
		agent,
		storage,
		image,
		hostUpload,
		onStderr: (line) => {
			const text = quiet ? `${line}\n` : dim(`[executa] ${line}\n`);
			writeStderrCooperative(text);
		}
	});
	if (!quiet) {
		const slugTag = (mode) => effectiveAppSlug ? `${mode} → app_slug=${effectiveAppSlug}${autoRegistered ? " (auto)" : ""}` : "disabled (no fixture, no --app-slug, no PAT)";
		console.log(bold(cyan("anna-app executa dev")));
		console.log(`  tool_id   ${dim(parsed.tool_id)}`);
		console.log(`  dir       ${dim(parsed.project_dir)}`);
		console.log(`  command   ${dim(parsed.command.join(" "))}`);
		if (autoRegistered && effectiveAppSlug) console.log(`  executa   ${dim(`auto-registered as ${effectiveAppSlug}`)}`);
		console.log(`  sampling  ${dim(opts.noSampling ? "disabled (--no-sampling)" : opts.mockSampling ? `mock (${opts.mockSampling})` : slugTag("real"))}`);
		console.log(`  agent     ${dim(opts.noAgent ? "disabled (--no-agent)" : opts.mockAgent ? `mock (${opts.mockAgent})` : slugTag("real"))}`);
		console.log(`  storage   ${dim(storageMode === "mock" ? `mock (${opts.mockStorage ?? "<no-fixture>"})` : storageMode === "real" ? `real → app_slug=${effectiveAppSlug ?? "<unset>"}` : storageMode)}`);
		console.log(`  image     ${dim(opts.noImage ? "disabled (--no-image)" : opts.mockImage ? `mock (${opts.mockImage})` : slugTag("real"))}`);
		console.log(`  upload    ${dim(opts.noUpload ? "disabled (--no-upload)" : opts.mockUpload ? `mock (${opts.mockUpload})` : slugTag("real"))}`);
	}
	let init;
	try {
		init = await runner.start();
	} catch (e) {
		console.error(red(`✗ executa failed to start: ${e.message}`));
		return 2;
	}
	if (!quiet) console.log(`  ${green("✓")} negotiated protocol ${bold(init.protocolVersion)}`);
	let exitCode = 0;
	try {
		if (oneShot) exitCode = await runOneShot(runner, opts, quiet);
		else exitCode = await runRepl(runner);
	} finally {
		await runner.stop();
	}
	return exitCode;
}
function prependDirIfMissing(spec, dir) {
	if (/\bdir\s*=/.test(spec)) return spec;
	return `dir=${dir},${spec}`;
}
async function runOneShot(runner, opts, quiet) {
	const print = (v) => {
		process.stdout.write(`${JSON.stringify(v, null, quiet ? 0 : 2)}\n`);
	};
	try {
		if (opts.describe) {
			print(await runner.describe());
			return 0;
		}
		if (opts.health) {
			print(await runner.health());
			return 0;
		}
		if (opts.invoke) {
			let args = {};
			if (opts.args) try {
				const parsed = JSON.parse(opts.args);
				if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new Error("--args must be a JSON object");
				args = parsed;
			} catch (e) {
				console.error(red(`✗ --args: ${e.message}`));
				return 2;
			}
			print(await runner.invoke(opts.invoke, args));
			return 0;
		}
		return 0;
	} catch (e) {
		console.error(red(`✗ ${e.message}`));
		return 1;
	}
}
async function runRepl(runner) {
	const rl = createInterface({
		input: process.stdin,
		output: process.stdout
	});
	console.log(yellow("  REPL ready. Commands: describe | health | invoke <tool> <json?> | quit"));
	activeRepl = { rl };
	for (;;) {
		let line;
		try {
			line = (await rl.question(cyan("executa> "))).trim();
		} catch {
			break;
		}
		if (!line) continue;
		if (line === "quit" || line === "exit" || line === ".q") break;
		try {
			if (line === "describe") {
				const out = await runner.describe();
				process.stdout.write(`${JSON.stringify(out, null, 2)}\n`);
			} else if (line === "health") {
				const out = await runner.health();
				process.stdout.write(`${JSON.stringify(out, null, 2)}\n`);
			} else if (line.startsWith("invoke")) {
				const rest = line.slice(6).trim();
				const sp = rest.indexOf(" ");
				const tool = sp === -1 ? rest : rest.slice(0, sp);
				const argsStr = sp === -1 ? "{}" : rest.slice(sp + 1).trim() || "{}";
				if (!tool) {
					console.error(red("usage: invoke <tool> <json-args?>"));
					continue;
				}
				let args;
				try {
					const parsed = JSON.parse(argsStr);
					if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new Error("args must be a JSON object");
					args = parsed;
				} catch (e) {
					console.error(red(`bad args: ${e.message}`));
					continue;
				}
				const out = await runner.invoke(tool, args);
				process.stdout.write(`${JSON.stringify(out, null, 2)}\n`);
			} else console.error(yellow(`unknown command: ${line}`));
		} catch (e) {
			console.error(red(`✗ ${e.message}`));
		}
	}
	activeRepl = null;
	rl.close();
	return 0;
}
/**
* Write to stderr without trampling the REPL prompt. When the REPL is
* idle (no `activeRepl`) or stdout is not a TTY, this is a plain
* `process.stderr.write`. When the REPL is at a prompt, we clear the
* prompt line on stdout, emit the stderr text, then redraw the prompt
* with any user input the reader had already buffered.
*/
function writeStderrCooperative(text) {
	const repl = activeRepl;
	if (!repl || !process.stdout.isTTY) {
		process.stderr.write(text);
		return;
	}
	readline.cursorTo(process.stdout, 0);
	readline.clearLine(process.stdout, 0);
	process.stderr.write(text);
	const rl = repl.rl;
	if (typeof rl._refreshLine === "function") rl._refreshLine();
	else process.stdout.write(cyan("executa> "));
}

//#endregion
export { runExecutaDev };
