// @anna-ai/cli/dist/executa-cache-Do1D7xC0.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { dirname, join, resolve } from "node:path";
import { existsSync, mkdirSync, readFileSync, renameSync, unlinkSync, writeFileSync } from "node:fs";
import { createHash } from "node:crypto";

//#region src/identity/executa-cache.ts
const CACHE_DIR = ".anna";
const CACHE_FILE = "executa.json";
function executaCachePath(cwd) {
	return resolve(cwd, CACHE_DIR, CACHE_FILE);
}
function readExecutaIdentity(cwd) {
	const p = executaCachePath(cwd);
	if (!existsSync(p)) return null;
	try {
		const raw = JSON.parse(readFileSync(p, "utf-8"));
		if (typeof raw.host === "string" && typeof raw.tool_id === "string" && typeof raw.executa_id === "number" && typeof raw.slug === "string" && (raw.executa_type === "tool" || raw.executa_type === "skill")) return {
			$schema: "anna-executa-identity/v1",
			host: canonicalHost(raw.host),
			executa_id: raw.executa_id,
			tool_id: raw.tool_id,
			slug: raw.slug,
			executa_type: raw.executa_type,
			first_published_at: raw.first_published_at
		};
	} catch {}
	return null;
}
function writeExecutaIdentity(cwd, identity) {
	const p = executaCachePath(cwd);
	mkdirSync(dirname(p), { recursive: true });
	const tmp = `${p}.tmp.${process.pid}`;
	writeFileSync(tmp, JSON.stringify({
		...identity,
		host: canonicalHost(identity.host)
	}, null, 2) + "\n", "utf-8");
	try {
		renameSync(tmp, p);
	} catch (e) {
		try {
			unlinkSync(tmp);
		} catch {}
		throw e;
	}
}
function executaCacheMatches(identity, host, slug) {
	if (!identity) return false;
	return identity.host === canonicalHost(host) && identity.slug === slug;
}
/**
* Remove the cache file (best-effort). Used when a cached identity
* points at a server-side row that no longer exists (404 on
* GET/PUT/POST against `executa_id`), so the next `executa publish`
* re-runs the draft+commit mint path with the idempotency key instead
* of dying with "Tool not found".
*/
function invalidateExecutaCache(cwd) {
	const p = executaCachePath(cwd);
	try {
		if (existsSync(p)) unlinkSync(p);
	} catch {}
}
/**
* Stable mint idempotency key — `sha256(host|slug|type)`.
*
* Sent on every `POST /my/drafts` so that a fresh CI runner without the
* `.anna/executa.json` cache short-circuits onto the existing draft /
* Executa instead of minting a brand-new `tool_id`.
*/
function mintIdempotencyKey(host, slug, executaType) {
	const canonical = canonicalHost(host);
	const h = createHash("sha256");
	h.update(canonical);
	h.update("|");
	h.update(slug);
	h.update("|");
	h.update(executaType);
	return h.digest("hex");
}
const _internal = {
	CACHE_DIR,
	CACHE_FILE: join(CACHE_DIR, CACHE_FILE)
};

//#endregion
export { executaCacheMatches, executaCachePath, invalidateExecutaCache, mintIdempotencyKey, readExecutaIdentity, writeExecutaIdentity };
