// @anna-ai/cli/dist/executa-reads-CjIr8BuD.js
import "./credentials-BTv2IfUZ.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { getMyTool, listMyTools, listToolVersions } from "./executas-BhFed1B-.js";
import { bold, cyan, dim, green } from "kleur/colors";

//#region src/commands/executa-reads.ts
async function runExecutaList(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const published = await listMyTools(client);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				executas: published
			}, null, 2));
			return 0;
		}
		console.log(bold(cyan(`my executas`)) + " " + dim(`(${host})`));
		if (published.length === 0) {
			console.log(dim("  (none — run `anna-app executa publish` from a project to mint one; dev harness placeholders show up under `anna-app apps list`)"));
			return 0;
		}
		for (const r of published) {
			const vis = r.visibility ? `[${r.visibility}]` : "";
			console.log(`  ${green("[published]")} ${r.tool_id} ` + dim(`(id=${r.id}, type=${r.executa_type}) ${vis}`));
		}
		return 0;
	});
}
async function findExecutaByRef(client, ref) {
	const rows = await listMyTools(client);
	const exact = rows.find((r) => r.tool_id === ref);
	if (exact) return exact;
	const slug = rows.find((r) => {
		const parts = r.tool_id.split("-");
		return parts.includes(ref);
	});
	return slug ?? null;
}
async function runExecutaStatus(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const ex = await findExecutaByRef(client, opts.ref);
		if (!ex) throw new CliError(`no executa matching "${opts.ref}" for current user on ${host}`, 5);
		const full = await getMyTool(client, ex.id);
		const versions = await listToolVersions(client, ex.id);
		const latest = versions[0] ?? null;
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				executa_id: full.id,
				tool_id: full.tool_id,
				name: full.name,
				executa_type: full.executa_type,
				visibility: full.visibility ?? null,
				is_published: full.is_published ?? false,
				version_count: versions.length,
				latest_version: latest ? {
					id: latest.id,
					version: latest.version,
					content_hash: latest.content_hash,
					in_published_app: latest.in_published_app ?? false
				} : null
			}, null, 2));
			return 0;
		}
		console.log(bold(cyan(full.tool_id)) + " " + dim(`(${host})`));
		console.log(`  ${dim("name      :")} ${full.name}`);
		console.log(`  ${dim("type      :")} ${full.executa_type}`);
		console.log(`  ${dim("visibility:")} ${full.visibility ?? "(unknown)"}`);
		console.log(`  ${dim("published :")} ${full.is_published ? green("yes") : "no"}`);
		if (latest) console.log(`  ${dim("latest    :")} v${latest.version} (id=${latest.id})`);
		else console.log(`  ${dim("latest    :")} ${dim("(no versions)")}`);
		return 0;
	});
}
async function runExecutaVersions(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const ex = await findExecutaByRef(client, opts.ref);
		if (!ex) throw new CliError(`no executa matching "${opts.ref}" for current user on ${host}`, 5);
		const versions = await listToolVersions(client, ex.id);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				executa_id: ex.id,
				tool_id: ex.tool_id,
				versions
			}, null, 2));
			return 0;
		}
		console.log(bold(cyan(ex.tool_id)) + " " + dim(`versions (${versions.length})`));
		if (versions.length === 0) {
			console.log(dim("  (none)"));
			return 0;
		}
		for (const v of versions) {
			const inApp = v.in_published_app ? green("[anchored]") : "          ";
			console.log(`  ${inApp} v${v.version} ${dim(`(id=${v.id}, hash=${v.content_hash.slice(0, 12)}…)`)}`);
		}
		return 0;
	});
}

//#endregion
export { runExecutaList, runExecutaStatus, runExecutaVersions };
