// @anna-ai/cli/dist/fixture-BJqc1IfX.js
import { resolve } from "node:path";
import kleur from "kleur";
import { readFile } from "node:fs/promises";

//#region src/fixture/parse.ts
const KNOWN_KINDS = new Set([
	"meta",
	"auth.refresh",
	"event",
	"req",
	"res"
]);
async function parseFixtureFile(path) {
	const raw = await readFile(path, "utf8");
	return parseFixtureText(raw);
}
function parseFixtureText(text) {
	const out = {
		lines: [],
		errors: []
	};
	const rows = text.split("\n");
	rows.forEach((row, idx) => {
		const lineNo = idx + 1;
		const trimmed = row.trim();
		if (!trimmed) return;
		let parsed;
		try {
			parsed = JSON.parse(trimmed);
		} catch (e) {
			out.errors.push({
				line: lineNo,
				message: `not valid JSON: ${e.message}`
			});
			return;
		}
		if (typeof parsed !== "object" || parsed === null) {
			out.errors.push({
				line: lineNo,
				message: "expected JSON object"
			});
			return;
		}
		const obj = parsed;
		const kind = obj.kind;
		if (!kind || !KNOWN_KINDS.has(kind)) {
			out.errors.push({
				line: lineNo,
				message: `unknown kind: ${kind ?? "(missing)"}`
			});
			return;
		}
		if (typeof obj.t !== "number" || !Number.isFinite(obj.t)) {
			out.errors.push({
				line: lineNo,
				message: "missing or non-numeric 't'"
			});
			return;
		}
		out.lines.push(obj);
	});
	return out;
}
function verifyFixture(parsed) {
	const errs = [...parsed.errors];
	let lastT = -Infinity;
	parsed.lines.forEach((ln, idx) => {
		if (ln.t < lastT) errs.push({
			line: idx + 1,
			message: `non-monotonic t (${ln.t} < ${lastT})`
		});
		lastT = ln.t;
	});
	if (parsed.lines.length > 0 && parsed.lines[0].kind !== "meta") errs.push({
		line: 1,
		message: "first line should be a 'meta' record"
	});
	const seenIds = new Set();
	parsed.lines.forEach((ln, idx) => {
		const lineNo = idx + 1;
		switch (ln.kind) {
			case "auth.refresh":
				if (typeof ln.token !== "string" || !ln.token) errs.push({
					line: lineNo,
					message: "auth.refresh missing token"
				});
				break;
			case "event":
				if (typeof ln.event !== "string" || !ln.event) errs.push({
					line: lineNo,
					message: "event missing 'event' name"
				});
				break;
			case "req":
				if (ln.ns == null || ln.method == null || ln.id == null) errs.push({
					line: lineNo,
					message: "req missing ns/method/id"
				});
				else {
					const k = `${ln.id}`;
					if (seenIds.has(k)) errs.push({
						line: lineNo,
						message: `duplicate req id: ${k}`
					});
					seenIds.add(k);
				}
				break;
			case "res":
				if (ln.id == null || typeof ln.ok !== "boolean") errs.push({
					line: lineNo,
					message: "res missing id/ok"
				});
				break;
			case "meta": break;
		}
	});
	const reqIds = new Set(parsed.lines.filter((l) => l.kind === "req").map((l) => `${l.id}`));
	for (const ln of parsed.lines) if (ln.kind === "res" && !reqIds.has(`${ln.id}`)) errs.push({
		line: 0,
		message: `res with id=${ln.id} has no matching req`
	});
	return errs;
}
function summarizeFixture(parsed) {
	const meta = parsed.lines.find((l) => l.kind === "meta") ?? null;
	const events_by_name = {};
	const reqs_by_method = {};
	let reqs_total = 0;
	let res_total = 0;
	let res_errors = 0;
	let auth_refresh_count = 0;
	const reqIds = new Map();
	const respondedIds = new Set();
	let lastT = 0;
	for (const ln of parsed.lines) {
		if (ln.t > lastT) lastT = ln.t;
		switch (ln.kind) {
			case "event":
				events_by_name[ln.event] = (events_by_name[ln.event] ?? 0) + 1;
				break;
			case "req": {
				reqs_total += 1;
				const key = `${ln.ns}.${ln.method}`;
				reqs_by_method[key] = (reqs_by_method[key] ?? 0) + 1;
				reqIds.set(`${ln.id}`, ln);
				break;
			}
			case "res":
				res_total += 1;
				if (!ln.ok) res_errors += 1;
				respondedIds.add(`${ln.id}`);
				break;
			case "auth.refresh":
				auth_refresh_count += 1;
				break;
			case "meta": break;
		}
	}
	const unmatched_req_ids = [];
	for (const [id, req] of reqIds) if (!respondedIds.has(id)) unmatched_req_ids.push(req.id);
	return {
		duration_ms: lastT,
		total_lines: parsed.lines.length,
		meta,
		events_by_name,
		reqs_by_method,
		reqs_total,
		res_total,
		res_errors,
		auth_refresh_count,
		unmatched_req_ids
	};
}

//#endregion
//#region src/commands/fixture.ts
async function runFixtureVerify(opts) {
	const path = resolve(opts.file);
	const parsed = await parseFixtureFile(path);
	const errors = verifyFixture(parsed);
	if (opts.json) process.stdout.write(`${JSON.stringify({
		ok: errors.length === 0,
		errors,
		lines: parsed.lines.length
	}, null, 2)}\n`);
	else if (errors.length === 0) process.stdout.write(`${kleur.green("✓")} ${path} — ${parsed.lines.length} records, no errors\n`);
	else {
		process.stdout.write(`${kleur.red("✗")} ${path} — ${errors.length} error(s):\n`);
		for (const e of errors) {
			const loc = e.line > 0 ? `:${e.line}` : "";
			process.stdout.write(`  ${kleur.red("•")} ${path}${loc}  ${e.message}\n`);
		}
	}
	return errors.length === 0 ? 0 : 1;
}
async function runFixtureSummarize(opts) {
	const path = resolve(opts.file);
	const parsed = await parseFixtureFile(path);
	const sum = summarizeFixture(parsed);
	if (opts.json) {
		process.stdout.write(`${JSON.stringify(sum, null, 2)}\n`);
		return 0;
	}
	const out = [];
	out.push(kleur.bold(`Fixture: ${path}`));
	out.push(`  records:        ${sum.total_lines}`);
	out.push(`  duration:       ${(sum.duration_ms / 1e3).toFixed(2)}s`);
	if (sum.meta) out.push(`  recorded by:    harness ${sum.meta.harness_version ?? "?"}  session=${sum.meta.session_id ?? "?"}`);
	out.push(`  requests:       ${sum.reqs_total}  (responses: ${sum.res_total}, errors: ${sum.res_errors})`);
	out.push(`  events:         ${Object.values(sum.events_by_name).reduce((a, b) => a + b, 0)}`);
	out.push(`  auth refresh:   ${sum.auth_refresh_count}`);
	if (sum.unmatched_req_ids.length) out.push(`  ${kleur.yellow("⚠ unmatched req ids")}: ${sum.unmatched_req_ids.slice(0, 8).join(", ")}${sum.unmatched_req_ids.length > 8 ? " …" : ""}`);
	if (sum.reqs_total > 0) {
		out.push("  by ns.method:");
		for (const [k, v] of Object.entries(sum.reqs_by_method).sort((a, b) => b[1] - a[1])) out.push(`    ${v.toString().padStart(4)}  ${k}`);
	}
	if (Object.keys(sum.events_by_name).length > 0) {
		out.push("  events:");
		for (const [k, v] of Object.entries(sum.events_by_name).sort((a, b) => b[1] - a[1])) out.push(`    ${v.toString().padStart(4)}  ${k}`);
	}
	process.stdout.write(`${out.join("\n")}\n`);
	return 0;
}
async function runFixtureReplay(opts) {
	const path = resolve(opts.file);
	const parsed = await parseFixtureFile(path);
	const errors = verifyFixture(parsed);
	if (errors.length > 0) {
		process.stdout.write(`${kleur.red("✗")} fixture has ${errors.length} verification error(s); fix before replay\n`);
		return 1;
	}
	const manifest = opts.manifest ? await readManifest(opts.manifest) : null;
	process.stdout.write(`${kleur.bold("replay plan")} ${path}` + (manifest ? `  (manifest: ${opts.manifest})` : "  (no manifest)") + "\n");
	let i = 0;
	for (const ln of parsed.lines) {
		i += 1;
		const tag = i.toString().padStart(4);
		const ts = `+${(ln.t / 1e3).toFixed(3)}s`.padStart(9);
		switch (ln.kind) {
			case "event": {
				const ev = ln;
				process.stdout.write(`  ${tag}  ${ts}  ${kleur.cyan("→ event")}      ${ev.event}\n`);
				break;
			}
			case "req": {
				const r = ln;
				process.stdout.write(`  ${tag}  ${ts}  ${kleur.yellow("← req  ")}     ${r.ns}.${r.method}  id=${r.id}\n`);
				break;
			}
			case "res": {
				const r = ln;
				const tone = r.ok ? kleur.green : kleur.red;
				process.stdout.write(`  ${tag}  ${ts}  ${tone("→ res  ")}     id=${r.id}  ok=${r.ok}\n`);
				break;
			}
			case "auth.refresh":
				process.stdout.write(`  ${tag}  ${ts}  ${kleur.magenta("→ auth ")}     refresh\n`);
				break;
			case "meta":
				process.stdout.write(`  ${tag}  ${ts}  ${kleur.gray("· meta ")}     start\n`);
				break;
		}
	}
	process.stdout.write("\n" + kleur.gray("live replay against `anna-app dev` harness is tracked as a Phase 6 follow-up; this MVP performs structural replay only.") + "\n");
	return 0;
}
async function readManifest(path) {
	try {
		const raw = await readFile(resolve(path), "utf8");
		return JSON.parse(raw);
	} catch (e) {
		process.stderr.write(`failed to read manifest at ${path}: ${e.message}\n`);
		return null;
	}
}

//#endregion
export { runFixtureReplay, runFixtureSummarize, runFixtureVerify };
