// @anna-ai/cli/dist/host_upload-sfo4UHlR.js
import { canonicalHost, getAccount } from "./credentials-BTv2IfUZ.js";
import { resolve } from "node:path";
import { existsSync, readFileSync } from "node:fs";
import { createHash } from "node:crypto";

//#region src/executa/host_upload.ts
var HostUploadBridge = class {
	mocks = [];
	cachedSession = null;
	constructor(opts) {
		this.opts = opts;
		if (opts.mode === "mock" && opts.mockFile) {
			const path = resolve(opts.mockFile);
			if (existsSync(path)) for (const line of readFileSync(path, "utf8").split(/\r?\n/)) {
				if (!line.trim() || line.startsWith("#")) continue;
				try {
					this.mocks.push(JSON.parse(line));
				} catch {}
			}
		}
	}
	async call(method, params) {
		if (method !== "host/uploadFile") throw withCode(new Error(`unsupported host method: ${method}`), -32601);
		if (this.opts.mode === "off") throw withCode(new Error("host upload disabled — rerun without `--no-upload`"), -32201);
		if (this.opts.mode === "mock") return this.mockCall(params);
		return this.realCall(params);
	}
	mockCall(params) {
		const mode = String(params.mode ?? "inline");
		const filename = String(params.filename ?? "");
		const candidates = this.mocks.filter((m) => m.ns === "host" && m.method === "uploadFile");
		const matched = candidates.find((m) => (!m.match?.modeEquals || m.match.modeEquals === mode) && (!m.match?.filenameIncludes || filename.includes(m.match.filenameIncludes))) ?? candidates[0];
		if (matched && matched.result) return matched.result;
		if (mode === "inline") {
			const b64 = String(params.content_b64 ?? "");
			const mime = String(params.mime_type ?? "application/octet-stream");
			const url = `data:${mime};base64,${b64}`;
			const r2Key = `mock/${hashShort(filename + ":" + b64)}/${filename || "blob"}`;
			return {
				download_url: url,
				r2_key: r2Key,
				size_bytes: Math.floor(b64.length * .75),
				expires_at: new Date(Date.now() + 3600 * 1e3).toISOString()
			};
		}
		if (mode === "negotiate") {
			const key = `mock/${hashShort(filename + Date.now())}/${filename || "blob"}`;
			return {
				put_url: `https://mock.local/${key}?signature=mock`,
				headers: { "content-type": String(params.mime_type ?? "") },
				r2_key: key,
				expires_at: new Date(Date.now() + 600 * 1e3).toISOString()
			};
		}
		if (mode === "confirm") {
			const key = String(params.r2_key ?? "");
			return {
				download_url: `https://mock.local/${key}`,
				r2_key: key,
				size_bytes: 0,
				expires_at: new Date(Date.now() + 3600 * 1e3).toISOString()
			};
		}
		throw withCode(new Error(`unknown upload mode: ${mode}`), -32203);
	}
	account() {
		const acc = getAccount(this.opts.account);
		if (!acc) throw withCode(new Error("no PAT on disk — run `anna-app login --host <nexus-url>` first (or pass `--mock-upload <fixture>` / `--no-upload`)"), -32201);
		if (acc.expires_at && acc.expires_at < Math.floor(Date.now() / 1e3)) throw withCode(new Error("PAT expired — run `anna-app login` again"), -32201);
		return acc;
	}
	async mint() {
		if (this.cachedSession && this.cachedSession.expiresAt - 30 > Math.floor(Date.now() / 1e3)) return this.cachedSession;
		const acc = this.account();
		if (!this.opts.appSlug) throw withCode(new Error("host upload bridge has no app_slug — pass `--app-slug <slug>`"), -32203);
		const url = `${canonicalHost(acc.host)}/api/v1/anna-apps/dev/session/mint`;
		const res = await fetch(url, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify({
				pat: acc.pat,
				kind: "complete",
				app_slug: this.opts.appSlug
			})
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`session.mint failed: HTTP ${res.status} ${text}`), -32207);
		}
		const body = await res.json();
		this.cachedSession = {
			appSessionToken: body.app_session_token,
			expiresAt: Math.floor(Date.now() / 1e3) + (body.expires_in || 600)
		};
		return this.cachedSession;
	}
	async realCall(params) {
		const acc = this.account();
		const session = await this.mint();
		const mode = String(params.mode ?? "inline");
		const path = mode === "negotiate" ? "/api/v1/copilot/app/upload/negotiate" : mode === "confirm" ? "/api/v1/copilot/app/upload/confirm" : "/api/v1/copilot/app/upload";
		const res = await fetch(`${canonicalHost(acc.host)}${path}`, {
			method: "POST",
			headers: {
				"content-type": "application/json",
				authorization: `Bearer ${session.appSessionToken}`
			},
			body: JSON.stringify(params)
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`HTTP ${res.status}: ${text}`), httpToUploadCode(res.status));
		}
		return res.json();
	}
};
function withCode(err, code) {
	err.rpcCode = code;
	return err;
}
function hashShort(s) {
	return createHash("sha256").update(s).digest("hex").slice(0, 12);
}
function httpToUploadCode(status) {
	if (status === 403) return -32201;
	if (status === 429) return -32202;
	if (status === 400) return -32203;
	if (status === 413) return -32204;
	if (status === 415) return -32205;
	if (status === 404) return -32212;
	if (status === 504) return -32208;
	return -32207;
}

//#endregion
export { HostUploadBridge };
