// @anna-ai/cli/dist/image-Bdi4Hu0m.js
import { canonicalHost, getAccount } from "./credentials-BTv2IfUZ.js";
import { resolve } from "node:path";
import { existsSync, readFileSync } from "node:fs";

//#region src/executa/image.ts
var ImageBridge = class {
	mocks = [];
	cachedSession = null;
	constructor(opts) {
		this.opts = opts;
		if (opts.mode === "mock" && opts.mockFile) {
			const path = resolve(opts.mockFile);
			if (existsSync(path)) for (const line of readFileSync(path, "utf8").split(/\r?\n/)) {
				if (!line.trim() || line.startsWith("#")) continue;
				try {
					this.mocks.push(JSON.parse(line));
				} catch {}
			}
		}
	}
	async generate(req) {
		if (this.opts.mode === "off") throw withCode(new Error("image generation disabled — rerun without `--no-image`"), -32101);
		if (this.opts.mode === "mock") return this.pickMock("generate", req.prompt);
		return this.realCall("/api/v1/copilot/app/image/generate", req);
	}
	async edit(req) {
		if (this.opts.mode === "off") throw withCode(new Error("image edit disabled — rerun without `--no-image`"), -32101);
		if (this.opts.mode === "mock") return this.pickMock("edit", req.prompt);
		return this.realCall("/api/v1/copilot/app/image/edit", req);
	}
	pickMock(method, prompt) {
		const wantedMethod = method === "generate" ? "generate" : "edit";
		const candidates = this.mocks.filter((m) => m.ns === "image" && m.method === wantedMethod);
		const matched = candidates.find((m) => m.match?.promptIncludes && prompt.includes(m.match.promptIncludes)) ?? candidates[0];
		if (matched && matched.result) return normaliseImageResult(matched.result);
		return {
			images: [{
				url: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=",
				mimeType: "image/png"
			}],
			model: "mock-image-model"
		};
	}
	account() {
		const acc = getAccount(this.opts.account);
		if (!acc) throw withCode(new Error("no PAT on disk — run `anna-app login --host <nexus-url>` first (or pass `--mock-image <fixture>` / `--no-image`)"), -32101);
		if (acc.expires_at && acc.expires_at < Math.floor(Date.now() / 1e3)) throw withCode(new Error("PAT expired — run `anna-app login` again"), -32101);
		return acc;
	}
	async mint() {
		if (this.cachedSession && this.cachedSession.expiresAt - 30 > Math.floor(Date.now() / 1e3)) return this.cachedSession;
		const acc = this.account();
		if (!this.opts.appSlug) throw withCode(new Error("image bridge has no app_slug — pass `--app-slug <slug>` to `anna-app executa dev` so nexus can attribute the spend"), -32104);
		const url = `${canonicalHost(acc.host)}/api/v1/anna-apps/dev/session/mint`;
		const res = await fetch(url, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify({
				pat: acc.pat,
				kind: "complete",
				app_slug: this.opts.appSlug
			})
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw withCode(new Error(`session.mint failed: HTTP ${res.status} ${text}`), -32103);
		}
		const body = await res.json();
		this.cachedSession = {
			appSessionToken: body.app_session_token,
			expiresAt: Math.floor(Date.now() / 1e3) + (body.expires_in || 600)
		};
		return this.cachedSession;
	}
	async realCall(path, body) {
		const acc = this.account();
		const session = await this.mint();
		const res = await fetch(`${canonicalHost(acc.host)}${path}`, {
			method: "POST",
			headers: {
				"content-type": "application/json",
				authorization: `Bearer ${session.appSessionToken}`
			},
			body: JSON.stringify(body)
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw withCode(
				new Error(`HTTP ${res.status}: ${text}`),
				// Map common HTTP statuses to canonical image codes; otherwise
				// let the host's body propagate.
				httpToImageCode(res.status)
);
		}
		const out = await res.json();
		return normaliseImageResult(out);
	}
};
function withCode(err, code) {
	err.rpcCode = code;
	return err;
}
function httpToImageCode(status) {
	if (status === 403) return -32101;
	if (status === 429) return -32102;
	if (status === 400) return -32104;
	if (status === 504) return -32105;
	return -32103;
}
function normaliseImageResult(raw) {
	if (!raw || typeof raw !== "object") return {
		images: [],
		model: "unknown"
	};
	const o = raw;
	const images = Array.isArray(o.images) ? o.images.map((img) => ({
		url: String(img.url ?? ""),
		mimeType: img.mimeType ? String(img.mimeType) : void 0,
		width: typeof img.width === "number" ? img.width : void 0,
		height: typeof img.height === "number" ? img.height : void 0
	})) : [];
	return {
		images,
		model: typeof o.model === "string" ? o.model : void 0,
		quota_used: o.quota_used && typeof o.quota_used === "object" ? o.quota_used : void 0,
		_meta: o._meta && typeof o._meta === "object" ? o._meta : void 0
	};
}

//#endregion
export { ImageBridge };
