// @anna-ai/cli/dist/login-CslsHbsq.js
import { canonicalHost, saveAccount } from "./credentials-BTv2IfUZ.js";

//#region src/commands/login.ts
const POLL_TIMEOUT_MS = 12 * 60 * 1e3;
function envBool(v) {
	return !!v && /^(1|true|yes|on)$/i.test(v);
}
async function tryOpenBrowser(url) {
	if (envBool(process.env.ANNA_LOGIN_NO_OPEN)) return;
	const opener = process.platform === "darwin" ? "open" : process.platform === "win32" ? "start" : "xdg-open";
	try {
		const { spawn } = await import("node:child_process");
		spawn(opener, [url], {
			detached: true,
			stdio: "ignore"
		}).unref();
	} catch {}
}
async function runLogin(opts) {
	const host = canonicalHost(opts.host);
	process.stdout.write(`▸ logging in to ${host}\n`);
	let startResp;
	try {
		const res = await fetch(`${host}/api/v1/anna-apps/dev/login/start`, {
			method: "POST",
			headers: { "content-type": "application/json" }
		});
		if (!res.ok) {
			console.error(`✗ /dev/login/start failed: HTTP ${res.status}`);
			return 1;
		}
		startResp = await res.json();
	} catch (e) {
		console.error(`✗ cannot reach nexus: ${e.message}`);
		return 1;
	}
	const verifyUrl = `${host}${startResp.verification_uri_complete}`;
	process.stdout.write(`\n  user_code: \x1b[1m${startResp.user_code}\x1b[0m\n`);
	process.stdout.write(`  open:      ${verifyUrl}\n`);
	process.stdout.write(`  (waiting up to ${Math.round(startResp.expires_in / 60)} min)\n\n`);
	if (!opts.noBrowser) await tryOpenBrowser(verifyUrl);
	const interval = Math.max(2, Math.min(15, startResp.interval || 5));
	const deadline = Date.now() + Math.min(POLL_TIMEOUT_MS, startResp.expires_in * 1e3);
	const spinner = [
		"⠋",
		"⠙",
		"⠹",
		"⠸",
		"⠼",
		"⠴",
		"⠦",
		"⠧",
		"⠇",
		"⠏"
	];
	let frame = 0;
	while (Date.now() < deadline) {
		process.stdout.write(`\r  ${spinner[frame++ % spinner.length]} polling...`);
		await new Promise((r) => setTimeout(r, interval * 1e3));
		let resp;
		try {
			resp = await fetch(`${host}/api/v1/anna-apps/dev/login/poll`, {
				method: "POST",
				headers: { "content-type": "application/json" },
				body: JSON.stringify({ device_code: startResp.device_code })
			});
		} catch (e) {
			process.stdout.write("\n");
			console.error(`✗ poll failed: ${e.message}`);
			return 1;
		}
		if (resp.status === 202) continue;
		if (resp.status === 410) {
			process.stdout.write("\n✗ device code expired\n");
			return 1;
		}
		if (resp.status === 200) {
			process.stdout.write("\n");
			const ok = await resp.json();
			const now = Math.floor(Date.now() / 1e3);
			saveAccount({
				host,
				user_id: null,
				pat: ok.pat,
				issued_at: now,
				expires_at: now + (ok.expires_in || 0),
				scopes: ok.scopes
			});
			console.log(`✓ logged in. PAT saved to ~/.config/anna/credentials.json`);
			console.log(`  expires in ~${Math.round((ok.expires_in || 0) / 86400)}d`);
			return 0;
		}
		process.stdout.write("\n");
		console.error(`✗ unexpected status: HTTP ${resp.status}`);
		return 1;
	}
	process.stdout.write("\n✗ timed out waiting for approval\n");
	return 1;
}

//#endregion
export { runLogin };
