// @anna-ai/cli/dist/manifest-CMc7RM57.js
import { CliError } from "./client-D-_z1ALk.js";
import { resolve } from "node:path";
import { existsSync, readFileSync } from "node:fs";

//#region src/publish/manifest.ts
const DISTRIBUTION_TYPES = new Set([
	"uv",
	"npm",
	"homebrew",
	"binary",
	"pipx",
	"local"
]);
/** Distribution types that install a named package from an index. */
const PACKAGE_BACKED = new Set([
	"uv",
	"npm",
	"homebrew",
	"pipx"
]);
const SLUG_RE = /^[a-z0-9][a-z0-9-]{1,78}[a-z0-9]$/;
const SEMVER_RE = /^\d+\.\d+\.\d+(?:[-+][0-9A-Za-z.-]+)?$/;
function readJson(path) {
	const raw = readFileSync(path, "utf-8");
	try {
		const parsed = JSON.parse(raw);
		if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) throw new CliError(`manifest at ${path} must be a JSON object`, 4);
		return parsed;
	} catch (e) {
		if (e instanceof CliError) throw e;
		throw new CliError(`failed to parse ${path}: ${e.message}`, 4);
	}
}
function loadAppManifest(cwd, manifestPath) {
	const path = resolve(cwd, manifestPath ?? "manifest.json");
	if (!existsSync(path)) throw new CliError(`no manifest.json at ${path}`, 4);
	const raw = readJson(path);
	const appJsonPath = resolve(cwd, "app.json");
	const hasAppJson = existsSync(appJsonPath);
	const appJsonRaw = hasAppJson ? readJson(appJsonPath) : null;
	const metadataPath = hasAppJson ? appJsonPath : path;
	const metadataRaw = appJsonRaw ?? raw;
	const metadataLabel = hasAppJson ? "app.json" : "manifest.json";
	const pickMeta = (k) => (appJsonRaw && pickString(appJsonRaw, k)) ?? pickString(raw, k);
	const slug = pickMeta("slug");
	const name = pickMeta("name") ?? slug ?? "";
	const version = pickMeta("version") ?? "0.1.0";
	if (!slug) throw new CliError(`${metadataLabel} missing required \`slug\` (3-80 chars, kebab-case)`, 4);
	if (!SLUG_RE.test(slug)) throw new CliError(`slug "${slug}" is not valid — must match ${SLUG_RE.source}`, 4);
	if (!SEMVER_RE.test(version)) throw new CliError(`${metadataLabel} \`version\` "${version}" is not semver`, 4);
	const bundledExecutas = parseBundledExecutas((appJsonRaw && appJsonRaw["bundled_executas"]) ?? raw["bundled_executas"], metadataLabel);
	return {
		kind: "app",
		path,
		raw,
		metadataPath,
		metadataRaw,
		cwd: resolve(cwd),
		bundledExecutas,
		slug,
		name,
		version,
		category: pickMeta("category"),
		tagline: pickMeta("tagline"),
		description: pickMeta("description"),
		changelog: pickMeta("changelog")
	};
}
const HANDLE_RE = /^[a-z0-9][a-z0-9-]{0,62}$/;
function parseBundledExecutas(value, label) {
	if (value === void 0 || value === null) return [];
	if (typeof value !== "object" || Array.isArray(value)) throw new CliError(`${label} \`bundled_executas\` must be an object mapping handle → { path }`, 4);
	const out = [];
	for (const [handle, decl] of Object.entries(value)) {
		if (!HANDLE_RE.test(handle)) throw new CliError(`bundled_executas handle "${handle}" is invalid — must be kebab-case, ≤63 chars`, 4);
		if (decl === null || typeof decl !== "object" || Array.isArray(decl)) throw new CliError(`bundled_executas["${handle}"] must be an object with a \`path\` field`, 4);
		const p = pickString(decl, "path");
		if (!p) throw new CliError(`bundled_executas["${handle}"] missing required \`path\``, 4);
		out.push({
			handle,
			path: p
		});
	}
	return out;
}
function loadExecutaManifest(cwd, manifestPath) {
	const path = resolve(cwd, manifestPath ?? "executa.json");
	if (!existsSync(path)) throw new CliError(`no executa.json at ${path}`, 4);
	const raw = readJson(path);
	const slug = pickString(raw, "slug");
	const name = pickString(raw, "name") ?? slug ?? "";
	const version = pickString(raw, "version") ?? "0.1.0";
	const t = pickString(raw, "executa_type") ?? "tool";
	if (t !== "tool" && t !== "skill") throw new CliError(`executa.json \`executa_type\` must be "tool" or "skill" (got "${t}")`, 4);
	if (!slug) throw new CliError(`executa.json missing required \`slug\` (3-80 chars, kebab-case)`, 4);
	if (!SLUG_RE.test(slug)) throw new CliError(`slug "${slug}" is not valid — must match ${SLUG_RE.source}`, 4);
	if (!SEMVER_RE.test(version)) throw new CliError(`executa.json \`version\` "${version}" is not semver`, 4);
	let skillContent;
	if (t === "skill") {
		const contentFile = pickString(raw, "content_file") ?? "SKILL.md";
		const contentPath = resolve(cwd, contentFile);
		if (!existsSync(contentPath)) throw new CliError(`skill executa.json references content file "${contentFile}" but ${contentPath} does not exist`, 4);
		skillContent = readFileSync(contentPath, "utf-8");
		if (!skillContent.trim()) throw new CliError(`skill content file "${contentFile}" is empty`, 4);
	}
	return {
		kind: "executa",
		path,
		raw,
		slug,
		name,
		version,
		executa_type: t,
		description: pickString(raw, "description") ?? void 0,
		changelog: pickString(raw, "changelog") ?? void 0,
		skillContent,
		executionMode: pickString(raw, "execution_mode") ?? void 0,
		distribution: parseExecutaDistribution(raw["distribution"], t)
	};
}
/**
* Parse + validate the optional `executa.json` `distribution` object.
*
* Returns `undefined` when absent (a dev-only Executa). Skills never carry
* distribution config — they are declarative Markdown, not installable
* processes — so a `distribution` block on a skill is rejected.
*
* Two shapes are accepted:
*
*   1. Flat (single mode) — the historical form:
*
*        "distribution": { "type": "binary", "binary_urls": { … } }
*
*   2. Multi-profile (switchable) — keep several modes side by side and flip
*      `active` to choose which one is published:
*
*        "distribution": {
*          "active": "binary",
*          "profiles": {
*            "local":  { "type": "local",  "executable_name": "…" },
*            "binary": { "type": "binary", "binary_urls": { … } }
*          }
*        }
*
*      Every profile is validated (so a typo in an *inactive* profile is
*      still caught), but the returned `ExecutaDistribution` carries the
*      fields of the *active* profile plus `active`/`profiles` for
*      introspection. Downstream publish code stays profile-agnostic.
*/
function parseExecutaDistribution(value, executaType) {
	if (value === void 0 || value === null) return void 0;
	if (typeof value !== "object" || Array.isArray(value)) throw new CliError("executa.json `distribution` must be an object with a `type` field (or `profiles` for the switchable multi-mode form)", 4);
	if (executaType === "skill") throw new CliError("executa.json `distribution` is only valid for `executa_type: \"tool\"` — skills are declarative Markdown and are not installed as a process", 4);
	const d = value;
	if (d["profiles"] !== void 0) return parseProfiledDistribution(d);
	return parseDistributionProfile(d, "distribution");
}
/**
* Resolve the multi-profile `distribution` form into the selected profile.
* Validates *all* profiles and records them under `profiles`.
*/
function parseProfiledDistribution(d) {
	if (d["type"] !== void 0) throw new CliError("executa.json `distribution` cannot set both `type` (flat form) and `profiles` (switchable form) — pick one", 4);
	const profilesRaw = d["profiles"];
	if (profilesRaw === null || typeof profilesRaw !== "object" || Array.isArray(profilesRaw)) throw new CliError("executa.json `distribution.profiles` must be an object mapping a profile name → distribution config", 4);
	const profileEntries = Object.entries(profilesRaw);
	if (profileEntries.length === 0) throw new CliError("executa.json `distribution.profiles` must declare at least one profile", 4);
	const profiles = {};
	for (const [pname, pval] of profileEntries) {
		if (pval === null || typeof pval !== "object" || Array.isArray(pval)) throw new CliError(`executa.json \`distribution.profiles["${pname}"]\` must be an object`, 4);
		profiles[pname] = parseDistributionProfile(pval, `distribution.profiles["${pname}"]`);
	}
	const active = pickString(d, "active");
	let selected;
	if (active !== void 0) {
		if (!(active in profiles)) throw new CliError(`executa.json \`distribution.active\` "${active}" is not a declared profile — expected one of ${Object.keys(profiles).join(" | ")}`, 4);
		selected = active;
	} else if (profileEntries.length === 1) selected = profileEntries[0][0];
	else throw new CliError(`executa.json \`distribution.active\` is required when more than one profile is declared — expected one of ${Object.keys(profiles).join(" | ")}`, 4);
	const resolved = profiles[selected];
	return {
		...resolved,
		active: selected,
		profiles
	};
}
/**
* Parse + validate a single distribution config (one flat block or one entry
* of a `profiles` map). `ctx` is the JSON path label used in error messages.
*/
function parseDistributionProfile(d, ctx) {
	const type = pickString(d, "type");
	if (!type) throw new CliError(`executa.json \`${ctx}.type\` is required`, 4);
	if (!DISTRIBUTION_TYPES.has(type)) throw new CliError(`executa.json \`${ctx}.type\` "${type}" is invalid — expected one of ` + [...DISTRIBUTION_TYPES].join(" | "), 4);
	const distType = type;
	const packageName = pickString(d, "package_name");
	const executableName = pickString(d, "executable_name");
	const supportsProtocol = pickBool(d, "supports_protocol");
	const binaryUrls = parseBinaryUrls(d["binary_urls"], ctx);
	const capabilities = parseCapabilities(d["capabilities"], ctx);
	if (PACKAGE_BACKED.has(distType) && !packageName) throw new CliError(`executa.json \`${ctx}.package_name\` is required for \`${ctx}.type: "${distType}"\``, 4);
	if (distType === "binary" && (!binaryUrls || Object.keys(binaryUrls).length === 0)) throw new CliError(`executa.json \`${ctx}.binary_urls\` is required (and non-empty) for \`${ctx}.type: "binary"\``, 4);
	return {
		type: distType,
		packageName,
		executableName,
		supportsProtocol,
		binaryUrls,
		capabilities
	};
}
function parseBinaryUrls(value, ctx = "distribution") {
	if (value === void 0 || value === null) return void 0;
	if (typeof value !== "object" || Array.isArray(value)) throw new CliError(`executa.json \`${ctx}.binary_urls\` must be an object mapping platform → URL string or asset dict`, 4);
	const out = {};
	for (const [platform, asset] of Object.entries(value)) {
		if (typeof asset === "string") {
			if (!asset.trim()) throw new CliError(`executa.json \`${ctx}.binary_urls["${platform}"]\` is empty`, 4);
			out[platform] = asset;
			continue;
		}
		if (asset && typeof asset === "object" && !Array.isArray(asset)) {
			const url = pickString(asset, "url");
			if (!url) throw new CliError(`executa.json \`${ctx}.binary_urls["${platform}"]\` asset dict missing required \`url\``, 4);
			out[platform] = asset;
			continue;
		}
		throw new CliError(`executa.json \`${ctx}.binary_urls["${platform}"]\` must be a URL string or an asset dict \`{ url, sha256?, size?, entrypoint?, format? }\``, 4);
	}
	return out;
}
function parseCapabilities(value, ctx = "distribution") {
	if (value === void 0 || value === null) return void 0;
	if (!Array.isArray(value) || value.some((v) => typeof v !== "string")) throw new CliError(`executa.json \`${ctx}.capabilities\` must be an array of strings`, 4);
	return value;
}
function pickBool(o, k) {
	const v = o[k];
	return typeof v === "boolean" ? v : void 0;
}
function pickString(o, k) {
	const v = o[k];
	return typeof v === "string" ? v : void 0;
}

//#endregion
export { loadAppManifest, loadExecutaManifest };
