// @anna-ai/cli/dist/mascot-wlYTJqMs.js
import kleur from "kleur";

//#region src/mascot.ts
const FACE = [
	255,
	227,
	234
];
const EYE = [
	122,
	90,
	101
];
const GEO = (() => {
	const VX0 = 252, VY0 = 280, VW = 520;
	const n = (x, y) => [(x - VX0) / VW, (y - VY0) / VW];
	const r = (v) => v / VW;
	const [fcx, fcy] = n(512, 540);
	const [lex, ley] = n(462, 520);
	const [rex, rey] = n(562, 520);
	return {
		face: {
			cx: fcx,
			cy: fcy,
			r: r(260)
		},
		eyeL: {
			cx: lex,
			cy: ley,
			r: r(34)
		},
		eyeR: {
			cx: rex,
			cy: rey,
			r: r(34)
		}
	};
})();
/** Sample the SVG at normalized (x, y) ∈ [0, 1]². Returns null for empty. */
function sampleSvg(x, y) {
	const inCircle = (cx, cy, r) => {
		const dx = x - cx, dy = y - cy;
		return dx * dx + dy * dy <= r * r;
	};
	let c = null;
	if (inCircle(GEO.face.cx, GEO.face.cy, GEO.face.r)) c = FACE;
	return c;
}
/**
* 2×2 supersampled pixel using **majority vote** (not averaging).
* Anti-aliases silhouette edges (filled vs empty) while keeping every
* pixel a pure palette color — no muddy blends between the pink face and
* the dark eyes. Ties resolve in favor of the topmost layer encountered.
*/
function pixel(px, py, w, h) {
	const counts = new Map();
	let nulls = 0;
	for (const oy of [.25, .75]) for (const ox of [.25, .75]) {
		const s = sampleSvg((px + ox) / w, (py + oy) / h);
		if (!s) {
			nulls++;
			continue;
		}
		const key = `${s[0]},${s[1]},${s[2]}`;
		const entry = counts.get(key);
		if (entry) entry.n++;
		else counts.set(key, {
			color: s,
			n: 1
		});
	}
	if (counts.size === 0) return null;
	let best = null;
	for (const e of counts.values()) if (!best || e.n > best.n) best = e;
	if (nulls > (best?.n ?? 0)) return null;
	return best.color;
}
const FG = (c) => `\x1b[38;2;${c[0]};${c[1]};${c[2]}m`;
const BG = (c) => `\x1b[48;2;${c[0]};${c[1]};${c[2]}m`;
const BG_DEFAULT = "\x1B[49m";
const RESET = "\x1B[0m";
function eq(a, b) {
	if (a === b) return true;
	if (!a || !b) return false;
	return a[0] === b[0] && a[1] === b[1] && a[2] === b[2];
}
/**
* Rasterize the face into a list of half-block lines.
* `cols` is character width; pixel rows are derived from the cell aspect
* so each pixel covers an (approximately) square area on screen.
*/
function renderFace(cols, cellAspect) {
	const W = cols;
	const charRows = Math.max(1, Math.round(cols / cellAspect));
	const H = charRows * 2;
	const eyeRow = Math.round(GEO.eyeL.cy * charRows - .5);
	const eyeColL = Math.max(0, Math.round(GEO.eyeL.cx * W) - 1);
	const eyeColR = Math.max(0, Math.round(GEO.eyeR.cx * W) - 1);
	const isEyeCell = (cx, cy) => cy === eyeRow && (cx >= eyeColL && cx < eyeColL + 2 || cx >= eyeColR && cx < eyeColR + 2);
	const rows = [];
	for (let cy = 0; cy < charRows; cy++) {
		let line = "";
		let lastFg = null;
		let lastBg = null;
		let lastBgDefault = true;
		for (let cx = 0; cx < W; cx++) {
			if (isEyeCell(cx, cy)) {
				if (!eq(lastFg, EYE)) {
					line += FG(EYE);
					lastFg = EYE;
				}
				if (lastBgDefault || !eq(lastBg, FACE)) {
					line += BG(FACE);
					lastBg = FACE;
					lastBgDefault = false;
				}
				line += "█";
				continue;
			}
			const top = pixel(cx, cy * 2, W, H);
			const bot = pixel(cx, cy * 2 + 1, W, H);
			if (!top && !bot) {
				if (!lastBgDefault) {
					line += BG_DEFAULT;
					lastBgDefault = true;
					lastBg = null;
				}
				line += " ";
				continue;
			}
			if (top && bot) {
				if (!eq(lastFg, top)) {
					line += FG(top);
					lastFg = top;
				}
				if (lastBgDefault || !eq(lastBg, bot)) {
					line += BG(bot);
					lastBg = bot;
					lastBgDefault = false;
				}
				line += "▀";
			} else if (top) {
				if (!eq(lastFg, top)) {
					line += FG(top);
					lastFg = top;
				}
				if (!lastBgDefault) {
					line += BG_DEFAULT;
					lastBgDefault = true;
					lastBg = null;
				}
				line += "▀";
			} else {
				if (!eq(lastFg, bot)) {
					line += FG(bot);
					lastFg = bot;
				}
				if (!lastBgDefault) {
					line += BG_DEFAULT;
					lastBgDefault = true;
					lastBg = null;
				}
				line += "▄";
			}
		}
		rows.push(line + RESET);
	}
	return rows;
}
function mascotEnabled() {
	if (process.env.ANNA_CLI_NO_MASCOT) return false;
	if (process.env.NO_COLOR) return false;
	if (process.env.ANNA_CLI_FORCE_MASCOT) return true;
	return Boolean(process.stdout.isTTY);
}
function pickWidth() {
	const override = Number.parseInt(process.env.ANNA_CLI_MASCOT_WIDTH ?? "", 10);
	if (Number.isFinite(override) && override >= 10 && override <= 120) return override;
	const cols = process.stdout.columns || 80;
	if (cols < 60) return 10;
	if (cols >= 120) return 16;
	return 13;
}
function pickAspect() {
	const v = Number.parseFloat(process.env.ANNA_CLI_MASCOT_ASPECT ?? "");
	if (Number.isFinite(v) && v >= 1.2 && v <= 3) return v;
	return 2.1;
}
/**
* Render the mascot as a multi-line string. Always returns the colored art
* (callers decide whether to print based on `mascotEnabled`).
*/
function renderMascot(message) {
	const cols = pickWidth();
	const aspect = pickAspect();
	const face = renderFace(cols, aspect);
	const pad = " ".repeat(2);
	const label = kleur.magenta().bold("Anna");
	const tagline = kleur.gray("Anna App developer CLI");
	const textLines = [`${label}  ${tagline}`];
	if (message) textLines.push(`${kleur.gray("›")} ${message}`);
	const start = Math.max(0, Math.floor((face.length - textLines.length) / 2));
	const lines = [];
	for (let i = 0; i < face.length; i++) {
		const tIdx = i - start;
		const text = tIdx >= 0 && tIdx < textLines.length ? "  " + textLines[tIdx] : "";
		lines.push(pad + face[i] + text);
	}
	return lines.join("\n");
}
/** Print the mascot to stderr if enabled. Stderr keeps stdout clean for piping. */
function printMascot(message) {
	if (!mascotEnabled()) return;
	process.stderr.write(renderMascot(message) + "\n\n");
}

//#endregion
export { printMascot };
