// @anna-ai/cli/dist/runner-DVu-o_U8.js
import { spawn } from "node:child_process";
import { createInterface } from "node:readline";

//#region src/executa/runner.ts
/** Spec describing where the host advertises its capabilities. */
const HOST_INITIALIZE_PARAMS = {
	protocolVersion: "2.0",
	client_capabilities: {
		sampling: {},
		agent: {},
		storage: {},
		image: {},
		"image.edit": {},
		upload: {}
	},
	client_info: {
		name: "anna-app-cli",
		version: "executa-dev"
	}
};
var ExecutaRunner = class {
	proc = null;
	nextId = 1;
	pending = new Map();
	exitWaiters = [];
	closed = false;
	/** Resolved by start() after the v2 handshake (or v1 fallback) succeeds. */
	initialized = null;
	constructor(opts) {
		this.opts = opts;
	}
	/** Spawn the subprocess and perform the initialize handshake. */
	async start() {
		if (this.proc) {
			if (!this.initialized) throw new Error("executa already starting");
			return this.initialized;
		}
		const [bin, ...args] = this.opts.command;
		if (!bin) throw new Error("ExecutaRunner: empty command");
		const env = {
			...process.env,
			...this.opts.env
		};
		env.ANNA_EXECUTA_DEV = "1";
		this.proc = spawn(bin, args, {
			cwd: this.opts.cwd,
			stdio: [
				"pipe",
				"pipe",
				"pipe"
			],
			env
		});
		const stderrSink = this.opts.onStderr ?? ((l) => process.stderr.write(`[executa] ${l}\n`));
		createInterface({ input: this.proc.stderr }).on("line", stderrSink);
		this.proc.on("exit", (code) => {
			this.closed = true;
			const waiters = this.exitWaiters.splice(0);
			for (const w of waiters) w(code);
			const err = new Error(`executa exited (code=${code ?? "null"})`);
			for (const p of this.pending.values()) p.reject(err);
			this.pending.clear();
		});
		this.proc.on("error", (e) => {
			this.closed = true;
			for (const p of this.pending.values()) p.reject(e);
			this.pending.clear();
		});
		createInterface({ input: this.proc.stdout }).on("line", (line) => {
			this.handleLine(line, stderrSink);
		});
		const initTimeout = this.opts.initTimeoutMs ?? 8e3;
		try {
			const raw = await this.callWithTimeout("initialize", HOST_INITIALIZE_PARAMS, initTimeout);
			const ver = String(raw.protocolVersion ?? "1.0");
			this.initialized = {
				protocolVersion: ver === "2.0" ? "2.0" : "1.0",
				raw
			};
		} catch (e) {
			const rpc = e.rpc;
			if (rpc && rpc.code === -32601) this.initialized = {
				protocolVersion: "1.0",
				raw: {}
			};
			else {
				await this.stop();
				throw e;
			}
		}
		return this.initialized;
	}
	/** Send a JSON-RPC request and await the response. */
	call(method, params = {}) {
		if (this.closed || !this.proc) return Promise.reject(new Error("executa not running"));
		const id = this.nextId++;
		const env = {
			jsonrpc: "2.0",
			id,
			method,
			params
		};
		return new Promise((resolve, reject) => {
			this.pending.set(id, {
				resolve: (v) => resolve(v),
				reject
			});
			this.proc.stdin.write(`${JSON.stringify(env)}\n`);
		});
	}
	/** Convenience: `describe → MANIFEST`. */
	describe() {
		return this.call("describe");
	}
	/** Convenience: `invoke {tool, arguments}`. */
	invoke(tool, args) {
		return this.call("invoke", {
			tool,
			arguments: args
		});
	}
	/** Convenience: `health → {status}`. */
	health() {
		return this.call("health");
	}
	/** Close stdin (graceful EOF shutdown) then SIGTERM after a grace period. */
	async stop(graceMs = 1500) {
		if (!this.proc) return;
		this.closed = true;
		const p = this.proc;
		try {
			p.stdin.end();
		} catch {}
		const waitExit = new Promise((resolve) => {
			this.exitWaiters.push(resolve);
			if (p.exitCode != null) resolve(p.exitCode);
		});
		const timer = new Promise((resolve) => {
			setTimeout(() => resolve("timeout"), graceMs);
		});
		const winner = await Promise.race([waitExit, timer]);
		if (winner === "timeout") try {
			p.kill("SIGTERM");
		} catch {}
		this.proc = null;
	}
	async callWithTimeout(method, params, ms) {
		const id = this.nextId++;
		const env = {
			jsonrpc: "2.0",
			id,
			method,
			params
		};
		return new Promise((resolve, reject) => {
			const to = setTimeout(() => {
				this.pending.delete(id);
				reject(new Error(`executa ${method} timed out after ${ms}ms`));
			}, ms);
			this.pending.set(id, {
				resolve: (v) => {
					clearTimeout(to);
					resolve(v);
				},
				reject: (e) => {
					clearTimeout(to);
					reject(e);
				}
			});
			this.proc.stdin.write(`${JSON.stringify(env)}\n`);
		});
	}
	handleLine(line, stderrSink) {
		if (!line.trim()) return;
		let env;
		try {
			env = JSON.parse(line);
		} catch {
			stderrSink(`non-json from executa stdout: ${line}`);
			return;
		}
		if (typeof env.method === "string") {
			this.handleReverse(env);
			return;
		}
		const id = env.id;
		if (typeof id !== "number") return;
		const pending = this.pending.get(id);
		if (!pending) return;
		this.pending.delete(id);
		if (env.error) {
			const e = env.error;
			const err = new Error(`[${e.code}] ${e.message}`);
			err.rpc = e;
			pending.reject(err);
		} else pending.resolve(env.result);
	}
	async handleReverse(env) {
		const id = env.id;
		const method = env.method;
		const params = env.params ?? {};
		const respond = (body) => {
			if (id === void 0 || id === null) return;
			if (!this.proc) return;
			const out = {
				jsonrpc: "2.0",
				id,
				...body
			};
			try {
				this.proc.stdin.write(`${JSON.stringify(out)}\n`);
			} catch {}
		};
		if (method === "sampling/createMessage") {
			if (!this.opts.sampling) {
				respond({ error: {
					code: -32008,
					message: "sampling not configured (use `anna-app login` and rerun without `--no-sampling`, or pass `--mock-sampling`)"
				} });
				return;
			}
			try {
				const result = await this.opts.sampling.createMessage(params);
				respond({ result });
			} catch (e) {
				const code = e.rpcCode ?? -32e3;
				respond({ error: {
					code,
					message: e.message
				} });
			}
			return;
		}
		if (method.startsWith("agent/")) {
			if (!this.opts.agent) {
				respond({ error: {
					code: -32041,
					message: "agent reverse-RPC not configured (rerun without `--no-agent`, or pass `--mock-agent <fixture>`)"
				} });
				return;
			}
			try {
				const result = await this.opts.agent.call(method, params);
				respond({ result });
			} catch (e) {
				const code = e.rpcCode ?? -32e3;
				respond({ error: {
					code,
					message: e.message
				} });
			}
			return;
		}
		if (method === "image/generate" || method === "image/edit") {
			if (!this.opts.image) {
				respond({ error: {
					code: -32101,
					message: "image not configured (rerun without `--no-image`, or pass `--mock-image <fixture>` / `--app-slug`)"
				} });
				return;
			}
			try {
				const result = method === "image/generate" ? await this.opts.image.generate(params) : await this.opts.image.edit(params);
				respond({ result });
			} catch (e) {
				const code = e.rpcCode ?? -32103;
				respond({ error: {
					code,
					message: e.message
				} });
			}
			return;
		}
		if (method === "host/uploadFile") {
			if (!this.opts.hostUpload) {
				respond({ error: {
					code: -32201,
					message: "host upload not configured (rerun without `--no-upload`, or pass `--mock-upload <fixture>` / `--app-slug`)"
				} });
				return;
			}
			try {
				const result = await this.opts.hostUpload.call(method, params);
				respond({ result });
			} catch (e) {
				const code = e.rpcCode ?? -32207;
				respond({ error: {
					code,
					message: e.message
				} });
			}
			return;
		}
		if (method.startsWith("storage/") || method.startsWith("files/")) {
			if (!this.opts.storage) {
				respond({ error: {
					code: -32021,
					message: "storage not configured (rerun with `--storage memory|mock|real`)"
				} });
				return;
			}
			try {
				const result = await this.opts.storage.call(method, params);
				respond({ result });
			} catch (e) {
				const code = e.rpcCode ?? -32e3;
				respond({ error: {
					code,
					message: e.message
				} });
			}
			return;
		}
		respond({ error: {
			code: -32601,
			message: `unknown reverse method: ${method}`
		} });
	}
};

//#endregion
export { ExecutaRunner };
