// @anna-ai/cli/dist/server-DtnMLKQS.js
import { canonicalHost, getAccount } from "./credentials-BTv2IfUZ.js";
import { BridgeRequestError } from "./bridge-DXH59CmP.js";
import { dirname, join, normalize, resolve } from "node:path";
import { createRequire } from "node:module";
import { createReadStream, existsSync, readFileSync, statSync, watch } from "node:fs";
import { fileURLToPath } from "node:url";
import { readFile } from "node:fs/promises";
import { createServer } from "node:http";
import { WebSocketServer } from "ws";
import { setTimeout as setTimeout$1 } from "node:timers/promises";

//#region src/harness/llm-bridge.ts
/** APS-mode ``storage`` methods the bridge forwards to ``/api/v1/storage/*``.
*  ``get/set/delete/list`` are the KV family; ``files_*`` are the object/blob
*  (APS Files) family, forwarded from an executa's ``files/*`` reverse-RPC. */
const STORAGE_APS_METHODS = new Set([
	"get",
	"set",
	"delete",
	"list",
	"files_init",
	"files_finalize",
	"files_url",
	"files_list",
	"files_delete"
]);
/** APS Files method → ``[httpMethod, subPath, payloadKind]``. Mirrors
*  `anna-app-cli/src/executa/storage.ts` ROUTING and matrix `storage.py`
*  so the app-harness path round-trips identically to `executa dev`. */
const FILES_ROUTES = {
	files_init: [
		"POST",
		"/files/init",
		"body"
	],
	files_finalize: [
		"POST",
		"/files/finalize",
		"body"
	],
	files_url: [
		"GET",
		"/files/url",
		"query"
	],
	files_list: [
		"GET",
		"/files",
		"query"
	],
	files_delete: [
		"DELETE",
		"/files",
		"query"
	]
};
/** App-direct ``anna.files.*`` method → ``[httpMethod, subPath, payloadKind]``.
*  These are the iframe-facing method names (ns=``"files"``) the browser SDK
*  emits via ``rt.call("files", method, args)``. Production nexus serves them
*  through ``anna_app_rpc_handlers/files.py`` with ``scope="app"`` /
*  ``owner_id=app_id`` (the handler ignores ``args.scope``); we force the same
*  scope here so the harness round-trips identically. Distinct from
*  ``FILES_ROUTES`` above, which is the executa reverse-RPC method naming
*  (``files_init`` …) forwarded with the executa's own (default ``user``) scope. */
const FILES_APP_ROUTES = {
	upload_init: [
		"POST",
		"/files/init",
		"body"
	],
	upload_finalize: [
		"POST",
		"/files/finalize",
		"body"
	],
	download_url: [
		"GET",
		"/files/url",
		"query"
	],
	list: [
		"GET",
		"/files",
		"query"
	],
	delete: [
		"DELETE",
		"/files",
		"query"
	]
};
const VALID_STORAGE_SCOPES = new Set([
	"user",
	"app",
	"tool"
]);
var LlmBridge = class {
	mintedAuto = new Map();
	mintedAgent = new Map();
	/** Single shared storage_token per LlmBridge instance — scope is per
	*  ``(user, app)`` server-side, not per-window, so caching one is
	*  correct. Re-minted automatically before expiry. */
	mintedStorage = null;
	mocks = [];
	streamCounter = 0;
	constructor(opts) {
		this.opts = opts;
		if (opts.mode === "mock" && opts.mockFile) {
			const path = resolve(opts.mockFile);
			if (existsSync(path)) {
				const lines = readFileSync(path, "utf8").split(/\r?\n/);
				for (const line of lines) {
					if (!line.trim() || line.startsWith("#")) continue;
					try {
						this.mocks.push(JSON.parse(line));
					} catch {}
				}
			}
		}
	}
	/** Returns true iff this bridge handles `(ns, method)` (i.e. the harness
	*  should NOT forward it to the in-process Python dispatcher).
	*
	*  Storage handling depends on ``storageMode``:
	*  - ``legacy`` (default): the bridge does NOT claim storage; the Python
	*    dispatcher implements ``anna.storage.*`` against the in-memory
	*    ``runtime_state`` dict. Pre-APS parity, works fully offline.
	*  - ``aps``: the bridge claims storage and forwards to real nexus APS
	*    via ``/api/v1/storage/*`` with a Bearer ``storage_token``. */
	handles(ns, method) {
		if (ns === "llm" && (method === "complete" || method === "embed")) return true;
		if (ns === "agent" && method.startsWith("session.")) return true;
		if (ns === "image" && (method === "generate" || method === "edit")) return true;
		if (ns === "upload" && (method === "inline" || method === "negotiate" || method === "confirm")) return true;
		if (ns === "storage" && this.opts.storageMode === "aps" && this.opts.mode === "real" && STORAGE_APS_METHODS.has(method)) return true;
		if (ns === "files" && this.opts.storageMode === "aps" && this.opts.mode === "real" && Object.prototype.hasOwnProperty.call(FILES_APP_ROUTES, method)) return true;
		return false;
	}
	/** Back-compat static alias — older code paths still call
	*  ``LlmBridge.handles(ns, method)`` without an instance. Returns the
	*  legacy (no-storage) decision since options aren't available here. */
	static handles(ns, method) {
		if (ns === "llm" && (method === "complete" || method === "embed")) return true;
		if (ns === "agent" && method.startsWith("session.")) return true;
		if (ns === "image" && (method === "generate" || method === "edit")) return true;
		if (ns === "upload" && (method === "inline" || method === "negotiate" || method === "confirm")) return true;
		return false;
	}
	/** Resolve the active account or throw with a friendly message. */
	account() {
		const acc = getAccount(this.opts.account);
		if (!acc) throw new Error("no PAT on disk — run `anna-app login --host <nexus-url>` first (or use `--no-llm` / `--mock-llm <fixture>` to develop offline)");
		if (acc.expires_at && acc.expires_at < Math.floor(Date.now() / 1e3)) throw new Error("PAT expired — run `anna-app login` again");
		return acc;
	}
	/** Mint (or reuse) an `app_session_token` for kind=complete (per window). */
	async mintComplete(windowUuid) {
		const cached = this.mintedAuto.get(windowUuid);
		if (cached && cached.expiresAt - 30 > Math.floor(Date.now() / 1e3)) return cached;
		const acc = this.account();
		const slug = this.requireAppSlug();
		const body = {
			pat: acc.pat,
			kind: "complete",
			app_slug: slug
		};
		const minted = await this.callMint(acc.host, body);
		const ms = {
			appSessionToken: minted.app_session_token,
			appSessionUuid: minted.app_session_uuid,
			expiresAt: Math.floor(Date.now() / 1e3) + (minted.expires_in || 600),
			isAgent: false,
			submode: null
		};
		this.mintedAuto.set(windowUuid, ms);
		return ms;
	}
	/** Mint a kind=agent session — called from `agent.session.create`. */
	async mintAgent(args) {
		const acc = this.account();
		const slug = this.requireAppSlug();
		const submode = args.submode ?? "auto";
		const body = {
			pat: acc.pat,
			kind: "agent",
			submode,
			fixed_client_id: args.fixed_client_id ?? args.fixedClientId ?? null,
			app_slug: slug,
			label: args.label ?? "anna-app dev",
			quota_caps: args.quotaCaps ?? null,
			system_prompt: args.systemPrompt ?? args.system_prompt ?? null
		};
		const minted = await this.callMint(acc.host, body);
		const ms = {
			appSessionToken: minted.app_session_token,
			appSessionUuid: minted.app_session_uuid,
			expiresAt: Math.floor(Date.now() / 1e3) + (minted.expires_in || 600),
			isAgent: true,
			submode
		};
		this.mintedAgent.set(ms.appSessionUuid, ms);
		return ms;
	}
	/** Re-mint an agent session token by ``app_session_uuid`` via the dev
	*  ``/dev/session/refresh`` endpoint (PAT + uuid — NO stale token needed).
	*
	*  This is the harness's token-free recovery path: even after the CLI
	*  process restarted (so ``mintedAgent`` is empty) or the previous token
	*  expired, the underlying ``AnnaAppSession`` row is still alive and the
	*  app remembers its ``app_session_uuid``. ``refresh`` slides the idle
	*  window and hands back a fresh short-lived token, fixing the forum
	*  report "uuid unusable after dev restart". Throws on 404/410 (the row
	*  is genuinely gone / revoked / idle-expired). */
	async callRefresh(apsUuid) {
		const acc = this.account();
		const url = `${canonicalHost(acc.host)}/api/v1/anna-apps/dev/session/refresh`;
		const res = await fetch(url, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify({
				pat: acc.pat,
				app_session_uuid: apsUuid
			})
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw new Error(`session.refresh failed: HTTP ${res.status} ${text}`);
		}
		const payload = await res.json();
		return {
			token: payload.app_session_token,
			expiresIn: payload.expires_in || 600,
			submode: payload.submode ?? null
		};
	}
	/** Revoke an agent session by uuid via ``/dev/session/revoke`` (PAT-auth).
	*
	*  Token-free: works even when the session token is long expired (which
	*  used to make the session impossible to release → quota leak). */
	async callRevoke(apsUuid) {
		const acc = this.account();
		const url = `${canonicalHost(acc.host)}/api/v1/anna-apps/dev/session/revoke`;
		const res = await fetch(url, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify({
				pat: acc.pat,
				app_session_uuid: apsUuid
			})
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw new Error(`session.revoke failed: HTTP ${res.status} ${text}`);
		}
	}
	/** Return a live agent session token for ``apsUuid``, re-minting via
	*  ``/dev/session/refresh`` on cache-miss or near-expiry.
	*
	*  Centralises the "is my token still good?" decision for every operation
	*  that needs an ``app_session_token`` (run / cancel / delete). On success
	*  the freshly-minted session is cached in ``mintedAgent`` so subsequent
	*  calls in the same process reuse it until the next near-expiry check. */
	async ensureLiveToken(apsUuid) {
		const now = Math.floor(Date.now() / 1e3);
		const cached = this.mintedAgent.get(apsUuid);
		if (cached && cached.expiresAt - 30 > now) return cached;
		const refreshed = await this.callRefresh(apsUuid);
		const ms = {
			appSessionToken: refreshed.token,
			appSessionUuid: apsUuid,
			expiresAt: Math.floor(Date.now() / 1e3) + refreshed.expiresIn,
			isAgent: true,
			submode: refreshed.submode ?? cached?.submode ?? null
		};
		this.mintedAgent.set(apsUuid, ms);
		return ms;
	}
	/** Mint (or reuse) a storage_token for APS forwarding.
	*
	*  Server-side ``/dev/storage/mint`` issues a JWT scoped to
	*  ``(user, app)`` with ``allowed_scopes ⊆ {user, app, tool}`` and a
	*  configurable TTL (default 600s) + ``max_calls`` budget (default 200).
	*  We re-mint when within 30s of expiry; budget exhaustion currently
	*  surfaces as an HTTP 401/403 which propagates back to the iframe.
	*
	*  Per-app, not per-window: the token doesn't carry a window or
	*  session uuid — storage scopes are ``user``/``app``/``tool``, not
	*  ``window``. Caching one per LlmBridge instance is correct. */
	async mintStorage() {
		const cached = this.mintedStorage;
		if (cached && cached.expiresAt - 30 > Math.floor(Date.now() / 1e3)) return cached;
		const acc = this.account();
		const slug = this.requireAppSlug();
		const body = {
			pat: acc.pat,
			app_slug: slug,
			ttl_seconds: 600,
			max_calls: 200
		};
		const url = `${canonicalHost(acc.host)}/api/v1/anna-apps/dev/storage/mint`;
		const res = await fetch(url, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify(body)
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw new Error(`storage.mint failed: HTTP ${res.status} ${text}`);
		}
		const payload = await res.json();
		const ms = {
			storageToken: payload.storage_token,
			expiresAt: Math.floor(Date.now() / 1e3) + (payload.expires_in || 600),
			allowedScopes: payload.allowed_scopes ?? [
				"user",
				"app",
				"tool"
			]
		};
		this.mintedStorage = ms;
		return ms;
	}
	/** Throws a friendly error if no appSlug was wired into the bridge. */
	requireAppSlug() {
		if (!this.opts.appSlug) throw new Error("llm bridge has no app_slug — the harness must register the dev app via POST /api/v1/anna-apps/dev/apps/register before minting sessions. (Run `anna-app login` once, then `anna-app dev` will auto-register the manifest slug.)");
		return this.opts.appSlug;
	}
	async callMint(host, body) {
		return this._callMintOnce(
			host,
			body,
			/*retryOnAppNotFound=*/
			true
);
	}
	async _callMintOnce(host, body, retryOnAppNotFound) {
		const url = `${canonicalHost(host)}/api/v1/anna-apps/dev/session/mint`;
		const res = await fetch(url, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify(body)
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			if (retryOnAppNotFound && res.status === 404 && /app slug .* not found/.test(text) && typeof this.opts.onAppSlugNotFound === "function") {
				try {
					await this.opts.onAppSlugNotFound();
				} catch (hookErr) {
					throw new Error(`session.mint failed: HTTP 404 ${text}\n  recovery hook failed: ${hookErr.message}`);
				}
				return this._callMintOnce(
					host,
					body,
					/*retryOnAppNotFound=*/
					false
);
			}
			throw new Error(`session.mint failed: HTTP ${res.status} ${text}`);
		}
		return await res.json();
	}
	/** Public entry — invoked by harness `proxyCall` for llm.* / agent.*. */
	async dispatch(args) {
		if (this.opts.mode === "off") return {
			ok: false,
			error: {
				code: "llm_disabled",
				message: "harness started with --no-llm"
			}
		};
		if (this.opts.mode === "mock") return this.dispatchMock(args);
		try {
			return await this.dispatchReal(args);
		} catch (e) {
			return {
				ok: false,
				error: {
					code: "transport",
					message: e.message
				}
			};
		}
	}
	async dispatchMock(args) {
		const content = String(args.args?.content ?? args.args?.messages ?? "");
		const entry = this.mocks.find((m) => m.ns === args.ns && m.method === args.method && (!m.match?.contentIncludes || content.includes(m.match.contentIncludes))) ?? this.mocks.find((m) => m.ns === args.ns && m.method === args.method);
		if (!entry) return {
			ok: true,
			result: args.ns === "llm" ? {
				role: "assistant",
				content: {
					type: "text",
					text: "(mock) no fixture matched"
				},
				model: "mock-model",
				stopReason: "endTurn"
			} : {
				app_session_uuid: "aps_mock",
				expires_in: 600,
				submode: "auto",
				granted_tools: []
			}
		};
		if (entry.events && entry.events.length > 0) {
			const sid = `mock-${++this.streamCounter}`;
			(async () => {
				let seq = 0;
				for (const ev of entry.events ?? []) {
					if (ev.delay_ms) await setTimeout$1(ev.delay_ms);
					seq += 1;
					args.onEvent("rpc.stream", {
						stream_id: sid,
						seq,
						payload: ev.payload,
						done: false
					});
				}
				args.onEvent("rpc.stream", {
					stream_id: sid,
					seq: seq + 1,
					payload: { event: "end" },
					done: true
				});
			})();
			return {
				ok: true,
				result: {
					stream_id: sid,
					run_id: "mock-run"
				}
			};
		}
		return {
			ok: true,
			result: entry.result ?? {}
		};
	}
	async dispatchReal(args) {
		const acc = this.account();
		if (args.ns === "llm" && args.method === "complete") {
			const ms = await this.mintComplete(args.windowUuid);
			const result = await this.postJson(`${canonicalHost(acc.host)}/api/v1/copilot/app/complete`, ms.appSessionToken, args.args);
			return {
				ok: true,
				result
			};
		}
		if (args.ns === "llm" && args.method === "embed") {
			const ms = await this.mintComplete(args.windowUuid);
			const result = await this.postJson(`${canonicalHost(acc.host)}/api/v1/copilot/app/embed`, ms.appSessionToken, args.args);
			return {
				ok: true,
				result
			};
		}
		if (args.ns === "agent") switch (args.method) {
			case "session.create": {
				const ms = await this.mintAgent(args.args);
				return {
					ok: true,
					result: {
						app_session_uuid: ms.appSessionUuid,
						expires_in: Math.max(0, ms.expiresAt - Math.floor(Date.now() / 1e3)),
						submode: ms.submode,
						fixed_client_id: args.args.fixed_client_id ?? null,
						granted_tools: []
					}
				};
			}
			case "session.run": {
				const apsUuid = String(args.args.app_session_uuid ?? "");
				let ms;
				try {
					ms = await this.ensureLiveToken(apsUuid);
				} catch (e) {
					return {
						ok: false,
						error: {
							code: "session_expired",
							message: e.message || "session expired or revoked"
						}
					};
				}
				const sid = `dev-${++this.streamCounter}`;
				this.pumpAgentRun({
					host: acc.host,
					token: ms.appSessionToken,
					body: {
						...args.args,
						stream: true
					},
					streamId: sid,
					windowUuid: args.windowUuid,
					onEvent: args.onEvent
				});
				return {
					ok: true,
					result: {
						stream_id: sid,
						run_id: String(args.args.run_id ?? sid)
					}
				};
			}
			case "session.refresh": {
				const apsUuid = String(args.args.app_session_uuid ?? "");
				try {
					const ms = await this.ensureLiveToken(apsUuid);
					return {
						ok: true,
						result: {
							app_session_uuid: ms.appSessionUuid,
							expires_in: Math.max(0, ms.expiresAt - Math.floor(Date.now() / 1e3)),
							submode: ms.submode
						}
					};
				} catch (e) {
					return {
						ok: false,
						error: {
							code: "session_expired",
							message: e.message || "session expired or revoked"
						}
					};
				}
			}
			case "session.cancel": {
				const apsUuid = String(args.args.app_session_uuid ?? "");
				let ms;
				try {
					ms = await this.ensureLiveToken(apsUuid);
				} catch (e) {
					return {
						ok: false,
						error: {
							code: "session_expired",
							message: e.message || "session expired or revoked"
						}
					};
				}
				const out = await this.postJson(`${canonicalHost(acc.host)}/api/v1/copilot/app/agent/cancel`, ms.appSessionToken, args.args);
				return {
					ok: true,
					result: out
				};
			}
			case "session.history": return {
				ok: false,
				error: {
					code: "not_supported",
					message: "agent.session.history is not exposed over HTTP; available only via in-process store"
				}
			};
			case "session.list": {
				const includeExpired = Boolean(args.args?.include_expired);
				const rawLimit = Number(args.args?.limit ?? 50);
				const limit = Math.max(1, Math.min(100, Number.isFinite(rawLimit) ? rawLimit : 50));
				if (this.opts.appSlug) try {
					const acc2 = this.account();
					const url = new URL(`${canonicalHost(acc2.host)}/api/v1/anna-apps/dev/sessions`);
					url.searchParams.set("pat", acc2.pat);
					url.searchParams.set("app_slug", this.opts.appSlug);
					url.searchParams.set("include_expired", String(includeExpired));
					url.searchParams.set("limit", String(limit));
					const res = await fetch(url, { method: "GET" });
					if (res.ok) {
						const body = await res.json();
						return {
							ok: true,
							result: { sessions: body.sessions ?? [] }
						};
					}
				} catch {}
				const now = Math.floor(Date.now() / 1e3);
				const sessions = [...this.mintedAgent.values()].filter((ms) => includeExpired || ms.expiresAt > now).slice(0, limit).map((ms) => ({
					app_session_uuid: ms.appSessionUuid,
					kind: "agent",
					submode: ms.submode,
					fixed_client_id: null,
					label: "anna-app dev",
					created_at: null,
					last_active_at: null,
					expires_at: new Date(ms.expiresAt * 1e3).toISOString()
				}));
				return {
					ok: true,
					result: { sessions }
				};
			}
			case "session.delete": {
				const apsUuid = String(args.args.app_session_uuid ?? "");
				const cached = this.mintedAgent.get(apsUuid);
				this.mintedAgent.delete(apsUuid);
				if (cached && cached.expiresAt - 5 > Math.floor(Date.now() / 1e3)) {
					const url = `${canonicalHost(acc.host)}/api/v1/copilot/app/sessions/${encodeURIComponent(apsUuid)}`;
					const res = await fetch(url, {
						method: "DELETE",
						headers: { authorization: `Bearer ${cached.appSessionToken}` }
					});
					if (res.ok) return {
						ok: true,
						result: { deleted: true }
					};
				}
				try {
					await this.callRevoke(apsUuid);
					return {
						ok: true,
						result: { deleted: true }
					};
				} catch (e) {
					return {
						ok: false,
						error: {
							code: "transport",
							message: e.message
						}
					};
				}
			}
		}
		if (args.ns === "image" && (args.method === "generate" || args.method === "edit")) {
			const ms = await this.mintComplete(args.windowUuid);
			const result = await this.postJson(`${canonicalHost(acc.host)}/api/v1/copilot/app/image/${args.method}`, ms.appSessionToken, args.args);
			return {
				ok: true,
				result
			};
		}
		if (args.ns === "upload") {
			const ms = await this.mintComplete(args.windowUuid);
			const path = args.method === "inline" ? "/api/v1/copilot/app/upload" : args.method === "negotiate" ? "/api/v1/copilot/app/upload/negotiate" : args.method === "confirm" ? "/api/v1/copilot/app/upload/confirm" : null;
			if (path != null) {
				const result = await this.postJson(`${canonicalHost(acc.host)}${path}`, ms.appSessionToken, args.args);
				return {
					ok: true,
					result
				};
			}
		}
		if (args.ns === "storage") {
			const sm = await this.mintStorage();
			const a = args.args;
			const base = `${canonicalHost(acc.host)}/api/v1/storage`;
			if (args.method === "get") {
				const qs = new URLSearchParams();
				qs.set("key", String(a.key ?? ""));
				if (a.scope) qs.set("scope", String(a.scope));
				const res = await fetch(`${base}/kv?${qs.toString()}`, {
					method: "GET",
					headers: { authorization: `Bearer ${sm.storageToken}` }
				});
				if (res.status === 404) return {
					ok: true,
					result: {
						value: null,
						exists: false
					}
				};
				if (!res.ok) {
					const text = await res.text().catch(() => "");
					throw new Error(`storage.get HTTP ${res.status}: ${text}`);
				}
				const entry = await res.json();
				if (entry.exists === void 0) entry.exists = true;
				return {
					ok: true,
					result: entry
				};
			}
			if (args.method === "set") {
				const qs = new URLSearchParams();
				if (a.scope) qs.set("scope", String(a.scope));
				const body = {
					key: a.key,
					value: a.value
				};
				if (a.if_match !== void 0) body.if_match = a.if_match;
				if (a.metadata !== void 0) body.metadata = a.metadata;
				if (a.tags !== void 0) body.tags = a.tags;
				if (a.ttl_seconds !== void 0) body.ttl_seconds = a.ttl_seconds;
				const res = await fetch(`${base}/kv?${qs.toString()}`, {
					method: "PUT",
					headers: {
						"content-type": "application/json",
						authorization: `Bearer ${sm.storageToken}`
					},
					body: JSON.stringify(body)
				});
				if (!res.ok) {
					const text = await res.text().catch(() => "");
					throw new Error(`storage.set HTTP ${res.status}: ${text}`);
				}
				return {
					ok: true,
					result: await res.json()
				};
			}
			if (args.method === "delete") {
				const qs = new URLSearchParams();
				qs.set("key", String(a.key ?? ""));
				if (a.if_match !== void 0) qs.set("if_match", String(a.if_match));
				if (a.scope) qs.set("scope", String(a.scope));
				const res = await fetch(`${base}/kv?${qs.toString()}`, {
					method: "DELETE",
					headers: { authorization: `Bearer ${sm.storageToken}` }
				});
				if (!res.ok) {
					const text = await res.text().catch(() => "");
					throw new Error(`storage.delete HTTP ${res.status}: ${text}`);
				}
				return {
					ok: true,
					result: await res.json()
				};
			}
			if (args.method === "list") {
				const qs = new URLSearchParams();
				if (a.prefix !== void 0) qs.set("prefix", String(a.prefix));
				if (a.cursor !== void 0) qs.set("cursor", String(a.cursor));
				if (a.limit !== void 0) qs.set("limit", String(a.limit));
				if (a.scope) qs.set("scope", String(a.scope));
				const res = await fetch(`${base}/list?${qs.toString()}`, {
					method: "GET",
					headers: { authorization: `Bearer ${sm.storageToken}` }
				});
				if (!res.ok) {
					const text = await res.text().catch(() => "");
					throw new Error(`storage.list HTTP ${res.status}: ${text}`);
				}
				return {
					ok: true,
					result: await res.json()
				};
			}
			const froute = FILES_ROUTES[args.method];
			if (froute) {
				const [httpMethod, subPath, payloadKind] = froute;
				const scope = typeof a.scope === "string" && VALID_STORAGE_SCOPES.has(a.scope) ? a.scope : "user";
				const url = new URL(`${base}${subPath}`);
				url.searchParams.set("scope", scope);
				let jsonBody;
				if (payloadKind === "query") for (const [k, v] of Object.entries(a)) {
					if (k === "scope" || v === void 0 || v === null) continue;
					url.searchParams.set(k, typeof v === "string" ? v : JSON.stringify(v));
				}
				else {
					const inner = {};
					for (const [k, v] of Object.entries(a)) {
						if (k === "scope" || v === void 0) continue;
						inner[k] = v;
					}
					jsonBody = JSON.stringify(inner);
				}
				const headers = { authorization: `Bearer ${sm.storageToken}` };
				if (jsonBody !== void 0) headers["content-type"] = "application/json";
				const res = await fetch(url, {
					method: httpMethod,
					headers,
					body: jsonBody
				});
				if (!res.ok) {
					const text = await res.text().catch(() => "");
					throw new Error(`storage.${args.method} HTTP ${res.status}: ${text}`);
				}
				const ct = res.headers.get("content-type") ?? "";
				if (args.method === "files_delete" && (res.status === 204 || !ct.includes("json"))) return {
					ok: true,
					result: { deleted: true }
				};
				return {
					ok: true,
					result: await res.json()
				};
			}
		}
		if (args.ns === "files") {
			const froute = FILES_APP_ROUTES[args.method];
			if (froute) {
				const sm = await this.mintStorage();
				const a = { ...args.args };
				if (args.method === "upload_init" && a.size !== void 0 && a.size_bytes === void 0) {
					a.size_bytes = a.size;
					delete a.size;
				}
				const base = `${canonicalHost(this.account().host)}/api/v1/storage`;
				const [httpMethod, subPath, payloadKind] = froute;
				const url = new URL(`${base}${subPath}`);
				url.searchParams.set("scope", "app");
				let jsonBody;
				if (payloadKind === "query") for (const [k, v] of Object.entries(a)) {
					if (k === "scope" || v === void 0 || v === null) continue;
					url.searchParams.set(k, typeof v === "string" ? v : JSON.stringify(v));
				}
				else {
					const inner = {};
					for (const [k, v] of Object.entries(a)) {
						if (k === "scope" || v === void 0) continue;
						inner[k] = v;
					}
					jsonBody = JSON.stringify(inner);
				}
				const headers = { authorization: `Bearer ${sm.storageToken}` };
				if (jsonBody !== void 0) headers["content-type"] = "application/json";
				const res = await fetch(url, {
					method: httpMethod,
					headers,
					body: jsonBody
				});
				if (!res.ok) {
					const text = await res.text().catch(() => "");
					throw new Error(`files.${args.method} HTTP ${res.status}: ${text}`);
				}
				const ct = res.headers.get("content-type") ?? "";
				if (args.method === "delete" && (res.status === 204 || !ct.includes("json"))) return {
					ok: true,
					result: { deleted: true }
				};
				return {
					ok: true,
					result: await res.json()
				};
			}
		}
		return {
			ok: false,
			error: {
				code: "unknown_method",
				message: `${args.ns}.${args.method}`
			}
		};
	}
	async postJson(url, token, body) {
		const res = await fetch(url, {
			method: "POST",
			headers: {
				"content-type": "application/json",
				authorization: `Bearer ${token}`
			},
			body: JSON.stringify(body)
		});
		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw new Error(`HTTP ${res.status}: ${text}`);
		}
		return res.json();
	}
	/** Consume an SSE response and forward each frame as `rpc.stream`. */
	async pumpAgentRun(args) {
		const url = `${canonicalHost(args.host)}/api/v1/copilot/app/agent`;
		let seq = 0;
		const emit = (payload, done) => {
			seq += 1;
			args.onEvent("rpc.stream", {
				stream_id: args.streamId,
				window_uuid: args.windowUuid,
				seq,
				payload,
				done
			});
		};
		try {
			const res = await fetch(url, {
				method: "POST",
				headers: {
					"content-type": "application/json",
					authorization: `Bearer ${args.token}`,
					accept: "text/event-stream"
				},
				body: JSON.stringify(args.body)
			});
			if (!res.ok || !res.body) {
				const text = await res.text().catch(() => "");
				emit({
					event: "error",
					code: "http",
					message: `HTTP ${res.status}: ${text}`
				}, true);
				return;
			}
			const reader = res.body.getReader();
			const decoder = new TextDecoder();
			let buf = "";
			while (true) {
				const { value, done } = await reader.read();
				if (done) break;
				buf += decoder.decode(value, { stream: true });
				let idx;
				while ((idx = buf.indexOf("\n\n")) >= 0) {
					const frame = buf.slice(0, idx);
					buf = buf.slice(idx + 2);
					const payload = parseSseFrame(frame);
					if (payload != null) emit(payload, false);
				}
			}
			if (buf.trim().length > 0) {
				const payload = parseSseFrame(buf);
				if (payload != null) emit(payload, false);
			}
			emit({ event: "end" }, true);
		} catch (e) {
			emit({
				event: "error",
				code: "transport",
				message: e.message
			}, true);
		}
	}
};
function parseSseFrame(frame) {
	const lines = frame.split(/\r?\n/);
	let dataLines = [];
	let eventName = null;
	for (const raw of lines) {
		if (!raw || raw.startsWith(":")) continue;
		if (raw.startsWith("data:")) dataLines.push(raw.slice(5).trimStart());
		else if (raw.startsWith("event:")) eventName = raw.slice(6).trim();
	}
	if (dataLines.length === 0) return null;
	const data = dataLines.join("\n");
	try {
		const obj = JSON.parse(data);
		if (eventName && obj && typeof obj === "object" && !("event" in obj)) return {
			...obj,
			event: eventName
		};
		return obj;
	} catch {
		return {
			event: eventName ?? "raw",
			text: data
		};
	}
}

//#endregion
//#region src/harness/server.ts
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const MIME = {
	".html": "text/html; charset=utf-8",
	".js": "application/javascript; charset=utf-8",
	".mjs": "application/javascript; charset=utf-8",
	".css": "text/css; charset=utf-8",
	".json": "application/json; charset=utf-8",
	".svg": "image/svg+xml",
	".png": "image/png",
	".jpg": "image/jpeg",
	".jpeg": "image/jpeg",
	".gif": "image/gif",
	".woff": "font/woff",
	".woff2": "font/woff2",
	".map": "application/json"
};
var HarnessServer = class {
	server = createServer((req, res) => this.handle(req, res));
	wss = null;
	sessionId = null;
	liveSockets = new Set();
	watchers = [];
	reloadDebounce = null;
	/** Pending events queued by the LLM bridge — drained alongside Python events. */
	llmEventQueue = [];
	llmBridge;
	constructor(cfg, bridge) {
		this.cfg = cfg;
		this.bridge = bridge;
		this.llmBridge = cfg.llm ? new LlmBridge(cfg.llm) : null;
		const HOST_OUTBOUND_ROUTES = [
			[
				"host.llm.complete",
				"llm",
				"complete"
			],
			[
				"host.llm.embed",
				"llm",
				"embed"
			],
			[
				"host.agent.session.create",
				"agent",
				"session.create"
			],
			[
				"host.agent.session.run",
				"agent",
				"session.run"
			],
			[
				"host.agent.session.cancel",
				"agent",
				"session.cancel"
			],
			[
				"host.agent.session.history",
				"agent",
				"session.history"
			],
			[
				"host.agent.session.refresh",
				"agent",
				"session.refresh"
			],
			[
				"host.agent.session.delete",
				"agent",
				"session.delete"
			],
			[
				"host.agent.session.list",
				"agent",
				"session.list"
			],
			[
				"host.image.generate",
				"image",
				"generate"
			],
			[
				"host.image.edit",
				"image",
				"edit"
			],
			[
				"host.upload.inline",
				"upload",
				"inline"
			],
			[
				"host.upload.negotiate",
				"upload",
				"negotiate"
			],
			[
				"host.upload.confirm",
				"upload",
				"confirm"
			],
			[
				"host.storage.get",
				"storage",
				"get"
			],
			[
				"host.storage.set",
				"storage",
				"set"
			],
			[
				"host.storage.delete",
				"storage",
				"delete"
			],
			[
				"host.storage.list",
				"storage",
				"list"
			],
			[
				"host.storage.files_init",
				"storage",
				"files_init"
			],
			[
				"host.storage.files_finalize",
				"storage",
				"files_finalize"
			],
			[
				"host.storage.files_url",
				"storage",
				"files_url"
			],
			[
				"host.storage.files_list",
				"storage",
				"files_list"
			],
			[
				"host.storage.files_delete",
				"storage",
				"files_delete"
			]
		];
		for (const [hostMethod, ns, dispatchMethod] of HOST_OUTBOUND_ROUTES) this.bridge.onRequest(hostMethod, async (params) => {
			if (this.llmBridge == null) throw new BridgeRequestError("llm_disabled", "harness started without an LlmBridge (use --no-llm to suppress this path or `anna-app login` for a real bridge)");
			const out = await this.llmBridge.dispatch({
				windowUuid: this.sessionId ?? "harness",
				ns,
				method: dispatchMethod,
				args: params,
				onEvent: (kind, payload) => {
					this.llmEventQueue.push({
						event: kind,
						payload
					});
				}
			});
			if (out.ok) return out.result;
			throw new BridgeRequestError(out.error.code, out.error.message);
		});
	}
	async listen() {
		if (this.cfg.executas && this.cfg.executas.length > 0) await this.bridge.call("executas.register", { executas: this.cfg.executas.map((e) => ({
			tool_id: e.tool_id,
			project_dir: e.project_dir,
			command: e.command ?? null
		})) });
		await new Promise((res, rej) => this.server.listen(this.cfg.port, () => res()).once("error", rej));
		this.wss = new WebSocketServer({ noServer: true });
		this.server.on("upgrade", (req, socket, head) => {
			const url = new URL(req.url ?? "/", "http://localhost");
			if (url.pathname !== "/ws") {
				socket.destroy();
				return;
			}
			this.wss.handleUpgrade(req, socket, head, (ws) => {
				const sid = url.searchParams.get("session_id");
				if (!sid || sid !== this.sessionId) {
					ws.close(1008, "unknown session_id");
					return;
				}
				this.liveSockets.add(ws);
				const timer = setInterval(async () => {
					try {
						const out = await this.bridge.call("session.drain_events", { session_id: sid });
						for (const ev of out.events) ws.send(JSON.stringify({
							kind: "event",
							...ev
						}));
						if (this.llmEventQueue.length > 0) {
							const drained = this.llmEventQueue.splice(0);
							for (const ev of drained) ws.send(JSON.stringify({
								kind: "event",
								...ev
							}));
						}
					} catch (e) {
						ws.close(1011, `drain failed: ${e.message}`);
						clearInterval(timer);
					}
				}, 200);
				ws.on("close", () => {
					this.liveSockets.delete(ws);
					clearInterval(timer);
				});
			});
		});
		if (this.cfg.watch) this.startWatcher();
	}
	async close() {
		for (const w of this.watchers) w.close();
		this.watchers = [];
		if (this.reloadDebounce) clearTimeout(this.reloadDebounce);
		if (this.wss) {
			for (const c of this.wss.clients) c.terminate();
			this.wss.close();
		}
		await new Promise((res) => this.server.close(() => res()));
	}
	startWatcher() {
		const broadcastReload = (path) => {
			if (this.reloadDebounce) clearTimeout(this.reloadDebounce);
			this.reloadDebounce = setTimeout(() => {
				const env = JSON.stringify({
					kind: "reload",
					path
				});
				for (const ws of this.liveSockets) if (ws.readyState === ws.OPEN) ws.send(env);
			}, 100);
		};
		try {
			this.watchers.push(watch(this.cfg.bundleDir, { recursive: true }, (_evt, filename) => {
				if (filename) broadcastReload(`bundle/${filename}`);
			}));
		} catch (e) {
			process.stderr.write(`[harness] watcher failed to attach to ${this.cfg.bundleDir}: ${e.message}\n`);
		}
	}
	async handle(req, res) {
		try {
			const url = new URL(req.url ?? "/", "http://localhost");
			const method = req.method ?? "GET";
			if (method === "GET" && (url.pathname === "/" || url.pathname === "/dashboard")) return await this.serveDashboard(res);
			if (method === "GET" && url.pathname === "/api/config") return this.json(res, 200, {
				app_slug: this.cfg.slug,
				view: this.cfg.view ?? null,
				bundle_base: `/anna-apps/${this.cfg.slug}/dev/${this.cfg.bundleEntry}`,
				executas: (this.cfg.executas ?? []).map((e) => e.tool_id),
				watch: !!this.cfg.watch
			});
			if (method === "POST" && url.pathname === "/api/session/create") return await this.createSession(res);
			if (method === "POST" && url.pathname === "/api/session/call") return await this.proxyCall(req, res);
			if (method === "POST" && url.pathname === "/api/session/refresh-token") return await this.refreshToken(res);
			if (method === "GET" && url.pathname.startsWith("/static/anna-apps/_sdk/")) return await this.serveSdk(url.pathname, res);
			if (method === "GET" && url.pathname.startsWith(`/anna-apps/${this.cfg.slug}/dev/`)) {
				const rel = url.pathname.replace(`/anna-apps/${this.cfg.slug}/dev/`, "");
				return await this.serveBundleAsset(rel, res);
			}
			this.text(res, 404, `not found: ${url.pathname}`);
		} catch (e) {
			this.text(res, 500, `harness error: ${e.message}`);
		}
	}
	async serveDashboard(res) {
		const file = join(__dirname, "dashboard.html");
		const html = await readFile(file, "utf-8");
		res.writeHead(200, {
			"content-type": MIME[".html"],
			"cache-control": "no-store"
		});
		res.end(html);
	}
	async createSession(res) {
		if (this.sessionId) {
			try {
				await this.bridge.call("session.close", { session_id: this.sessionId });
			} catch {}
			this.sessionId = null;
		}
		const out = await this.bridge.call("session.create", {
			user_id: this.cfg.userId,
			manifest: this.cfg.manifest,
			view: this.cfg.view,
			entry_payload: this.cfg.entryPayload ?? {},
			app_slug: this.cfg.slug
		});
		this.sessionId = out.session_id;
		this.json(res, 200, out);
	}
	async proxyCall(req, res) {
		const body = await readBody(req);
		let parsed;
		try {
			parsed = JSON.parse(body);
		} catch {
			return this.json(res, 400, {
				ok: false,
				error: {
					code: "bad_request",
					message: "invalid json body"
				}
			});
		}
		if (this.llmBridge != null && this.llmBridge.handles(parsed.ns, parsed.method)) {
			const out = await this.llmBridge.dispatch({
				windowUuid: this.sessionId ?? "harness",
				ns: parsed.ns,
				method: parsed.method,
				args: parsed.args ?? {},
				onEvent: (kind, payload) => {
					this.llmEventQueue.push({
						event: kind,
						payload
					});
				}
			});
			this.json(res, 200, out);
			return;
		}
		try {
			const out = await this.bridge.call("session.call", {
				session_id: parsed.session_id,
				ns: parsed.ns,
				method: parsed.method,
				args: parsed.args ?? {}
			});
			this.json(res, 200, out);
		} catch (e) {
			this.json(res, 200, {
				ok: false,
				error: {
					code: "transport",
					message: e.message
				}
			});
		}
	}
	async refreshToken(res) {
		if (!this.sessionId) return this.json(res, 400, { error: "no active session" });
		try {
			const out = await this.bridge.call("session.refresh_token", { session_id: this.sessionId });
			this.json(res, 200, out);
		} catch (e) {
			this.json(res, 500, { error: e.message });
		}
	}
	async serveSdk(pathname, res) {
		const sdkRel = pathname.replace(/^\/static\/anna-apps\/_sdk\/[^/]+\//, "");
		const distRoot = this.resolveSdkDistRoot();
		if (!distRoot) return this.text(res, 500, "@anna-ai/app-runtime is not installed and no in-tree packages/anna-app-runtime/dist was found.");
		const abs = resolve(distRoot, sdkRel);
		if (!abs.startsWith(distRoot)) return this.text(res, 403, "forbidden");
		return this.serveFile(abs, res);
	}
	/**
	* Resolve the directory the browser SDK (`index.js` + friends) is served
	* from. When `--matrix-nexus-root` is set we prefer the in-tree
	* `packages/anna-app-runtime/dist/` so contributors run byte-identical
	* with the unpublished workspace SDK (parity with the Python bridge, which
	* already runs in-tree in `nexus-source` mode). Falls back to the
	* published npm package resolved from anna-app-cli's own node_modules.
	* Returns null if neither is available.
	*/
	resolveSdkDistRoot() {
		const root = this.cfg.matrixNexusRoot;
		if (root) {
			const inTree = resolve(root, "packages", "anna-app-runtime", "dist");
			if (existsSync(join(inTree, "index.js"))) return inTree;
		}
		try {
			const req = createRequire(import.meta.url);
			return dirname(req.resolve("@anna-ai/app-runtime"));
		} catch {
			return null;
		}
	}
	async serveBundleAsset(rel, res) {
		const abs = resolve(this.cfg.bundleDir, normalize(rel));
		if (!abs.startsWith(resolve(this.cfg.bundleDir))) return this.text(res, 403, "forbidden");
		return this.serveFile(abs, res);
	}
	async serveFile(abs, res) {
		let stat;
		try {
			stat = statSync(abs);
		} catch {
			return this.text(res, 404, `not found: ${abs}`);
		}
		if (!stat.isFile()) return this.text(res, 404, "not a file");
		const ext = abs.slice(abs.lastIndexOf("."));
		res.writeHead(200, {
			"content-type": MIME[ext] ?? "application/octet-stream",
			"cache-control": "no-store",
			"content-length": String(stat.size)
		});
		createReadStream(abs).pipe(res);
	}
	json(res, status, body) {
		const text = JSON.stringify(body);
		res.writeHead(status, {
			"content-type": MIME[".json"],
			"content-length": String(Buffer.byteLength(text)),
			"cache-control": "no-store"
		});
		res.end(text);
	}
	text(res, status, body) {
		res.writeHead(status, {
			"content-type": "text/plain; charset=utf-8",
			"content-length": String(Buffer.byteLength(body))
		});
		res.end(body);
	}
};
function readBody(req) {
	return new Promise((resolve$1, reject) => {
		const chunks = [];
		req.on("data", (c) => chunks.push(c));
		req.on("end", () => resolve$1(Buffer.concat(chunks).toString("utf-8")));
		req.on("error", reject);
	});
}

//#endregion
export { HarnessServer };
