// @anna-ai/cli/dist/token-hGGfoge2.js
import "./credentials-BTv2IfUZ.js";
import "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { bold, cyan, dim, green, red, yellow } from "kleur/colors";

//#region src/api/tokens.ts
async function listTokens(client, includeRevoked = false) {
	const r = await client.request({
		path: "/api/v1/anna-apps/dev/tokens",
		query: { include_revoked: includeRevoked ? "true" : void 0 }
	});
	return r.data?.tokens ?? [];
}
async function revokeToken(client, id) {
	await client.request({
		method: "POST",
		path: `/api/v1/anna-apps/dev/tokens/${id}/revoke`,
		allowStatuses: [204]
	});
}
async function listScopes(client) {
	const r = await client.request({ path: "/api/v1/anna-apps/dev/tokens/scopes" });
	return r.data?.scopes ?? [];
}

//#endregion
//#region src/commands/token.ts
async function runTokenList(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const tokens = await listTokens(client, opts.includeRevoked ?? false);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				tokens
			}, null, 2));
			return 0;
		}
		console.log(bold(cyan(`PATs for ${host}`)));
		if (tokens.length === 0) {
			console.log(dim(`  (no tokens — run \`anna-app login\` to mint one)`));
			return 0;
		}
		for (const t of tokens) {
			const scopes = Array.isArray(t.scopes) ? t.scopes.join(",") : t.scopes ?? "";
			const tag = t.revoked_at ? red("[revoked]") : green("[active]");
			const label = t.label ? ` "${t.label}"` : "";
			console.log(`  ${tag} #${t.id}${label} ${dim(`(scopes=[${scopes}], jti=${t.jti})`)}`);
		}
		return 0;
	});
}
async function runTokenRevoke(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		await revokeToken(client, opts.id);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				token_id: opts.id,
				revoked: true
			}, null, 2));
			return 0;
		}
		console.log(yellow(`✓ token #${opts.id} revoked on ${host}`));
		return 0;
	});
}
async function runTokenScopes(opts) {
	return withErrorHandling(async () => {
		const { client, host } = resolveClient({ account: opts.account });
		const scopes = await listScopes(client);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				scopes
			}, null, 2));
			return 0;
		}
		console.log(bold(cyan(`available PAT scopes`)) + " " + dim(`(${host})`));
		for (const s of scopes) console.log(`  ${green(s.name.padEnd(18))} ${dim(s.description)}`);
		return 0;
	});
}

//#endregion
export { runTokenList, runTokenRevoke, runTokenScopes };
