// @anna-ai/cli/dist/whoami-ChzwOaN2.js
import { credentialsAreLooselyPermissioned, maskPat, readCredentials } from "./credentials-BTv2IfUZ.js";

//#region src/commands/whoami.ts
async function runWhoami(opts = {}) {
	const data = readCredentials();
	if (Object.keys(data.accounts).length === 0) {
		if (opts.json) process.stdout.write(JSON.stringify({
			accounts: [],
			current: null
		}) + "\n");
		else console.log("(no accounts — run `anna-app login --host <nexus-url>`)");
		return 0;
	}
	if (opts.json) {
		const out = {
			current: data.current,
			accounts: Object.entries(data.accounts).map(([k, v]) => ({
				host: k,
				user_id: v.user_id,
				pat_preview: maskPat(v.pat),
				issued_at: v.issued_at,
				expires_at: v.expires_at,
				scopes: v.scopes
			}))
		};
		process.stdout.write(JSON.stringify(out, null, 2) + "\n");
		return 0;
	}
	if (credentialsAreLooselyPermissioned()) console.warn("⚠ ~/.config/anna/credentials.json has loose permissions; run `chmod 600` on it.");
	for (const [k, v] of Object.entries(data.accounts)) {
		const tag = k === data.current ? " (current)" : "";
		const expDays = Math.max(0, Math.round((v.expires_at - Date.now() / 1e3) / 86400));
		console.log(`• ${k}${tag}`);
		console.log(`    pat:        ${maskPat(v.pat)}`);
		console.log(`    user_id:    ${v.user_id ?? "(unknown)"}`);
		console.log(`    scopes:     ${v.scopes ?? "(none)"}`);
		console.log(`    expires_in: ~${expDays}d`);
	}
	return 0;
}

//#endregion
export { runWhoami };
