// @anna-ai/cli/dist/working-orchestration-CX0ONKIq.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { createApp, findAppBySlug, getApp } from "./apps-73pdBAFE.js";
import { CliError } from "./client-D-_z1ALk.js";
import { parseExecutaIdOverrides, readExecutasLock, substituteBundledRefs, validateBundledHandles, writeBundleToolIdSidecar, writeExecutasLock } from "./bundled-executas-_H1A9M50.js";
import { runExecutaPublish } from "./executa-publish-Xmm4DMO9.js";
import { loadExecutaManifest } from "./manifest-CMc7RM57.js";
import { runExecutaInstall } from "./executa-install-A-BgavhA.js";
import { appCacheMatches, readAppIdentity, writeAppIdentity } from "./app-cache-BqnTkxUw.js";
import { join, resolve } from "node:path";
import { dim, green, yellow } from "kleur/colors";
import { homedir } from "node:os";

//#region src/publish/working-orchestration.ts
async function resolveBundledExecutas(params) {
	const { manifest, cwd, host, dryRun = false, freeze, json = false } = params;
	const overrides = parseExecutaIdOverrides(params.executaId);
	const active = !params.noBundledExecutas && (manifest.bundledExecutas.length > 0 || overrides.length > 0);
	if (!active) return {
		publishManifestRaw: manifest.raw,
		resolved: {}
	};
	const { referenced } = validateBundledHandles(manifest.raw, manifest.bundledExecutas, overrides);
	const overrideMap = new Map(overrides.map((o) => [o.handle, o.toolId]));
	const lock = readExecutasLock(cwd, host);
	const resolved = {};
	const lockEntries = {};
	for (const decl of manifest.bundledExecutas) {
		if (overrideMap.has(decl.handle)) {
			const id = overrideMap.get(decl.handle);
			resolved[decl.handle] = id;
			lockEntries[decl.handle] = {
				tool_id: id,
				path: decl.path,
				source: "override"
			};
			if (!json) console.log(dim(`  bundled:${decl.handle} → ${id} (override)`));
			continue;
		}
		if (params.skipExecutaPublish) {
			const cachedId = lock?.executas[decl.handle]?.tool_id;
			if (!cachedId) throw new CliError(`--skip-executa-publish set but no cached tool_id for "${decl.handle}" in .anna/executas.lock.json (run without --skip-executa-publish first, or pass --executa-id ${decl.handle}=<tool-id>)`, 4);
			resolved[decl.handle] = cachedId;
			lockEntries[decl.handle] = {
				tool_id: cachedId,
				path: decl.path,
				source: "cache"
			};
			if (!json) console.log(dim(`  bundled:${decl.handle} → ${cachedId} (cache)`));
			continue;
		}
		const execCwd = resolve(cwd, decl.path);
		if (dryRun) {
			if (!json) console.log(yellow(`  [dry-run] would publish bundled executa "${decl.handle}" from ${decl.path}`));
			continue;
		}
		if (!json) console.log(dim(`  ${freeze ? "publishing" : "registering"} bundled executa "${decl.handle}" (${decl.path})${freeze ? "" : " [no-freeze]"}…`));
		let captured;
		const code = await runExecutaPublish({
			cwd: execCwd,
			account: params.account,
			quiet: true,
			freeze,
			capture: (r) => {
				captured = r;
			}
		});
		if (code !== 0 || !captured) throw new CliError(`failed to ${freeze ? "publish" : "register"} bundled executa "${decl.handle}" (${decl.path})`, code === 0 ? 4 : code);
		resolved[decl.handle] = captured.tool_id;
		lockEntries[decl.handle] = {
			tool_id: captured.tool_id,
			path: decl.path,
			source: params.lockSource ?? (freeze ? "published" : "working")
		};
		if (!json) console.log(green(`  ✓ bundled:${decl.handle} → ${captured.tool_id}`) + dim(` (v${captured.version}${captured.idempotent_hit ? ", unchanged" : ""})`));
	}
	for (const o of overrides) if (!(o.handle in resolved)) {
		resolved[o.handle] = o.toolId;
		lockEntries[o.handle] = {
			tool_id: o.toolId,
			source: "override"
		};
		if (!json) console.log(dim(`  bundled:${o.handle} → ${o.toolId} (override)`));
	}
	const missing = [...referenced].filter((h) => !(h in resolved));
	if (missing.length > 0 && !dryRun) throw new CliError(`could not resolve bundled handle(s): ${missing.sort().join(", ")}`, 4);
	let publishManifestRaw = manifest.raw;
	if (!dryRun) {
		const merged = {
			$schema: "anna-executas-lock/v1",
			host: canonicalHost(host),
			executas: {
				...lock?.executas ?? {},
				...lockEntries
			}
		};
		writeExecutasLock(cwd, merged);
		publishManifestRaw = substituteBundledRefs(manifest.raw, resolved);
		const bundleEntry = manifest.raw.ui?.bundle?.entry;
		if (bundleEntry && Object.keys(resolved).length > 0) {
			const sidecar = writeBundleToolIdSidecar(resolve(cwd, "bundle"), resolved);
			if (!json) console.log(dim(`  wrote ${sidecar.replace(cwd + "/", "")}`));
		}
	}
	return {
		publishManifestRaw,
		resolved
	};
}
/**
* Install a local-dev launcher shim for every bundled executa whose *active*
* distribution profile is `local`, under its minted `tool_id`. This closes
* the manual gap after `apps push`: a `local` executa never triggers a server
* `install_plugin` RPC, so the only way it turns "Running" on an Agent is the
* **Rediscover Local** flow — which needs an on-disk shim named exactly the
* minted `tool_id`. Binary/uv/npm/etc. profiles are skipped (the Agent
* installs those itself), as are skills (no runtime process) and dev-only
* executas with no distribution block.
*
* Failures are non-fatal: a shim that can't be written warns but does not
* fail the surrounding `apps push`.
*/
async function installLocalBundledShims(params) {
	const { manifest, cwd, resolved, json = false } = params;
	const binDir = params.binDir ?? join(homedir(), ".anna", "executa", "bin");
	const installed = [];
	for (const decl of manifest.bundledExecutas) {
		const toolId = resolved[decl.handle];
		if (!toolId) continue;
		const dir = resolve(cwd, decl.path);
		let exManifest;
		try {
			exManifest = loadExecutaManifest(dir);
		} catch {
			continue;
		}
		if (exManifest.distribution?.type !== "local") continue;
		const code = await runExecutaInstall({
			dir,
			toolId,
			binDir: params.binDir,
			force: true,
			quiet: true
		});
		if (code === 0) installed.push({
			handle: decl.handle,
			tool_id: toolId,
			shim_path: join(binDir, toolId)
		});
		else if (!json) console.log(yellow(`  ! could not install local dev shim for "${decl.handle}" (exit ${code})`));
	}
	return installed;
}
/**
* Resolve (creating if absent) the `AnnaApp` row, persisting the identity
* cache before any subsequent network call. Returns `null` only in dry-run
* mode when the app does not exist yet (nothing to create).
*/
async function resolveAppIdentity(params) {
	const { client, cwd, host, manifest, dryRun = false, json = false } = params;
	let cached = readAppIdentity(cwd);
	if (cached && cached.host !== canonicalHost(host)) cached = null;
	if (cached && appCacheMatches(cached, host, manifest.slug)) try {
		const live = await getApp(client, cached.app_id);
		return {
			appId: live.id,
			serverSlug: live.slug,
			firstPublish: false,
			fromCache: true
		};
	} catch (e) {
		if (!(e instanceof CliError && e.status === 404)) throw e;
		if (!json) console.log(yellow(`↻ cached app_id=${cached.app_id} (slug=${cached.slug}) no longer exists on ${host}; re-resolving by slug…`));
	}
	const found = await findAppBySlug(client, manifest.slug);
	let appId;
	let serverSlug;
	let serverHandle;
	let firstPublish = false;
	if (found) {
		appId = found.id;
		serverSlug = found.slug;
		serverHandle = typeof found.handle === "string" ? found.handle : void 0;
	} else {
		if (dryRun) {
			if (!json) console.log(yellow(`[dry-run] would create AnnaApp slug=${manifest.slug}`));
			return null;
		}
		let created;
		try {
			created = await createApp(client, {
				slug: manifest.slug,
				name: manifest.name,
				category: manifest.category,
				tagline: manifest.tagline,
				description: manifest.description
			});
		} catch (e) {
			if (e instanceof CliError && e.status === 409) {
				const body = e.body;
				const code = body?.detail?.code;
				if (code === "HANDLE_REQUIRED") throw new CliError("you must set a developer handle before creating an app.\n  Run:  anna-app account set-handle <handle>\n  Your apps will then publish as @<handle>/" + manifest.slug + ".", 5, {
					status: 409,
					body
				});
			}
			throw e;
		}
		appId = created.id;
		serverSlug = created.slug;
		serverHandle = typeof created.handle === "string" ? created.handle : void 0;
		firstPublish = true;
	}
	writeAppIdentity(cwd, {
		$schema: "anna-app-identity/v1",
		host: canonicalHost(host),
		app_id: appId,
		slug: serverSlug,
		handle: serverHandle,
		first_published_at: cached?.first_published_at ?? new Date().toISOString()
	});
	return {
		appId,
		serverSlug,
		firstPublish,
		fromCache: false
	};
}
/**
* Resolve an EXISTING app via an explicit `slug` or the local
* `.anna/app.json` identity cache (no create). Used by verbs that operate on
* an already-pushed app (`cut`, `discard`). Throws `CliError(…, 5)` when the
* app cannot be resolved.
*/
async function resolveAppBySlugOrCache(client, host, opts) {
	if (opts.slug) {
		const app = await findAppBySlug(client, opts.slug);
		if (!app) throw new CliError(`no app with slug "${opts.slug}" on ${host}`, 5);
		return getApp(client, app.id);
	}
	const cwd = resolve(opts.cwd ?? process.cwd());
	const cached = readAppIdentity(cwd);
	if (cached && cached.host === canonicalHost(host)) try {
		return await getApp(client, cached.app_id);
	} catch (e) {
		if (e instanceof CliError && e.status === 404) {
			const app = await findAppBySlug(client, cached.slug);
			if (app) return getApp(client, app.id);
		} else throw e;
	}
	throw new CliError("cannot resolve app — pass --slug <slug> or run from a project dir with a .anna/app.json cache (run `anna-app apps push` first)", 5);
}

//#endregion
export { installLocalBundledShims, resolveAppBySlugOrCache, resolveAppIdentity, resolveBundledExecutas };
